{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3b2b0324-6e05-46ef-9833-25348c1c3f06",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "from collections import defaultdict\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "48f0557d-f5dc-4788-9423-9f80cfa59b50",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_depth(assembly_id, min_depth=3):\n",
    "    reference_fasta = os.path.join(reference_genome_path, f'{assembly_id}.fasta')\n",
    "    depth_file = os.path.join(depth_file_path, f\"{assembly_id}.depth\")\n",
    "    \n",
    "    with open(reference_fasta, \"r\") as handle:\n",
    "        genome_length = 0\n",
    "        for record in SeqIO.parse(handle, \"fasta\"):\n",
    "            genome_length += len(record.seq)\n",
    "\n",
    "    genome_pos_count = 0\n",
    "    genome_totol_count = 0\n",
    "\n",
    "    with open(depth_file, \"r\") as depth:\n",
    "        for line in depth.readlines():\n",
    "\n",
    "            pos = int(line.split(\"\\t\")[1])\n",
    "            depth = int(line.strip().split(\"\\t\")[2])\n",
    "\n",
    "            if depth >= min_depth:\n",
    "                genome_pos_count += 1\n",
    "                genome_totol_count += depth\n",
    "\n",
    "    breadth_coverage = genome_pos_count/genome_length\n",
    "    depth_coverage = genome_totol_count/genome_pos_count\n",
    "    \n",
    "    return breadth_coverage, depth_coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "92daff8f-3cf6-4beb-b488-a1d993758b5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_directory = \"/home/Users/yl181/seqscreen_nano/ZymoBIOMICS.STD.Even.ont.minimap2\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cd9707e7-b2b2-4662-9c94-69ab53215c67",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_metadata = pd.read_csv(os.path.join(output_directory, 'reference_metadata.csv'))\n",
    "reference_genome_path = os.path.join(output_directory, 'reference_genomes')\n",
    "depth_file_path = os.path.join(output_directory, 'depth_files')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "522b42af-864d-41dd-8554-39bc90f7362a",
   "metadata": {},
   "outputs": [],
   "source": [
    "min_depth = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e80c840a-c137-44ce-9cd7-d6d855bbe27d",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = reference_metadata[reference_metadata['Downloaded']]\n",
    "\n",
    "breadth_coverage_list = []\n",
    "depth_coverage_list = []\n",
    "for assembly_id in downloaded_assemblies['Assembly Accession ID']:\n",
    "    breadth_coverage, depth_coverage = calculate_depth(assembly_id, min_depth=min_depth)\n",
    "    breadth_coverage_list.append(breadth_coverage)\n",
    "    depth_coverage_list.append(depth_coverage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "64f4c7d6-b483-4601-9677-5f54683367be",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies['Breadth Coverage'] = breadth_coverage_list\n",
    "downloaded_assemblies['Depth Coverage'] = depth_coverage_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "5f45e3a4-b818-42b0-9787-f3c43ac059a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies.to_csv(os.path.join(output_directory, 'coverage.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "ab759fef-9879-4397-ba3f-8cc6da1ff008",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Assembly Accession ID</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>Is Representative</th>\n",
       "      <th>Assembly Level</th>\n",
       "      <th>Organism of Assembly</th>\n",
       "      <th>Downloaded</th>\n",
       "      <th>Breadth Coverage</th>\n",
       "      <th>Depth Coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1280</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>GCF_000013425.1</td>\n",
       "      <td>University of Oklahoma Health Sciences Center</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>True</td>\n",
       "      <td>0.923425</td>\n",
       "      <td>28.811838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1613</td>\n",
       "      <td>Limosilactobacillus fermentum</td>\n",
       "      <td>GCF_022819245.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Limosilactobacillus fermentum</td>\n",
       "      <td>True</td>\n",
       "      <td>0.887391</td>\n",
       "      <td>37.477806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1351</td>\n",
       "      <td>Enterococcus faecalis</td>\n",
       "      <td>GCF_001598635.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Enterococcus faecalis</td>\n",
       "      <td>True</td>\n",
       "      <td>0.895596</td>\n",
       "      <td>28.436297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1423</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>GCF_000009045.1</td>\n",
       "      <td>BSNR</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>True</td>\n",
       "      <td>0.871755</td>\n",
       "      <td>28.929011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>287</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>GCF_000006765.1</td>\n",
       "      <td>PathoGenesis Corporation</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>True</td>\n",
       "      <td>0.969221</td>\n",
       "      <td>9.803418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5207</td>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>GCA_022832995.1</td>\n",
       "      <td>Broad Institute</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>True</td>\n",
       "      <td>0.434962</td>\n",
       "      <td>1.476958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1639</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "      <td>European Consortium</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Listeria monocytogenes EGD-e</td>\n",
       "      <td>True</td>\n",
       "      <td>0.941636</td>\n",
       "      <td>31.114196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1642</td>\n",
       "      <td>Listeria innocua</td>\n",
       "      <td>GCF_009648575.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Listeria innocua</td>\n",
       "      <td>True</td>\n",
       "      <td>0.909327</td>\n",
       "      <td>29.911235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>GCF_000008865.2</td>\n",
       "      <td>GIRC</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia coli O157:H7 str. Sakai</td>\n",
       "      <td>True</td>\n",
       "      <td>0.794427</td>\n",
       "      <td>21.601007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1638</td>\n",
       "      <td>Listeria ivanovii</td>\n",
       "      <td>GCF_000763515.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Listeria ivanovii subsp. ivanovii</td>\n",
       "      <td>True</td>\n",
       "      <td>0.849922</td>\n",
       "      <td>27.917540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>28901</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>GCF_000006945.2</td>\n",
       "      <td>Washington University Genome Sequencing Center</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>True</td>\n",
       "      <td>0.929836</td>\n",
       "      <td>21.580957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1392</td>\n",
       "      <td>Bacillus anthracis</td>\n",
       "      <td>GCF_000008445.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus anthracis str. 'Ames Ancestor'</td>\n",
       "      <td>True</td>\n",
       "      <td>0.080132</td>\n",
       "      <td>23.575129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>623</td>\n",
       "      <td>Shigella flexneri</td>\n",
       "      <td>GCF_000006925.2</td>\n",
       "      <td>Microbial Genome Center of ChMPH</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Shigella flexneri 2a str. 301</td>\n",
       "      <td>True</td>\n",
       "      <td>0.834539</td>\n",
       "      <td>22.208736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>GCF_000240185.1</td>\n",
       "      <td>Bioinformatics &amp; Microbial Genomics, State Ke...</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae HS11286</td>\n",
       "      <td>True</td>\n",
       "      <td>0.567063</td>\n",
       "      <td>18.940874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4932</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>GCF_000146045.2</td>\n",
       "      <td>Saccharomyces Genome Database</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Saccharomyces cerevisiae S288C</td>\n",
       "      <td>True</td>\n",
       "      <td>0.630432</td>\n",
       "      <td>1.926546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>96241</td>\n",
       "      <td>Bacillus spizizenii</td>\n",
       "      <td>GCF_006094475.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus subtilis subsp. spizizenii ATCC 6633 ...</td>\n",
       "      <td>True</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>29.770779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1590</td>\n",
       "      <td>Lactiplantibacillus plantarum</td>\n",
       "      <td>GCF_003269405.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Lactiplantibacillus plantarum</td>\n",
       "      <td>True</td>\n",
       "      <td>0.045929</td>\n",
       "      <td>43.137614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1352</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>GCF_009734005.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>True</td>\n",
       "      <td>0.232760</td>\n",
       "      <td>18.738459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>294</td>\n",
       "      <td>Pseudomonas fluorescens</td>\n",
       "      <td>GCF_000730425.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pseudomonas fluorescens</td>\n",
       "      <td>True</td>\n",
       "      <td>0.460645</td>\n",
       "      <td>7.494233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1643</td>\n",
       "      <td>Listeria welshimeri</td>\n",
       "      <td>GCF_900187315.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Listeria welshimeri</td>\n",
       "      <td>True</td>\n",
       "      <td>0.886593</td>\n",
       "      <td>29.402895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>176275</td>\n",
       "      <td>Beauveria bassiana</td>\n",
       "      <td>GCA_014607475.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Beauveria bassiana</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000996</td>\n",
       "      <td>17.986237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Taxonomy ID                        Species Assembly Accession ID  \\\n",
       "0          1280          Staphylococcus aureus       GCF_000013425.1   \n",
       "1          1613  Limosilactobacillus fermentum       GCF_022819245.1   \n",
       "2          1351          Enterococcus faecalis       GCF_001598635.1   \n",
       "3          1423              Bacillus subtilis       GCF_000009045.1   \n",
       "4           287         Pseudomonas aeruginosa       GCF_000006765.1   \n",
       "5          5207        Cryptococcus neoformans       GCA_022832995.1   \n",
       "6          1639         Listeria monocytogenes       GCF_000196035.1   \n",
       "7          1642               Listeria innocua       GCF_009648575.1   \n",
       "8           562               Escherichia coli       GCF_000008865.2   \n",
       "9          1638              Listeria ivanovii       GCF_000763515.1   \n",
       "10        28901            Salmonella enterica       GCF_000006945.2   \n",
       "11         1392             Bacillus anthracis       GCF_000008445.1   \n",
       "12          623              Shigella flexneri       GCF_000006925.2   \n",
       "13          573          Klebsiella pneumoniae       GCF_000240185.1   \n",
       "14         4932       Saccharomyces cerevisiae       GCF_000146045.2   \n",
       "15        96241            Bacillus spizizenii       GCF_006094475.1   \n",
       "16         1590  Lactiplantibacillus plantarum       GCF_003269405.1   \n",
       "17         1352           Enterococcus faecium       GCF_009734005.1   \n",
       "18          294        Pseudomonas fluorescens       GCF_000730425.1   \n",
       "19         1643            Listeria welshimeri       GCF_900187315.1   \n",
       "20       176275             Beauveria bassiana       GCA_014607475.1   \n",
       "\n",
       "                                      Source Database  Is Representative  \\\n",
       "0       University of Oklahoma Health Sciences Center               True   \n",
       "1                                         NCBI RefSeq               True   \n",
       "2                                         NCBI RefSeq              False   \n",
       "3                                                BSNR               True   \n",
       "4                            PathoGenesis Corporation               True   \n",
       "5                                     Broad Institute              False   \n",
       "6                                 European Consortium               True   \n",
       "7                                         NCBI RefSeq               True   \n",
       "8                                                GIRC               True   \n",
       "9                                         NCBI RefSeq               True   \n",
       "10     Washington University Genome Sequencing Center               True   \n",
       "11                                        NCBI RefSeq               True   \n",
       "12                   Microbial Genome Center of ChMPH               True   \n",
       "13   Bioinformatics & Microbial Genomics, State Ke...               True   \n",
       "14                      Saccharomyces Genome Database               True   \n",
       "15                                        NCBI RefSeq               True   \n",
       "16                                        NCBI RefSeq               True   \n",
       "17                                        NCBI RefSeq               True   \n",
       "18                                        NCBI RefSeq              False   \n",
       "19                                        NCBI RefSeq               True   \n",
       "20                                                NaN              False   \n",
       "\n",
       "     Assembly Level                               Organism of Assembly  \\\n",
       "0   Complete Genome      Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "1   Complete Genome                      Limosilactobacillus fermentum   \n",
       "2   Complete Genome                              Enterococcus faecalis   \n",
       "3   Complete Genome         Bacillus subtilis subsp. subtilis str. 168   \n",
       "4   Complete Genome                        Pseudomonas aeruginosa PAO1   \n",
       "5   Complete Genome                            Cryptococcus neoformans   \n",
       "6   Complete Genome                       Listeria monocytogenes EGD-e   \n",
       "7   Complete Genome                                   Listeria innocua   \n",
       "8   Complete Genome                Escherichia coli O157:H7 str. Sakai   \n",
       "9   Complete Genome                  Listeria ivanovii subsp. ivanovii   \n",
       "10  Complete Genome  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "11  Complete Genome            Bacillus anthracis str. 'Ames Ancestor'   \n",
       "12  Complete Genome                      Shigella flexneri 2a str. 301   \n",
       "13  Complete Genome    Klebsiella pneumoniae subsp. pneumoniae HS11286   \n",
       "14  Complete Genome                     Saccharomyces cerevisiae S288C   \n",
       "15  Complete Genome  Bacillus subtilis subsp. spizizenii ATCC 6633 ...   \n",
       "16  Complete Genome                      Lactiplantibacillus plantarum   \n",
       "17  Complete Genome                               Enterococcus faecium   \n",
       "18  Complete Genome                            Pseudomonas fluorescens   \n",
       "19  Complete Genome                                Listeria welshimeri   \n",
       "20  Complete Genome                                 Beauveria bassiana   \n",
       "\n",
       "    Downloaded  Breadth Coverage  Depth Coverage  \n",
       "0         True          0.923425       28.811838  \n",
       "1         True          0.887391       37.477806  \n",
       "2         True          0.895596       28.436297  \n",
       "3         True          0.871755       28.929011  \n",
       "4         True          0.969221        9.803418  \n",
       "5         True          0.434962        1.476958  \n",
       "6         True          0.941636       31.114196  \n",
       "7         True          0.909327       29.911235  \n",
       "8         True          0.794427       21.601007  \n",
       "9         True          0.849922       27.917540  \n",
       "10        True          0.929836       21.580957  \n",
       "11        True          0.080132       23.575129  \n",
       "12        True          0.834539       22.208736  \n",
       "13        True          0.567063       18.940874  \n",
       "14        True          0.630432        1.926546  \n",
       "15        True          1.000000       29.770779  \n",
       "16        True          0.045929       43.137614  \n",
       "17        True          0.232760       18.738459  \n",
       "18        True          0.460645        7.494233  \n",
       "19        True          0.886593       29.402895  \n",
       "20        True          0.000996       17.986237  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "downloaded_assemblies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "05e571c8-90cd-41b6-90b2-b35a9f471a91",
   "metadata": {},
   "outputs": [],
   "source": [
    "zymo_theoretical_abundance = dict()\n",
    "with open('/home/Users/yl181/seqscreen_nano/ZymoBIOMICS.STD.refseq.v2/theoretical_composition.txt', 'r') as ground_truth_f:\n",
    "    for line in ground_truth_f.readlines():\n",
    "        tax_name = line.strip().split(\" - \")[0]\n",
    "        abundance = float(line.strip().split(\" - \")[1])/100\n",
    "        zymo_theoretical_abundance[tax_name] = abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "90229db1-5e09-48d0-a30c-4b681da47636",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Listeria monocytogenes': 0.12,\n",
       " 'Pseudomonas aeruginosa': 0.12,\n",
       " 'Bacillus subtilis': 0.12,\n",
       " 'Escherichia coli': 0.12,\n",
       " 'Salmonella enterica': 0.12,\n",
       " 'Limosilactobacillus fermentum': 0.12,\n",
       " 'Enterococcus faecalis': 0.12,\n",
       " 'Staphylococcus aureus': 0.12,\n",
       " 'Saccharomyces cerevisiae': 0.02,\n",
       " 'Cryptococcus neoformans': 0.02}"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zymo_theoretical_abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "751ff257-6a25-445a-a034-fc71d10d884e",
   "metadata": {},
   "outputs": [],
   "source": [
    "theoretical_abundance = []\n",
    "labels = []\n",
    "for taxname in downloaded_assemblies['Species']:\n",
    "    try:\n",
    "        theoretical_abundance.append(zymo_theoretical_abundance[taxname])\n",
    "        labels.append(\"TP\")\n",
    "    except KeyError:\n",
    "        theoretical_abundance.append(0)\n",
    "        labels.append(\"FP\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "6c80895a-0f8d-4f69-ac7b-a11ee03690ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies['Theoretical Abundance'] = theoretical_abundance\n",
    "downloaded_assemblies['Labels'] = labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "e311cf1d-b839-4cf7-a1e0-2b53d83505df",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = downloaded_assemblies[['Taxonomy ID', 'Species', 'Breadth Coverage', 'Depth Coverage', 'Theoretical Abundance', 'Labels']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "bd8f7a2c-6733-40f5-89d1-215d1a98ce26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Breadth Coverage</th>\n",
       "      <th>Depth Coverage</th>\n",
       "      <th>Theoretical Abundance</th>\n",
       "      <th>Labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>96241</td>\n",
       "      <td>Bacillus spizizenii</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>29.770779</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>287</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>0.969221</td>\n",
       "      <td>9.803418</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1639</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>0.941636</td>\n",
       "      <td>31.114196</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>28901</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>0.929836</td>\n",
       "      <td>21.580957</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1280</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>0.923425</td>\n",
       "      <td>28.811838</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1642</td>\n",
       "      <td>Listeria innocua</td>\n",
       "      <td>0.909327</td>\n",
       "      <td>29.911235</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1351</td>\n",
       "      <td>Enterococcus faecalis</td>\n",
       "      <td>0.895596</td>\n",
       "      <td>28.436297</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1613</td>\n",
       "      <td>Limosilactobacillus fermentum</td>\n",
       "      <td>0.887391</td>\n",
       "      <td>37.477806</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1643</td>\n",
       "      <td>Listeria welshimeri</td>\n",
       "      <td>0.886593</td>\n",
       "      <td>29.402895</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1423</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>0.871755</td>\n",
       "      <td>28.929011</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1638</td>\n",
       "      <td>Listeria ivanovii</td>\n",
       "      <td>0.849922</td>\n",
       "      <td>27.917540</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>623</td>\n",
       "      <td>Shigella flexneri</td>\n",
       "      <td>0.834539</td>\n",
       "      <td>22.208736</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>0.794427</td>\n",
       "      <td>21.601007</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4932</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>0.630432</td>\n",
       "      <td>1.926546</td>\n",
       "      <td>0.02</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>0.567063</td>\n",
       "      <td>18.940874</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>294</td>\n",
       "      <td>Pseudomonas fluorescens</td>\n",
       "      <td>0.460645</td>\n",
       "      <td>7.494233</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5207</td>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>0.434962</td>\n",
       "      <td>1.476958</td>\n",
       "      <td>0.02</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1352</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>0.232760</td>\n",
       "      <td>18.738459</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1392</td>\n",
       "      <td>Bacillus anthracis</td>\n",
       "      <td>0.080132</td>\n",
       "      <td>23.575129</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1590</td>\n",
       "      <td>Lactiplantibacillus plantarum</td>\n",
       "      <td>0.045929</td>\n",
       "      <td>43.137614</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>176275</td>\n",
       "      <td>Beauveria bassiana</td>\n",
       "      <td>0.000996</td>\n",
       "      <td>17.986237</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Taxonomy ID                        Species  Breadth Coverage  \\\n",
       "15        96241            Bacillus spizizenii          1.000000   \n",
       "4           287         Pseudomonas aeruginosa          0.969221   \n",
       "6          1639         Listeria monocytogenes          0.941636   \n",
       "10        28901            Salmonella enterica          0.929836   \n",
       "0          1280          Staphylococcus aureus          0.923425   \n",
       "7          1642               Listeria innocua          0.909327   \n",
       "2          1351          Enterococcus faecalis          0.895596   \n",
       "1          1613  Limosilactobacillus fermentum          0.887391   \n",
       "19         1643            Listeria welshimeri          0.886593   \n",
       "3          1423              Bacillus subtilis          0.871755   \n",
       "9          1638              Listeria ivanovii          0.849922   \n",
       "12          623              Shigella flexneri          0.834539   \n",
       "8           562               Escherichia coli          0.794427   \n",
       "14         4932       Saccharomyces cerevisiae          0.630432   \n",
       "13          573          Klebsiella pneumoniae          0.567063   \n",
       "18          294        Pseudomonas fluorescens          0.460645   \n",
       "5          5207        Cryptococcus neoformans          0.434962   \n",
       "17         1352           Enterococcus faecium          0.232760   \n",
       "11         1392             Bacillus anthracis          0.080132   \n",
       "16         1590  Lactiplantibacillus plantarum          0.045929   \n",
       "20       176275             Beauveria bassiana          0.000996   \n",
       "\n",
       "    Depth Coverage  Theoretical Abundance Labels  \n",
       "15       29.770779                   0.00     FP  \n",
       "4         9.803418                   0.12     TP  \n",
       "6        31.114196                   0.12     TP  \n",
       "10       21.580957                   0.12     TP  \n",
       "0        28.811838                   0.12     TP  \n",
       "7        29.911235                   0.00     FP  \n",
       "2        28.436297                   0.12     TP  \n",
       "1        37.477806                   0.12     TP  \n",
       "19       29.402895                   0.00     FP  \n",
       "3        28.929011                   0.12     TP  \n",
       "9        27.917540                   0.00     FP  \n",
       "12       22.208736                   0.00     FP  \n",
       "8        21.601007                   0.12     TP  \n",
       "14        1.926546                   0.02     TP  \n",
       "13       18.940874                   0.00     FP  \n",
       "18        7.494233                   0.00     FP  \n",
       "5         1.476958                   0.02     TP  \n",
       "17       18.738459                   0.00     FP  \n",
       "11       23.575129                   0.00     FP  \n",
       "16       43.137614                   0.00     FP  \n",
       "20       17.986237                   0.00     FP  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df.sort_values(['Breadth Coverage', 'Theoretical Abundance'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "d4300b16-df2c-4139-aed6-dfc7071dde4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Breadth Coverage</th>\n",
       "      <th>Depth Coverage</th>\n",
       "      <th>Theoretical Abundance</th>\n",
       "      <th>Labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1280</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>0.923425</td>\n",
       "      <td>28.811838</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>623</td>\n",
       "      <td>Shigella flexneri</td>\n",
       "      <td>0.834539</td>\n",
       "      <td>22.208736</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>28901</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>0.929836</td>\n",
       "      <td>21.580957</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4932</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>0.630432</td>\n",
       "      <td>1.926546</td>\n",
       "      <td>0.02</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>294</td>\n",
       "      <td>Pseudomonas fluorescens</td>\n",
       "      <td>0.460645</td>\n",
       "      <td>7.494233</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>287</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>0.969221</td>\n",
       "      <td>9.803418</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1643</td>\n",
       "      <td>Listeria welshimeri</td>\n",
       "      <td>0.886593</td>\n",
       "      <td>29.402895</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1639</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>0.941636</td>\n",
       "      <td>31.114196</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1638</td>\n",
       "      <td>Listeria ivanovii</td>\n",
       "      <td>0.849922</td>\n",
       "      <td>27.917540</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1642</td>\n",
       "      <td>Listeria innocua</td>\n",
       "      <td>0.909327</td>\n",
       "      <td>29.911235</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1613</td>\n",
       "      <td>Limosilactobacillus fermentum</td>\n",
       "      <td>0.887391</td>\n",
       "      <td>37.477806</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1590</td>\n",
       "      <td>Lactiplantibacillus plantarum</td>\n",
       "      <td>0.045929</td>\n",
       "      <td>43.137614</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>0.567063</td>\n",
       "      <td>18.940874</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>0.794427</td>\n",
       "      <td>21.601007</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1352</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>0.232760</td>\n",
       "      <td>18.738459</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1351</td>\n",
       "      <td>Enterococcus faecalis</td>\n",
       "      <td>0.895596</td>\n",
       "      <td>28.436297</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5207</td>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>0.434962</td>\n",
       "      <td>1.476958</td>\n",
       "      <td>0.02</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>176275</td>\n",
       "      <td>Beauveria bassiana</td>\n",
       "      <td>0.000996</td>\n",
       "      <td>17.986237</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1423</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>0.871755</td>\n",
       "      <td>28.929011</td>\n",
       "      <td>0.12</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>96241</td>\n",
       "      <td>Bacillus spizizenii</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>29.770779</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1392</td>\n",
       "      <td>Bacillus anthracis</td>\n",
       "      <td>0.080132</td>\n",
       "      <td>23.575129</td>\n",
       "      <td>0.00</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Taxonomy ID                        Species  Breadth Coverage  \\\n",
       "0          1280          Staphylococcus aureus          0.923425   \n",
       "12          623              Shigella flexneri          0.834539   \n",
       "10        28901            Salmonella enterica          0.929836   \n",
       "14         4932       Saccharomyces cerevisiae          0.630432   \n",
       "18          294        Pseudomonas fluorescens          0.460645   \n",
       "4           287         Pseudomonas aeruginosa          0.969221   \n",
       "19         1643            Listeria welshimeri          0.886593   \n",
       "6          1639         Listeria monocytogenes          0.941636   \n",
       "9          1638              Listeria ivanovii          0.849922   \n",
       "7          1642               Listeria innocua          0.909327   \n",
       "1          1613  Limosilactobacillus fermentum          0.887391   \n",
       "16         1590  Lactiplantibacillus plantarum          0.045929   \n",
       "13          573          Klebsiella pneumoniae          0.567063   \n",
       "8           562               Escherichia coli          0.794427   \n",
       "17         1352           Enterococcus faecium          0.232760   \n",
       "2          1351          Enterococcus faecalis          0.895596   \n",
       "5          5207        Cryptococcus neoformans          0.434962   \n",
       "20       176275             Beauveria bassiana          0.000996   \n",
       "3          1423              Bacillus subtilis          0.871755   \n",
       "15        96241            Bacillus spizizenii          1.000000   \n",
       "11         1392             Bacillus anthracis          0.080132   \n",
       "\n",
       "    Depth Coverage  Theoretical Abundance Labels  \n",
       "0        28.811838                   0.12     TP  \n",
       "12       22.208736                   0.00     FP  \n",
       "10       21.580957                   0.12     TP  \n",
       "14        1.926546                   0.02     TP  \n",
       "18        7.494233                   0.00     FP  \n",
       "4         9.803418                   0.12     TP  \n",
       "19       29.402895                   0.00     FP  \n",
       "6        31.114196                   0.12     TP  \n",
       "9        27.917540                   0.00     FP  \n",
       "7        29.911235                   0.00     FP  \n",
       "1        37.477806                   0.12     TP  \n",
       "16       43.137614                   0.00     FP  \n",
       "13       18.940874                   0.00     FP  \n",
       "8        21.601007                   0.12     TP  \n",
       "17       18.738459                   0.00     FP  \n",
       "2        28.436297                   0.12     TP  \n",
       "5         1.476958                   0.02     TP  \n",
       "20       17.986237                   0.00     FP  \n",
       "3        28.929011                   0.12     TP  \n",
       "15       29.770779                   0.00     FP  \n",
       "11       23.575129                   0.00     FP  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df.sort_values(['Species', 'Theoretical Abundance'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "5564542f-e1bb-4593-86d4-d31f70da4be1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for taxname in zymo_theoretical_abundance:\n",
    "    if taxname not in set(downloaded_assemblies['Species']):\n",
    "        print(taxname)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "c3d9af41-9ba8-4f3c-a37e-21a84941db75",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = result_df.sort_values(['Breadth Coverage', 'Theoretical Abundance'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "ce302cc7-d194-4e07-9c4c-731e53de4d42",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp[temp['Breadth Coverage'] >= 0.5].to_csv(os.path.join(output_directory, 'coverage_above_50.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f03a50d-1cf0-4d35-8fbd-30ea9e5ce3c9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
