{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "63e057ac-51aa-4a96-9ede-b76fe7ba7407",
   "metadata": {},
   "source": [
    "# Dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "251cef2a-37b5-46e6-9fda-4546d1c14315",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import argparse\n",
    "import pathlib\n",
    "import sys\n",
    "import subprocess\n",
    "import math\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "from multiprocessing import Pool, Manager\n",
    "from itertools import repeat\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import pandas as pd\n",
    "from ete3 import NCBITaxa\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3911e5c3-e6b4-41f6-a06b-0fc2337e0e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.reference_finder import prepare_reference_genomes\n",
    "from utils.alignment import run_minimap2, run_bwa, sort_samfile, samtools_calculate_coverage\n",
    "from utils.summary import alignment_summary, merge_reference_fasta, call_present_absent\n",
    "from utils.ani import samtools_merged_consensus, ani_summary\n",
    "from utils.input_parsing import parsing_input_f, filter_input_df, get_seq2assembly_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f88611f9-3109-4a1f-812d-a03f3de37519",
   "metadata": {},
   "source": [
    "# Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "180315a5-b912-4a11-8706-8fb35d20d185",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = pd.read_csv('/home/Users/yl181/magnet/zymo_bowtie2_a_k2_0.002/magnet_results.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7266ce2a-5c8e-48e8-b414-5f3d46f5d087",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = call_present_absent(result_df, min_coverage_score=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6b77b29b-54b2-49a1-b7e2-e8e61513585e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Assembly Accession ID</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>Is Representative</th>\n",
       "      <th>Assembly Level</th>\n",
       "      <th>Organism of Assembly</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Total Length</th>\n",
       "      <th>Downloaded</th>\n",
       "      <th>Species</th>\n",
       "      <th>...</th>\n",
       "      <th>Secondary Expected</th>\n",
       "      <th>Secondary Score</th>\n",
       "      <th>Secondary Depth</th>\n",
       "      <th>Primary Breadth</th>\n",
       "      <th>Primary Expected</th>\n",
       "      <th>Primary Score</th>\n",
       "      <th>Primary Depth</th>\n",
       "      <th>Consensus ANI</th>\n",
       "      <th>Combined PS and ANI (Sqrt(ANI)xPSx100)</th>\n",
       "      <th>Presence/Absence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>96241</td>\n",
       "      <td>GCF_006094475.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus spizizenii ATCC 6633 = JCM 2499</td>\n",
       "      <td>ATCC 6633</td>\n",
       "      <td>4045538.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus spizizenii</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>132.511000</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>68.478268</td>\n",
       "      <td>0.999977</td>\n",
       "      <td>100.00</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>287</td>\n",
       "      <td>GCF_000006765.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>6264404.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.967680</td>\n",
       "      <td>65.565285</td>\n",
       "      <td>0.967013</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.967013</td>\n",
       "      <td>56.030159</td>\n",
       "      <td>0.994526</td>\n",
       "      <td>96.44</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28901</td>\n",
       "      <td>GCF_000006945.2</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>LT2</td>\n",
       "      <td>4951383.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.935623</td>\n",
       "      <td>99.628887</td>\n",
       "      <td>0.928550</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.928550</td>\n",
       "      <td>64.813931</td>\n",
       "      <td>0.989556</td>\n",
       "      <td>92.37</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1280</td>\n",
       "      <td>GCF_000013425.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>NCTC 8325</td>\n",
       "      <td>2821361.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.921382</td>\n",
       "      <td>160.315715</td>\n",
       "      <td>0.917263</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.917263</td>\n",
       "      <td>83.129377</td>\n",
       "      <td>0.993009</td>\n",
       "      <td>91.41</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1639</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Listeria monocytogenes EGD-e</td>\n",
       "      <td>EGD-e</td>\n",
       "      <td>2944528.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.931354</td>\n",
       "      <td>129.529645</td>\n",
       "      <td>0.911952</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.911952</td>\n",
       "      <td>77.818246</td>\n",
       "      <td>0.947604</td>\n",
       "      <td>88.77</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1351</td>\n",
       "      <td>GCF_001598635.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Enterococcus faecalis</td>\n",
       "      <td>LD33</td>\n",
       "      <td>2803429.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Enterococcus faecalis</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.890446</td>\n",
       "      <td>132.768305</td>\n",
       "      <td>0.889125</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.889125</td>\n",
       "      <td>101.940818</td>\n",
       "      <td>0.991167</td>\n",
       "      <td>88.52</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1613</td>\n",
       "      <td>GCF_029961225.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Limosilactobacillus fermentum</td>\n",
       "      <td>EFEL6800</td>\n",
       "      <td>2103331.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Limosilactobacillus fermentum</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.892914</td>\n",
       "      <td>212.556811</td>\n",
       "      <td>0.885748</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.885748</td>\n",
       "      <td>136.140257</td>\n",
       "      <td>0.993539</td>\n",
       "      <td>88.29</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>623</td>\n",
       "      <td>GCF_000006925.2</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Shigella flexneri 2a str. 301</td>\n",
       "      <td>301</td>\n",
       "      <td>4828820.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Shigella flexneri</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.889687</td>\n",
       "      <td>114.269365</td>\n",
       "      <td>0.786340</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.786340</td>\n",
       "      <td>32.865802</td>\n",
       "      <td>0.980880</td>\n",
       "      <td>77.88</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>562</td>\n",
       "      <td>GCF_000008865.2</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia coli O157:H7 str. Sakai</td>\n",
       "      <td>Sakai substr. RIMD 0509952</td>\n",
       "      <td>5594605.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.808360</td>\n",
       "      <td>97.632873</td>\n",
       "      <td>0.712230</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.712230</td>\n",
       "      <td>33.563861</td>\n",
       "      <td>0.977177</td>\n",
       "      <td>70.41</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1642</td>\n",
       "      <td>GCF_009648575.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Listeria innocua</td>\n",
       "      <td>CFSAN044836</td>\n",
       "      <td>2922148.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Listeria innocua</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.862307</td>\n",
       "      <td>111.905010</td>\n",
       "      <td>0.536577</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.536578</td>\n",
       "      <td>25.403433</td>\n",
       "      <td>0.908522</td>\n",
       "      <td>51.14</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>72361</td>\n",
       "      <td>GCF_003667885.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus vallismortis</td>\n",
       "      <td>Bac111</td>\n",
       "      <td>3958372.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus vallismortis</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.881393</td>\n",
       "      <td>104.355670</td>\n",
       "      <td>0.203398</td>\n",
       "      <td>0.449790</td>\n",
       "      <td>0.452206</td>\n",
       "      <td>2.937331</td>\n",
       "      <td>0.931382</td>\n",
       "      <td>43.64</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1654388</td>\n",
       "      <td>GCF_900636685.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Staphylococcus schweitzeri</td>\n",
       "      <td>NCTC13712</td>\n",
       "      <td>2784939.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Staphylococcus schweitzeri</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.861931</td>\n",
       "      <td>144.854998</td>\n",
       "      <td>0.430713</td>\n",
       "      <td>0.994473</td>\n",
       "      <td>0.433107</td>\n",
       "      <td>12.068378</td>\n",
       "      <td>0.916871</td>\n",
       "      <td>41.47</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>483913</td>\n",
       "      <td>GCF_023238325.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus inaquosorum</td>\n",
       "      <td>A65.1</td>\n",
       "      <td>4242130.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus inaquosorum</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.861323</td>\n",
       "      <td>105.754378</td>\n",
       "      <td>0.223387</td>\n",
       "      <td>0.543606</td>\n",
       "      <td>0.410936</td>\n",
       "      <td>3.511335</td>\n",
       "      <td>0.937964</td>\n",
       "      <td>39.80</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>260554</td>\n",
       "      <td>GCF_004006435.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus halotolerans</td>\n",
       "      <td>ZB201702</td>\n",
       "      <td>4154245.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus halotolerans</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.851494</td>\n",
       "      <td>90.357620</td>\n",
       "      <td>0.143939</td>\n",
       "      <td>0.359235</td>\n",
       "      <td>0.400682</td>\n",
       "      <td>3.092184</td>\n",
       "      <td>0.902012</td>\n",
       "      <td>38.05</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>227866</td>\n",
       "      <td>GCF_012225885.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus tequilensis</td>\n",
       "      <td>EA-CB0015</td>\n",
       "      <td>4012371.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus tequilensis</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.849856</td>\n",
       "      <td>105.969970</td>\n",
       "      <td>0.183283</td>\n",
       "      <td>0.458545</td>\n",
       "      <td>0.399705</td>\n",
       "      <td>3.347207</td>\n",
       "      <td>0.930512</td>\n",
       "      <td>38.56</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>72360</td>\n",
       "      <td>GCF_012648005.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus mojavensis</td>\n",
       "      <td>UCMB5075</td>\n",
       "      <td>4031121.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus mojavensis</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.858009</td>\n",
       "      <td>90.663038</td>\n",
       "      <td>0.128208</td>\n",
       "      <td>0.328128</td>\n",
       "      <td>0.390727</td>\n",
       "      <td>3.101836</td>\n",
       "      <td>0.902625</td>\n",
       "      <td>37.12</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1423</td>\n",
       "      <td>GCF_000009045.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>168</td>\n",
       "      <td>4215606.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.862585</td>\n",
       "      <td>102.397199</td>\n",
       "      <td>0.193634</td>\n",
       "      <td>0.510101</td>\n",
       "      <td>0.379599</td>\n",
       "      <td>3.685022</td>\n",
       "      <td>0.930759</td>\n",
       "      <td>36.62</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>573</td>\n",
       "      <td>GCF_000240185.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae HS11286</td>\n",
       "      <td>HS11286</td>\n",
       "      <td>5682322.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.426917</td>\n",
       "      <td>71.521912</td>\n",
       "      <td>0.088700</td>\n",
       "      <td>0.236148</td>\n",
       "      <td>0.375613</td>\n",
       "      <td>3.036957</td>\n",
       "      <td>0.882375</td>\n",
       "      <td>35.28</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2892442</td>\n",
       "      <td>GCF_020911845.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pseudomonas sp. CIP-10</td>\n",
       "      <td>CIP-10</td>\n",
       "      <td>6473352.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Pseudomonas sp. CIP-10</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.301415</td>\n",
       "      <td>52.264418</td>\n",
       "      <td>0.056747</td>\n",
       "      <td>0.156827</td>\n",
       "      <td>0.361844</td>\n",
       "      <td>3.006012</td>\n",
       "      <td>0.877682</td>\n",
       "      <td>33.90</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>54736</td>\n",
       "      <td>GCF_000439255.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Salmonella bongori N268-08</td>\n",
       "      <td>N268-08</td>\n",
       "      <td>4773537.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Salmonella bongori</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.805211</td>\n",
       "      <td>84.664805</td>\n",
       "      <td>0.274417</td>\n",
       "      <td>0.790406</td>\n",
       "      <td>0.347185</td>\n",
       "      <td>5.694112</td>\n",
       "      <td>0.913124</td>\n",
       "      <td>33.18</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1311</td>\n",
       "      <td>GCF_001552035.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Streptococcus agalactiae</td>\n",
       "      <td>NGBS128</td>\n",
       "      <td>2079123.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Streptococcus agalactiae</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.055425</td>\n",
       "      <td>772.158157</td>\n",
       "      <td>0.018327</td>\n",
       "      <td>0.055622</td>\n",
       "      <td>0.329498</td>\n",
       "      <td>3.122534</td>\n",
       "      <td>0.869289</td>\n",
       "      <td>30.72</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>208962</td>\n",
       "      <td>GCF_016904755.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia albertii</td>\n",
       "      <td>Sample 167</td>\n",
       "      <td>4631903.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia albertii</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.800222</td>\n",
       "      <td>92.606287</td>\n",
       "      <td>0.200037</td>\n",
       "      <td>0.608517</td>\n",
       "      <td>0.328729</td>\n",
       "      <td>4.688133</td>\n",
       "      <td>0.904501</td>\n",
       "      <td>31.26</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1396</td>\n",
       "      <td>GCF_002220285.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus cereus</td>\n",
       "      <td>FORC_047</td>\n",
       "      <td>5836971.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus cereus</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.053919</td>\n",
       "      <td>641.472836</td>\n",
       "      <td>0.017304</td>\n",
       "      <td>0.054238</td>\n",
       "      <td>0.319038</td>\n",
       "      <td>3.222602</td>\n",
       "      <td>0.870186</td>\n",
       "      <td>29.76</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>164546</td>\n",
       "      <td>GCF_022406855.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Cupriavidus taiwanensis</td>\n",
       "      <td>SWF66294</td>\n",
       "      <td>6782863.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Cupriavidus taiwanensis</td>\n",
       "      <td>...</td>\n",
       "      <td>0.997388</td>\n",
       "      <td>0.081080</td>\n",
       "      <td>73.546897</td>\n",
       "      <td>0.014609</td>\n",
       "      <td>0.047258</td>\n",
       "      <td>0.309141</td>\n",
       "      <td>3.313685</td>\n",
       "      <td>0.875291</td>\n",
       "      <td>28.92</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2044467</td>\n",
       "      <td>GCF_005843885.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia sp. E4742</td>\n",
       "      <td>E4742</td>\n",
       "      <td>5120753.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia sp. E4742</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.770260</td>\n",
       "      <td>108.055180</td>\n",
       "      <td>0.274972</td>\n",
       "      <td>0.927370</td>\n",
       "      <td>0.296508</td>\n",
       "      <td>9.536773</td>\n",
       "      <td>0.915722</td>\n",
       "      <td>28.37</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1499973</td>\n",
       "      <td>GCF_002900365.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia marmotae</td>\n",
       "      <td>HT073016</td>\n",
       "      <td>4896291.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia marmotae</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.813909</td>\n",
       "      <td>95.923526</td>\n",
       "      <td>0.220992</td>\n",
       "      <td>0.757247</td>\n",
       "      <td>0.291836</td>\n",
       "      <td>6.406079</td>\n",
       "      <td>0.912023</td>\n",
       "      <td>27.87</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>492670</td>\n",
       "      <td>GCF_000769555.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus velezensis</td>\n",
       "      <td>JS25R</td>\n",
       "      <td>4014440.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus velezensis</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.390627</td>\n",
       "      <td>91.963779</td>\n",
       "      <td>0.073708</td>\n",
       "      <td>0.255968</td>\n",
       "      <td>0.287956</td>\n",
       "      <td>4.011359</td>\n",
       "      <td>0.883434</td>\n",
       "      <td>27.07</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>564</td>\n",
       "      <td>GCF_020097475.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia fergusonii</td>\n",
       "      <td>FDAARGOS_1499</td>\n",
       "      <td>4645928.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia fergusonii</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.699085</td>\n",
       "      <td>89.949381</td>\n",
       "      <td>0.257754</td>\n",
       "      <td>0.934275</td>\n",
       "      <td>0.275887</td>\n",
       "      <td>10.561352</td>\n",
       "      <td>0.909611</td>\n",
       "      <td>26.31</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1352</td>\n",
       "      <td>GCF_009734005.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>SRR24</td>\n",
       "      <td>2919198.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.119192</td>\n",
       "      <td>269.461884</td>\n",
       "      <td>0.046409</td>\n",
       "      <td>0.249496</td>\n",
       "      <td>0.186010</td>\n",
       "      <td>6.184292</td>\n",
       "      <td>0.869586</td>\n",
       "      <td>17.35</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29385</td>\n",
       "      <td>GCF_007814115.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Staphylococcus saprophyticus</td>\n",
       "      <td>1A</td>\n",
       "      <td>2635789.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Staphylococcus saprophyticus</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.231856</td>\n",
       "      <td>158.609517</td>\n",
       "      <td>0.066855</td>\n",
       "      <td>0.370481</td>\n",
       "      <td>0.180454</td>\n",
       "      <td>6.922310</td>\n",
       "      <td>0.869870</td>\n",
       "      <td>16.83</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>78331</td>\n",
       "      <td>GCF_000253375.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Mycobacterium canettii CIPT 140010059</td>\n",
       "      <td>CIPT 140010059</td>\n",
       "      <td>4482059.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Mycobacterium canettii</td>\n",
       "      <td>...</td>\n",
       "      <td>0.739821</td>\n",
       "      <td>0.011292</td>\n",
       "      <td>161.169516</td>\n",
       "      <td>0.001858</td>\n",
       "      <td>0.011227</td>\n",
       "      <td>0.165485</td>\n",
       "      <td>6.076979</td>\n",
       "      <td>0.894239</td>\n",
       "      <td>15.65</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>29382</td>\n",
       "      <td>GCF_003956025.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Staphylococcus cohnii</td>\n",
       "      <td>FDAARGOS_538</td>\n",
       "      <td>2700845.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Staphylococcus cohnii</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.221629</td>\n",
       "      <td>167.008188</td>\n",
       "      <td>0.062271</td>\n",
       "      <td>0.681949</td>\n",
       "      <td>0.091313</td>\n",
       "      <td>18.395722</td>\n",
       "      <td>0.869676</td>\n",
       "      <td>8.52</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Taxonomy ID Assembly Accession ID         Source Database  \\\n",
       "0         96241       GCF_006094475.1  SOURCE_DATABASE_REFSEQ   \n",
       "1           287       GCF_000006765.1  SOURCE_DATABASE_REFSEQ   \n",
       "2         28901       GCF_000006945.2  SOURCE_DATABASE_REFSEQ   \n",
       "3          1280       GCF_000013425.1  SOURCE_DATABASE_REFSEQ   \n",
       "4          1639       GCF_000196035.1  SOURCE_DATABASE_REFSEQ   \n",
       "5          1351       GCF_001598635.1  SOURCE_DATABASE_REFSEQ   \n",
       "6          1613       GCF_029961225.1  SOURCE_DATABASE_REFSEQ   \n",
       "7           623       GCF_000006925.2  SOURCE_DATABASE_REFSEQ   \n",
       "8           562       GCF_000008865.2  SOURCE_DATABASE_REFSEQ   \n",
       "9          1642       GCF_009648575.1  SOURCE_DATABASE_REFSEQ   \n",
       "10        72361       GCF_003667885.1  SOURCE_DATABASE_REFSEQ   \n",
       "11      1654388       GCF_900636685.1  SOURCE_DATABASE_REFSEQ   \n",
       "12       483913       GCF_023238325.1  SOURCE_DATABASE_REFSEQ   \n",
       "13       260554       GCF_004006435.1  SOURCE_DATABASE_REFSEQ   \n",
       "14       227866       GCF_012225885.1  SOURCE_DATABASE_REFSEQ   \n",
       "15        72360       GCF_012648005.1  SOURCE_DATABASE_REFSEQ   \n",
       "16         1423       GCF_000009045.1  SOURCE_DATABASE_REFSEQ   \n",
       "17          573       GCF_000240185.1  SOURCE_DATABASE_REFSEQ   \n",
       "18      2892442       GCF_020911845.1  SOURCE_DATABASE_REFSEQ   \n",
       "19        54736       GCF_000439255.1  SOURCE_DATABASE_REFSEQ   \n",
       "20         1311       GCF_001552035.1  SOURCE_DATABASE_REFSEQ   \n",
       "21       208962       GCF_016904755.1  SOURCE_DATABASE_REFSEQ   \n",
       "22         1396       GCF_002220285.1  SOURCE_DATABASE_REFSEQ   \n",
       "23       164546       GCF_022406855.1  SOURCE_DATABASE_REFSEQ   \n",
       "24      2044467       GCF_005843885.1  SOURCE_DATABASE_REFSEQ   \n",
       "25      1499973       GCF_002900365.1  SOURCE_DATABASE_REFSEQ   \n",
       "26       492670       GCF_000769555.1  SOURCE_DATABASE_REFSEQ   \n",
       "27          564       GCF_020097475.1  SOURCE_DATABASE_REFSEQ   \n",
       "28         1352       GCF_009734005.1  SOURCE_DATABASE_REFSEQ   \n",
       "29        29385       GCF_007814115.1  SOURCE_DATABASE_REFSEQ   \n",
       "30        78331       GCF_000253375.1  SOURCE_DATABASE_REFSEQ   \n",
       "31        29382       GCF_003956025.1  SOURCE_DATABASE_REFSEQ   \n",
       "\n",
       "    Is Representative   Assembly Level  \\\n",
       "0                True  Complete Genome   \n",
       "1                True  Complete Genome   \n",
       "2                True  Complete Genome   \n",
       "3                True  Complete Genome   \n",
       "4                True  Complete Genome   \n",
       "5               False  Complete Genome   \n",
       "6                True  Complete Genome   \n",
       "7                True  Complete Genome   \n",
       "8                True  Complete Genome   \n",
       "9                True  Complete Genome   \n",
       "10               True  Complete Genome   \n",
       "11               True  Complete Genome   \n",
       "12               True  Complete Genome   \n",
       "13               True  Complete Genome   \n",
       "14               True  Complete Genome   \n",
       "15               True  Complete Genome   \n",
       "16               True  Complete Genome   \n",
       "17               True  Complete Genome   \n",
       "18              False  Complete Genome   \n",
       "19               True  Complete Genome   \n",
       "20               True  Complete Genome   \n",
       "21               True  Complete Genome   \n",
       "22               True  Complete Genome   \n",
       "23              False  Complete Genome   \n",
       "24              False  Complete Genome   \n",
       "25              False  Complete Genome   \n",
       "26               True  Complete Genome   \n",
       "27               True  Complete Genome   \n",
       "28               True  Complete Genome   \n",
       "29               True  Complete Genome   \n",
       "30               True  Complete Genome   \n",
       "31              False  Complete Genome   \n",
       "\n",
       "                                 Organism of Assembly  \\\n",
       "0            Bacillus spizizenii ATCC 6633 = JCM 2499   \n",
       "1                         Pseudomonas aeruginosa PAO1   \n",
       "2   Salmonella enterica subsp. enterica serovar Ty...   \n",
       "3       Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "4                        Listeria monocytogenes EGD-e   \n",
       "5                               Enterococcus faecalis   \n",
       "6                       Limosilactobacillus fermentum   \n",
       "7                       Shigella flexneri 2a str. 301   \n",
       "8                 Escherichia coli O157:H7 str. Sakai   \n",
       "9                                    Listeria innocua   \n",
       "10                              Bacillus vallismortis   \n",
       "11                         Staphylococcus schweitzeri   \n",
       "12                               Bacillus inaquosorum   \n",
       "13                              Bacillus halotolerans   \n",
       "14                               Bacillus tequilensis   \n",
       "15                                Bacillus mojavensis   \n",
       "16         Bacillus subtilis subsp. subtilis str. 168   \n",
       "17    Klebsiella pneumoniae subsp. pneumoniae HS11286   \n",
       "18                             Pseudomonas sp. CIP-10   \n",
       "19                         Salmonella bongori N268-08   \n",
       "20                           Streptococcus agalactiae   \n",
       "21                               Escherichia albertii   \n",
       "22                                    Bacillus cereus   \n",
       "23                            Cupriavidus taiwanensis   \n",
       "24                              Escherichia sp. E4742   \n",
       "25                               Escherichia marmotae   \n",
       "26                                Bacillus velezensis   \n",
       "27                             Escherichia fergusonii   \n",
       "28                               Enterococcus faecium   \n",
       "29                       Staphylococcus saprophyticus   \n",
       "30              Mycobacterium canettii CIPT 140010059   \n",
       "31                              Staphylococcus cohnii   \n",
       "\n",
       "                        Strain  Total Length  Downloaded  \\\n",
       "0                    ATCC 6633     4045538.0        True   \n",
       "1                         PAO1     6264404.0        True   \n",
       "2                          LT2     4951383.0        True   \n",
       "3                    NCTC 8325     2821361.0        True   \n",
       "4                        EGD-e     2944528.0        True   \n",
       "5                         LD33     2803429.0        True   \n",
       "6                     EFEL6800     2103331.0        True   \n",
       "7                          301     4828820.0        True   \n",
       "8   Sakai substr. RIMD 0509952     5594605.0        True   \n",
       "9                  CFSAN044836     2922148.0        True   \n",
       "10                      Bac111     3958372.0        True   \n",
       "11                   NCTC13712     2784939.0        True   \n",
       "12                       A65.1     4242130.0        True   \n",
       "13                    ZB201702     4154245.0        True   \n",
       "14                   EA-CB0015     4012371.0        True   \n",
       "15                    UCMB5075     4031121.0        True   \n",
       "16                         168     4215606.0        True   \n",
       "17                     HS11286     5682322.0        True   \n",
       "18                      CIP-10     6473352.0        True   \n",
       "19                     N268-08     4773537.0        True   \n",
       "20                     NGBS128     2079123.0        True   \n",
       "21                  Sample 167     4631903.0        True   \n",
       "22                    FORC_047     5836971.0        True   \n",
       "23                    SWF66294     6782863.0        True   \n",
       "24                       E4742     5120753.0        True   \n",
       "25                    HT073016     4896291.0        True   \n",
       "26                       JS25R     4014440.0        True   \n",
       "27               FDAARGOS_1499     4645928.0        True   \n",
       "28                       SRR24     2919198.0        True   \n",
       "29                          1A     2635789.0        True   \n",
       "30              CIPT 140010059     4482059.0        True   \n",
       "31                FDAARGOS_538     2700845.0        True   \n",
       "\n",
       "                          Species  ...  Secondary Expected  Secondary Score  \\\n",
       "0             Bacillus spizizenii  ...            1.000000         1.000000   \n",
       "1          Pseudomonas aeruginosa  ...            1.000000         0.967680   \n",
       "2             Salmonella enterica  ...            1.000000         0.935623   \n",
       "3           Staphylococcus aureus  ...            1.000000         0.921382   \n",
       "4          Listeria monocytogenes  ...            1.000000         0.931354   \n",
       "5           Enterococcus faecalis  ...            1.000000         0.890446   \n",
       "6   Limosilactobacillus fermentum  ...            1.000000         0.892914   \n",
       "7               Shigella flexneri  ...            1.000000         0.889687   \n",
       "8                Escherichia coli  ...            1.000000         0.808360   \n",
       "9                Listeria innocua  ...            1.000000         0.862307   \n",
       "10          Bacillus vallismortis  ...            1.000000         0.881393   \n",
       "11     Staphylococcus schweitzeri  ...            1.000000         0.861931   \n",
       "12           Bacillus inaquosorum  ...            1.000000         0.861323   \n",
       "13          Bacillus halotolerans  ...            1.000000         0.851494   \n",
       "14           Bacillus tequilensis  ...            1.000000         0.849856   \n",
       "15            Bacillus mojavensis  ...            1.000000         0.858009   \n",
       "16              Bacillus subtilis  ...            1.000000         0.862585   \n",
       "17          Klebsiella pneumoniae  ...            1.000000         0.426917   \n",
       "18         Pseudomonas sp. CIP-10  ...            1.000000         0.301415   \n",
       "19             Salmonella bongori  ...            1.000000         0.805211   \n",
       "20       Streptococcus agalactiae  ...            1.000000         0.055425   \n",
       "21           Escherichia albertii  ...            1.000000         0.800222   \n",
       "22                Bacillus cereus  ...            1.000000         0.053919   \n",
       "23        Cupriavidus taiwanensis  ...            0.997388         0.081080   \n",
       "24          Escherichia sp. E4742  ...            1.000000         0.770260   \n",
       "25           Escherichia marmotae  ...            1.000000         0.813909   \n",
       "26            Bacillus velezensis  ...            1.000000         0.390627   \n",
       "27         Escherichia fergusonii  ...            1.000000         0.699085   \n",
       "28           Enterococcus faecium  ...            1.000000         0.119192   \n",
       "29   Staphylococcus saprophyticus  ...            1.000000         0.231856   \n",
       "30         Mycobacterium canettii  ...            0.739821         0.011292   \n",
       "31          Staphylococcus cohnii  ...            1.000000         0.221629   \n",
       "\n",
       "    Secondary Depth  Primary Breadth  Primary Expected  Primary Score  \\\n",
       "0        132.511000         0.999999          1.000000       0.999999   \n",
       "1         65.565285         0.967013          1.000000       0.967013   \n",
       "2         99.628887         0.928550          1.000000       0.928550   \n",
       "3        160.315715         0.917263          1.000000       0.917263   \n",
       "4        129.529645         0.911952          1.000000       0.911952   \n",
       "5        132.768305         0.889125          1.000000       0.889125   \n",
       "6        212.556811         0.885748          1.000000       0.885748   \n",
       "7        114.269365         0.786340          1.000000       0.786340   \n",
       "8         97.632873         0.712230          1.000000       0.712230   \n",
       "9        111.905010         0.536577          0.999999       0.536578   \n",
       "10       104.355670         0.203398          0.449790       0.452206   \n",
       "11       144.854998         0.430713          0.994473       0.433107   \n",
       "12       105.754378         0.223387          0.543606       0.410936   \n",
       "13        90.357620         0.143939          0.359235       0.400682   \n",
       "14       105.969970         0.183283          0.458545       0.399705   \n",
       "15        90.663038         0.128208          0.328128       0.390727   \n",
       "16       102.397199         0.193634          0.510101       0.379599   \n",
       "17        71.521912         0.088700          0.236148       0.375613   \n",
       "18        52.264418         0.056747          0.156827       0.361844   \n",
       "19        84.664805         0.274417          0.790406       0.347185   \n",
       "20       772.158157         0.018327          0.055622       0.329498   \n",
       "21        92.606287         0.200037          0.608517       0.328729   \n",
       "22       641.472836         0.017304          0.054238       0.319038   \n",
       "23        73.546897         0.014609          0.047258       0.309141   \n",
       "24       108.055180         0.274972          0.927370       0.296508   \n",
       "25        95.923526         0.220992          0.757247       0.291836   \n",
       "26        91.963779         0.073708          0.255968       0.287956   \n",
       "27        89.949381         0.257754          0.934275       0.275887   \n",
       "28       269.461884         0.046409          0.249496       0.186010   \n",
       "29       158.609517         0.066855          0.370481       0.180454   \n",
       "30       161.169516         0.001858          0.011227       0.165485   \n",
       "31       167.008188         0.062271          0.681949       0.091313   \n",
       "\n",
       "    Primary Depth  Consensus ANI  Combined PS and ANI (Sqrt(ANI)xPSx100)  \\\n",
       "0       68.478268       0.999977                                  100.00   \n",
       "1       56.030159       0.994526                                   96.44   \n",
       "2       64.813931       0.989556                                   92.37   \n",
       "3       83.129377       0.993009                                   91.41   \n",
       "4       77.818246       0.947604                                   88.77   \n",
       "5      101.940818       0.991167                                   88.52   \n",
       "6      136.140257       0.993539                                   88.29   \n",
       "7       32.865802       0.980880                                   77.88   \n",
       "8       33.563861       0.977177                                   70.41   \n",
       "9       25.403433       0.908522                                   51.14   \n",
       "10       2.937331       0.931382                                   43.64   \n",
       "11      12.068378       0.916871                                   41.47   \n",
       "12       3.511335       0.937964                                   39.80   \n",
       "13       3.092184       0.902012                                   38.05   \n",
       "14       3.347207       0.930512                                   38.56   \n",
       "15       3.101836       0.902625                                   37.12   \n",
       "16       3.685022       0.930759                                   36.62   \n",
       "17       3.036957       0.882375                                   35.28   \n",
       "18       3.006012       0.877682                                   33.90   \n",
       "19       5.694112       0.913124                                   33.18   \n",
       "20       3.122534       0.869289                                   30.72   \n",
       "21       4.688133       0.904501                                   31.26   \n",
       "22       3.222602       0.870186                                   29.76   \n",
       "23       3.313685       0.875291                                   28.92   \n",
       "24       9.536773       0.915722                                   28.37   \n",
       "25       6.406079       0.912023                                   27.87   \n",
       "26       4.011359       0.883434                                   27.07   \n",
       "27      10.561352       0.909611                                   26.31   \n",
       "28       6.184292       0.869586                                   17.35   \n",
       "29       6.922310       0.869870                                   16.83   \n",
       "30       6.076979       0.894239                                   15.65   \n",
       "31      18.395722       0.869676                                    8.52   \n",
       "\n",
       "    Presence/Absence  \n",
       "0            Present  \n",
       "1            Present  \n",
       "2            Present  \n",
       "3            Present  \n",
       "4            Present  \n",
       "5            Present  \n",
       "6            Present  \n",
       "7            Present  \n",
       "8            Present  \n",
       "9             Absent  \n",
       "10            Absent  \n",
       "11            Absent  \n",
       "12            Absent  \n",
       "13            Absent  \n",
       "14            Absent  \n",
       "15            Absent  \n",
       "16            Absent  \n",
       "17            Absent  \n",
       "18            Absent  \n",
       "19            Absent  \n",
       "20            Absent  \n",
       "21            Absent  \n",
       "22            Absent  \n",
       "23            Absent  \n",
       "24            Absent  \n",
       "25            Absent  \n",
       "26            Absent  \n",
       "27            Absent  \n",
       "28            Absent  \n",
       "29            Absent  \n",
       "30            Absent  \n",
       "31            Absent  \n",
       "\n",
       "[32 rows x 21 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2347ee7f-71b9-4fe1-8aeb-78cbead638b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ground_truth_abd = {\n",
    "\"Listeria monocytogenes\": 12.0,\n",
    "\"Pseudomonas aeruginosa\": 12.0,\n",
    "\"Bacillus spizizenii\": 12.0,\n",
    "\"Escherichia coli\": 12.0,\n",
    "\"Salmonella enterica\": 12.0,\n",
    "\"Limosilactobacillus fermentum\": 12.0,\n",
    "\"Enterococcus faecalis\": 12.0,\n",
    "\"Staphylococcus aureus\": 12.0,\n",
    "\"Saccharomyces cerevisiae\": 2.0,\n",
    "\"Cryptococcus neoformans\": 2.0,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "084599c5-3974-42fc-8164-ad72f8d69c16",
   "metadata": {},
   "outputs": [],
   "source": [
    "ground_truth_set = set(ground_truth_abd.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "95d3a570-3faa-4019-bf76-418e2bcf3022",
   "metadata": {},
   "outputs": [],
   "source": [
    "positive_calls = set(result_df[result_df['Presence/Absence'] == 'Present']['Species'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "23f662c6-ffd7-459c-9e70-292dcb55b157",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ground_truth_set.intersection(positive_calls))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cc7843c3-0465-4141-a495-1288c6f38d69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Cryptococcus neoformans', 'Saccharomyces cerevisiae'}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ground_truth_set - positive_calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7db00924-de19-44f9-8cf8-9a9bd7fc1991",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Assembly Accession ID</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>Is Representative</th>\n",
       "      <th>Assembly Level</th>\n",
       "      <th>Organism of Assembly</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Total Length</th>\n",
       "      <th>Downloaded</th>\n",
       "      <th>Species</th>\n",
       "      <th>...</th>\n",
       "      <th>Secondary Expected</th>\n",
       "      <th>Secondary Score</th>\n",
       "      <th>Secondary Depth</th>\n",
       "      <th>Primary Breadth</th>\n",
       "      <th>Primary Expected</th>\n",
       "      <th>Primary Score</th>\n",
       "      <th>Primary Depth</th>\n",
       "      <th>Consensus ANI</th>\n",
       "      <th>Combined PS and ANI (Sqrt(ANI)xPSx100)</th>\n",
       "      <th>Presence/Absence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>623</td>\n",
       "      <td>GCF_000006925.2</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Shigella flexneri 2a str. 301</td>\n",
       "      <td>301</td>\n",
       "      <td>4828820.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Shigella flexneri</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.889687</td>\n",
       "      <td>114.269365</td>\n",
       "      <td>0.78634</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.78634</td>\n",
       "      <td>32.865802</td>\n",
       "      <td>0.98088</td>\n",
       "      <td>77.88</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Taxonomy ID Assembly Accession ID         Source Database  \\\n",
       "7          623       GCF_000006925.2  SOURCE_DATABASE_REFSEQ   \n",
       "\n",
       "   Is Representative   Assembly Level           Organism of Assembly Strain  \\\n",
       "7               True  Complete Genome  Shigella flexneri 2a str. 301    301   \n",
       "\n",
       "   Total Length  Downloaded            Species  ...  Secondary Expected  \\\n",
       "7     4828820.0        True  Shigella flexneri  ...                 1.0   \n",
       "\n",
       "   Secondary Score  Secondary Depth  Primary Breadth  Primary Expected  \\\n",
       "7         0.889687       114.269365          0.78634               1.0   \n",
       "\n",
       "   Primary Score  Primary Depth  Consensus ANI  \\\n",
       "7        0.78634      32.865802        0.98088   \n",
       "\n",
       "   Combined PS and ANI (Sqrt(ANI)xPSx100)  Presence/Absence  \n",
       "7                                   77.88           Present  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df[result_df['Species'].isin(list(positive_calls - ground_truth_set))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b6e1e63b-657b-4080-a147-f64a110b1104",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Assembly Accession ID</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>Is Representative</th>\n",
       "      <th>Assembly Level</th>\n",
       "      <th>Organism of Assembly</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Total Length</th>\n",
       "      <th>Downloaded</th>\n",
       "      <th>Species</th>\n",
       "      <th>...</th>\n",
       "      <th>Secondary Expected</th>\n",
       "      <th>Secondary Score</th>\n",
       "      <th>Secondary Depth</th>\n",
       "      <th>Primary Breadth</th>\n",
       "      <th>Primary Expected</th>\n",
       "      <th>Primary Score</th>\n",
       "      <th>Primary Depth</th>\n",
       "      <th>Consensus ANI</th>\n",
       "      <th>Combined PS and ANI (Sqrt(ANI)xPSx100)</th>\n",
       "      <th>Presence/Absence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Taxonomy ID, Assembly Accession ID, Source Database, Is Representative, Assembly Level, Organism of Assembly, Strain, Total Length, Downloaded, Species, Secondary Breadth, Secondary Expected, Secondary Score, Secondary Depth, Primary Breadth, Primary Expected, Primary Score, Primary Depth, Consensus ANI, Combined PS and ANI (Sqrt(ANI)xPSx100), Presence/Absence]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 21 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df[result_df['Species'].isin(list(ground_truth_set - positive_calls))]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11f6b5e1-a48d-424d-bc43-92b5bca25376",
   "metadata": {},
   "source": [
    "# Argument Parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ec9c12ec-9dbf-4045-80a6-7d3107371e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = argparse.ArgumentParser(description=\"Universal Taxonomic Classification Verifier.\")\n",
    "\n",
    "parser.add_argument(\"-c\", \"--classification\", type=pathlib.Path, required=True, help=\"Path to the Taxonomic Classification Report. Accepting csv/tsv file format, other text formats are treated as tsv.\")\n",
    "parser.add_argument(\"-i\", \"--fastq\", type=pathlib.Path, required=True, help=\"Path to the first fastq file.\")\n",
    "parser.add_argument(\"-I\", \"--fastq2\", type=pathlib.Path, required=False, help=\"Path to the second fastq file for paired-end reads.\")\n",
    "parser.add_argument(\"-m\", \"--mode\", type=str, required=False, choices=['ont', 'illumina'], help=\"Modes for different sequencing platforms [ont, illumina]. Default:[ont]\",  default='ont')\n",
    "parser.add_argument(\"-o\", \"--output\", type=pathlib.Path, required=True, help=\"Path to the output directory.\")\n",
    "parser.add_argument(\"-t\", \"--taxid-idx\", type=int, required=False, help=\"The column index (0-based) of the taxids. Default:[0]\", default=0)\n",
    "parser.add_argument(\"-a\", \"--abundance-idx\", type=int, required=False, help=\"The column index (0-based) of the abundance. Default:[None]\")\n",
    "parser.add_argument(\"--min-abundance\", type=float, required=False, help=\"Minimum abundance (0-1) for pre-filtering, exclude taxa below the threshold.\", default=0)\n",
    "parser.add_argument(\"--min-mapq\", type=int, required=False, help=\"Minimum MAPQ for primary alignments. Default:[20]\", default=20)\n",
    "parser.add_argument(\"--min-covscore\", type=float, required=False, help=\"Minimum Coverage Score for supplementary alignments. Default:[0.7]\", default=0.7)\n",
    "parser.add_argument(\"--threads\", type=int, required=False, help=\"Number of threads for Multi-threading. Default:[1]\", default=1)\n",
    "parser.add_argument(\"--include-mag\", action='store_true', required=False, help=\"Include metagenomic assemble genomes. Default:[False]\")\n",
    "parser.set_defaults(include_mag=False)\n",
    "parser.add_argument(\"--subspecies\", action='store_true', required=False, help=\"Verify taxonomic classification at subspecies rank. Default:[False]\")\n",
    "parser.set_defaults(subspecies=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3d072a11-1f7d-45c7-80e8-df8aae1818de",
   "metadata": {},
   "outputs": [],
   "source": [
    "args = parser.parse_args(['-c', '/home/Users/ns58/emu40/Sim-Output/Large-uniform-8-0/Mob-combined-async/Mob-combined-async_rel-abundance.tsv',\n",
    "                          '-i', '/home/Users/ns58/emu40/Sim-Data/Samples/Large_30species_rs3_startx10._ONT_8_0.fastq',\n",
    "                          '-o', '../test',\n",
    "                          '-t', str(0),\n",
    "                          '--threads', str(40),\n",
    "                          '--subspecies'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f3e41278-9f73-4642-940f-f3d6ac1be60c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-c /home/Users/ns58/emu40/Sim-Output/Large-uniform-8-0/Mob-combined-async/Mob-combined-async_rel-abundance.tsv -i /home/Users/ns58/emu40/Sim-Data/Samples/Large_30species_rs3_startx10._ONT_8_0.fastq -o ../test -t 0 --threads 40 --subspecies\n"
     ]
    }
   ],
   "source": [
    "print(' '.join(['-c', '/home/Users/ns58/emu40/Sim-Output/Large-uniform-8-0/Mob-combined-async/Mob-combined-async_rel-abundance.tsv',\n",
    "                          '-i', '/home/Users/ns58/emu40/Sim-Data/Samples/Large_30species_rs3_startx10._ONT_8_0.fastq',\n",
    "                          '-o', '../test',\n",
    "                          '-t', str(0),\n",
    "                          '--threads', str(40),\n",
    "                          '--subspecies']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7f48f6c6-ac9e-4912-b335-8072777ea25c",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tsv = args.classification\n",
    "input_fastq = args.fastq\n",
    "input_fastq2 = args.fastq2\n",
    "mode = args.mode\n",
    "working_directory = args.output\n",
    "\n",
    "taxid_col_idx = args.taxid_idx\n",
    "abundance_col_idx = args.abundance_idx\n",
    "min_abundance = args.min_abundance\n",
    "min_mapq = args.min_mapq\n",
    "min_coverage_score = args.min_covscore\n",
    "threads = args.threads\n",
    "\n",
    "if args.include_mag:\n",
    "    mag_flag = 'all'\n",
    "else:\n",
    "    mag_flag = 'exclude'\n",
    "    \n",
    "call_subspecies = args.subspecies\n",
    "\n",
    "sep = '\\t'\n",
    "if str(input_file)[-3:] == 'csv':\n",
    "    sep = ','"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa6cfd78-3468-4cbd-9c93-078ba8ce07d5",
   "metadata": {},
   "source": [
    "# Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b5a8b208-4aa2-481c-8e3c-927c6634ff8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_taxa_db = NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c088dedd-fe11-416f-8985-2f359970c932",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(working_directory):\n",
    "    os.mkdir(working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "eecc0004-9f46-4b9f-8d1a-acc1dc971184",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unassigned is not a valid taxid.\n"
     ]
    }
   ],
   "source": [
    "input_df, min_abundance = parsing_input_f(input_file, sep, input_tsv, taxid_col_idx, abundance_col_idx, min_abundance)\n",
    "# make valid_kingdom a variable?\n",
    "valid_taxids = filter_input_df(input_df, min_abundance, ncbi_taxa_db, valid_kingdom={2, 4751, 2157, 10239}, ret_subspecies=call_subspecies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f5bf1e11-2199-42ac-9550-d3af818b2d66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1246       \t GCF_007954605.1 \t Leuconostoc lactis             \t CBA3625    \t Complete Genome\n",
      "173262     \t Genome Not Found.\n",
      "927083     \t GCF_000737325.1 \t Sandaracinus amylolyticus      \t DSM 5366.. \t Complete Genome\n",
      "63363      \t GCF_000008625.1 \t Aquifex aeolicus VF5           \t VF5        \t Complete Genome\n",
      "658062     \t GCF_000014005.1 \t Candidatus Koribacter versat.. \t Ellin345   \t Complete Genome\n",
      "1293412    \t GCF_006542355.1 \t Swingsia samuiensis            \t AH83       \t Complete Genome\n",
      "2203724    \t GCF_004102045.2 \t Oenococcus sicerae             \t UCMA1522.. \t Complete Genome\n",
      "664643     \t GCF_024347035.1 \t Vibrio plantisponsor           \t CECT 758.. \t Complete Genome\n",
      "623        \t GCF_000006925.2 \t Shigella flexneri 2a str. 30.. \t 301        \t Complete Genome\n",
      "154981     \t GCF_025137635.1 \t Aliiroseovarius crassostreae   \t S044       \t Complete Genome\n",
      "2305508    \t GCF_003576455.2 \t Mucilaginibacter celer         \t HYN0043    \t Complete Genome\n",
      "436        \t GCF_014843995.1 \t Novacetimonas hansenii         \t C110       \t Complete Genome\n",
      "136241     \t GCF_000238995.1 \t Mycoplasma haemocanis str. I.. \t Illinois   \t Complete Genome\n",
      "380        \t GCF_003177055.1 \t Sinorhizobium fredii CCBAU 2.. \t CCBAU 25.. \t Complete Genome\n",
      "445709     \t GCF_001017775.3 \t Pandoraea thiooxydans          \t DSM 2532.. \t Complete Genome\n",
      "413403     \t GCF_024347075.1 \t Vibrio rarus                   \t LMG 2367.. \t Complete Genome\n",
      "1851148    \t GCF_001999965.1 \t Limihaloglobus sulfuriphilus   \t SM-Chi-D.. \t Complete Genome\n",
      "28028      \t GCF_001941465.1 \t Corynebacterium flavescens     \t OJ8        \t Complete Genome\n",
      "1197717    \t GCF_001701045.1 \t Cloacibacillus porcorum        \t CL-84 (T.. \t Complete Genome\n",
      "507626     \t GCF_001545155.1 \t Halomonas chromatireducens     \t AGD 8-3    \t Complete Genome\n",
      "1389020    \t GCF_021442325.1 \t Agromyces marinus              \t DSM 2615.. \t Complete Genome\n",
      "190148     \t GCF_001693515.2 \t Bradyrhizobium paxllaeri       \t LMTR 21    \t Complete Genome\n",
      "1761016    \t GCF_003368325.1 \t Paraburkholderia caffeinilyt.. \t CF1        \t Complete Genome\n",
      "392416     \t GCF_001951175.1 \t Companilactobacillus crustor.. \t MN047      \t Complete Genome\n",
      "2712892    \t GCF_013363675.1 \t Streptomyces sp. CAI-24        \t CAI-24     \t Contig\n",
      "67258      \t GCF_002804165.1 \t Streptomyces cavourensis       \t 1AS2a      \t Complete Genome\n",
      "1169747    \t GCF_013363475.1 \t Streptomyces sp. KAI-26        \t KAI-26     \t Contig\n",
      "2838874    \t GCF_018619515.1 \t Streptomyces sp. CYG21         \t CYG21      \t Contig\n",
      "2838873    \t GCF_018619535.1 \t Streptomyces sp. CYG20         \t CYG20      \t Contig\n",
      "945844     \t GCF_003143515.1 \t Massilia oculi                 \t CCUG 434.. \t Complete Genome\n",
      "1815583    \t GCF_014297595.1 \t Neisseria musculi              \t NW831      \t Complete Genome\n",
      "2558283    \t GCF_004569405.1 \t Neisseria sp. WF04             \t WF04       \t Contig\n",
      "577        \t GCF_012029655.1 \t Raoultella terrigena           \t JH01       \t Complete Genome\n",
      "336203     \t GCF_015074805.1 \t Sphingobium fuliginis          \t KK22       \t Complete Genome\n",
      "2878542    \t GCF_024611135.1 \t Sphingobium sp. CFD-2          \t CFD-2      \t Complete Genome\n",
      "2940617    \t GCF_024810075.1 \t Raoultella sp. BIGb0132        \t BIGb0132   \t Contig\n",
      "46354      \t GCF_000147695.2 \t Thermoanaerobacter wiegelii .. \t Rt8.B1     \t Complete Genome\n",
      "1516       \t GCF_900102395.1 \t Thermoanaerobacter thermohyd.. \t DSM 569    \t Scaffold\n",
      "1686286    \t GCF_006547305.1 \t Corynebacterium phoceense      \t 272        \t Contig\n",
      "2358       \t GCF_000266945.1 \t Desulfomonile tiedjei DSM 67.. \t DSM 6799   \t Complete Genome\n",
      "562        \t GCF_000008865.2 \t Escherichia coli O157:H7 str.. \t Sakai su.. \t Complete Genome\n",
      "2093857    \t GCF_005121165.3 \t Dysosmobacter welbionis        \t J115       \t Complete Genome\n",
      "28901      \t GCF_000006945.2 \t Salmonella enterica subsp. e.. \t LT2        \t Complete Genome\n",
      "2733484    \t GCF_017868775.1 \t Halomonas sulfidoxydans        \t MCCC 1A1.. \t Complete Genome\n",
      "615        \t GCF_000783915.2 \t Serratia marcescens            \t FDAARGOS.. \t Complete Genome\n",
      "468058     \t GCF_003148845.1 \t Mucilaginibacter oryzae        \t DSM 1997.. \t Scaffold\n",
      "47229      \t GCF_001866515.1 \t Massilia timonae               \t NEU        \t Contig\n",
      "2490855    \t GCF_030008055.1 \t Lachnoanaerobaculum gingival.. \t SHL-2023.. \t Complete Genome\n",
      "121428     \t GCF_002288285.1 \t Sphingobium xenophagum         \t C1         \t Complete Genome\n",
      "399        \t GCF_000731315.1 \t Neorhizobium galegae bv. ori.. \t HAMBI 54.. \t Complete Genome\n",
      "2920428    \t GCF_024195285.1 \t Aurantimonas marianensis       \t LRZ36      \t Contig\n",
      "384        \t GCF_001679785.1 \t Rhizobium leguminosarum        \t Vaf10      \t Complete Genome\n",
      "1076926    \t GCF_021052325.1 \t Rhizobium laguerreae           \t WSM1455    \t Complete Genome\n",
      "396        \t GCF_001664385.1 \t Rhizobium phaseoli             \t R650       \t Complete Genome\n",
      "2726076    \t GCF_023701945.1 \t Rhizobium sp. CG5              \t CG5        \t Scaffold\n",
      "3041166    \t GCF_950101795.1 \t Nocardioides sp. T2.26MG-1     \t            \t Scaffold\n",
      "287        \t GCF_000006765.1 \t Pseudomonas aeruginosa PAO1    \t PAO1       \t Complete Genome\n",
      "317        \t GCF_018394375.1 \t Pseudomonas syringae           \t Susan213.. \t Complete Genome\n",
      "285080     \t GCF_012396155.1 \t Cellulomonas septica           \t ATCC BAA.. \t Contig\n",
      "3055039    \t GCF_030380805.1 \t Rhizobium sp. S163             \t S163       \t Scaffold\n",
      "3021407    \t GCF_030182485.1 \t Rhizobium sp. CNPSo 3490       \t CNPSo 34.. \t Contig\n",
      "2919500    \t GCF_022760965.1 \t Acidovorax sp. W1-6            \t W1-6       \t Contig\n",
      "2847975    \t GCF_025128055.1 \t Synechococcus sp. CS-1329      \t CS-1329    \t Contig\n",
      "114707     \t Genome Not Found.\n",
      "134537     \t GCF_000961515.1 \t Paraburkholderia fungorum      \t ATCC BAA.. \t Complete Genome\n",
      "244566     \t GCF_001708485.1 \t Pseudomonas lurida             \t L228       \t Complete Genome\n"
     ]
    }
   ],
   "source": [
    "# can be parallelized\n",
    "reference_metadata = prepare_reference_genomes(valid_taxids, working_directory, ncbi_taxa_db, mag_flag=mag_flag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3b6276e5-36bd-4cfd-a37a-907f3c66c8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = reference_metadata[reference_metadata['Downloaded']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1f84719b-1713-43c5-81c9-581f261593e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq2assembly_dict = get_seq2assembly_dict(working_directory, downloaded_assemblies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2ef6787d-0419-4855-a344-c48b85224f8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_fasta = merge_reference_fasta(list(downloaded_assemblies['Assembly Accession ID']), working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "49967f36-999b-4180-a013-673aa62fedc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "aligner_output = run_minimap2(input_fastq, reference_fasta, 'merged', working_directory, threads=threads)\n",
    "sort_samfile('merged', aligner_output, working_directory, min_mapq=0, threads=threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "02130814-ad1c-4e55-85f0-910ffca64952",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 26.2 ms, sys: 254 ms, total: 280 ms\n",
      "Wall time: 2min 18s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "pool = Pool(processes=threads)\n",
    "pool.starmap(samtools_calculate_coverage, zip(repeat(working_directory), [True, False]))\n",
    "pool.close()\n",
    "pool.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4c7d2852-61dd-4595-b97a-2c8a14dcb558",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = alignment_summary(downloaded_assemblies,\n",
    "                                          working_directory,\n",
    "                                          seq2assembly_dict,\n",
    "                                          include_supp=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c59f498c-edd1-4146-9d0c-94afaf57918b",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = alignment_summary(downloaded_assemblies,\n",
    "                                          working_directory,\n",
    "                                          seq2assembly_dict,\n",
    "                                          include_supp=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9736777b-5376-4ba9-b750-c8f628ad2ef3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.03 s, sys: 392 ms, total: 1.43 s\n",
      "Wall time: 3min 31s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "consensus_record_dict = samtools_merged_consensus(working_directory, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "68cbb5fa-2ce8-40ab-b798-651c954134f8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5.29 s, sys: 14.9 s, total: 20.2 s\n",
      "Wall time: 37.5 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "downloaded_assemblies = ani_summary(downloaded_assemblies, consensus_record_dict, working_directory, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "052662c7-a05d-49f7-a88e-e54eb58910e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = call_present_absent(downloaded_assemblies, min_coverage_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "fb5825f7-1eaa-45b2-a984-fc47188e5fdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies.sort_values(['Primary Score'], ascending=False).to_csv(os.path.join(working_directory, \n",
    "                                                                                          'magnet_results.csv'), \n",
    "                                                                             index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84f204bd-20d8-4fe9-b42b-0d3675d24dfe",
   "metadata": {},
   "source": [
    "# Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "45a6004d-8bb3-4709-a46f-ca1bb2f5968c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Assembly Accession ID</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>Is Representative</th>\n",
       "      <th>Assembly Level</th>\n",
       "      <th>Organism of Assembly</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Total Length</th>\n",
       "      <th>Downloaded</th>\n",
       "      <th>Species</th>\n",
       "      <th>...</th>\n",
       "      <th>Secondary Breadth</th>\n",
       "      <th>Secondary Expected</th>\n",
       "      <th>Secondary Score</th>\n",
       "      <th>Secondary Depth</th>\n",
       "      <th>Primary Breadth</th>\n",
       "      <th>Primary Expected</th>\n",
       "      <th>Primary Score</th>\n",
       "      <th>Primary Depth</th>\n",
       "      <th>Combined PS and ANI (Sqrt(ANI)xPSx100)</th>\n",
       "      <th>Presence/Absence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1246</td>\n",
       "      <td>GCF_007954605.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Leuconostoc lactis</td>\n",
       "      <td>CBA3625</td>\n",
       "      <td>1791608.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Leuconostoc lactis</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>99.860177</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>98.808125</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>927083</td>\n",
       "      <td>GCF_000737325.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Sandaracinus amylolyticus</td>\n",
       "      <td>DSM 53668</td>\n",
       "      <td>10327335.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Sandaracinus amylolyticus</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>49.606000</td>\n",
       "      <td>0.999992</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999992</td>\n",
       "      <td>49.413092</td>\n",
       "      <td>100.00</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>63363</td>\n",
       "      <td>GCF_000008625.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Aquifex aeolicus VF5</td>\n",
       "      <td>VF5</td>\n",
       "      <td>1590791.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Aquifex aeolicus</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999932</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999932</td>\n",
       "      <td>49.586588</td>\n",
       "      <td>0.999932</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999932</td>\n",
       "      <td>49.447579</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>658062</td>\n",
       "      <td>GCF_000014005.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Candidatus Koribacter versatilis Ellin345</td>\n",
       "      <td>Ellin345</td>\n",
       "      <td>5650368.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Candidatus Koribacter versatilis</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999977</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999977</td>\n",
       "      <td>98.981260</td>\n",
       "      <td>0.999977</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999977</td>\n",
       "      <td>98.947459</td>\n",
       "      <td>100.00</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1293412</td>\n",
       "      <td>GCF_006542355.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Swingsia samuiensis</td>\n",
       "      <td>AH83</td>\n",
       "      <td>2193407.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Swingsia samuiensis</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999929</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999929</td>\n",
       "      <td>100.964187</td>\n",
       "      <td>0.999929</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999929</td>\n",
       "      <td>99.078553</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2203724</td>\n",
       "      <td>GCF_004102045.2</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Oenococcus sicerae</td>\n",
       "      <td>UCMA15228</td>\n",
       "      <td>1664394.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Oenococcus sicerae</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999871</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999871</td>\n",
       "      <td>49.581504</td>\n",
       "      <td>0.999871</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999871</td>\n",
       "      <td>49.436986</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>664643</td>\n",
       "      <td>GCF_024347035.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Vibrio plantisponsor</td>\n",
       "      <td>CECT 7581</td>\n",
       "      <td>4509247.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Vibrio plantisponsor</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999960</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999960</td>\n",
       "      <td>50.684451</td>\n",
       "      <td>0.999960</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999960</td>\n",
       "      <td>49.405227</td>\n",
       "      <td>100.00</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>623</td>\n",
       "      <td>GCF_000006925.2</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Shigella flexneri 2a str. 301</td>\n",
       "      <td>301</td>\n",
       "      <td>4828820.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Shigella flexneri</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>22.716979</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>19.765296</td>\n",
       "      <td>99.95</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>154981</td>\n",
       "      <td>GCF_025137635.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Aliiroseovarius crassostreae</td>\n",
       "      <td>S044</td>\n",
       "      <td>3367659.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Aliiroseovarius crassostreae</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999794</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999794</td>\n",
       "      <td>19.906297</td>\n",
       "      <td>0.999794</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999794</td>\n",
       "      <td>19.789472</td>\n",
       "      <td>99.98</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2305508</td>\n",
       "      <td>GCF_003576455.2</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Mucilaginibacter celer</td>\n",
       "      <td>HYN0043</td>\n",
       "      <td>7149106.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Mucilaginibacter celer</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999760</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999760</td>\n",
       "      <td>19.861267</td>\n",
       "      <td>0.999760</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999760</td>\n",
       "      <td>19.789550</td>\n",
       "      <td>99.98</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>436</td>\n",
       "      <td>GCF_014843995.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Novacetimonas hansenii</td>\n",
       "      <td>C110</td>\n",
       "      <td>3549186.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Novacetimonas hansenii</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999419</td>\n",
       "      <td>0.999955</td>\n",
       "      <td>0.999464</td>\n",
       "      <td>10.000399</td>\n",
       "      <td>0.999419</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>0.999470</td>\n",
       "      <td>9.888364</td>\n",
       "      <td>99.95</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>136241</td>\n",
       "      <td>GCF_000238995.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Mycoplasma haemocanis str. Illinois</td>\n",
       "      <td>Illinois</td>\n",
       "      <td>919992.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Mycoplasma haemocanis</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999188</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999188</td>\n",
       "      <td>19.812887</td>\n",
       "      <td>0.999188</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999188</td>\n",
       "      <td>19.806581</td>\n",
       "      <td>99.92</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>380</td>\n",
       "      <td>GCF_003177055.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Sinorhizobium fredii CCBAU 25509</td>\n",
       "      <td>CCBAU 25509</td>\n",
       "      <td>6810864.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Sinorhizobium fredii</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999976</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999976</td>\n",
       "      <td>103.333486</td>\n",
       "      <td>0.999976</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999976</td>\n",
       "      <td>100.712422</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>445709</td>\n",
       "      <td>GCF_001017775.3</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pandoraea thiooxydans</td>\n",
       "      <td>DSM 25325</td>\n",
       "      <td>4464186.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Pandoraea thiooxydans</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999933</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999933</td>\n",
       "      <td>19.914034</td>\n",
       "      <td>0.999933</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999933</td>\n",
       "      <td>19.786825</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>413403</td>\n",
       "      <td>GCF_024347075.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Vibrio rarus</td>\n",
       "      <td>LMG 23674</td>\n",
       "      <td>3877193.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Vibrio rarus</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999865</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999865</td>\n",
       "      <td>20.375531</td>\n",
       "      <td>0.999865</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999865</td>\n",
       "      <td>19.762609</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1851148</td>\n",
       "      <td>GCF_001999965.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Limihaloglobus sulfuriphilus</td>\n",
       "      <td>SM-Chi-D1</td>\n",
       "      <td>3868357.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Limihaloglobus sulfuriphilus</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>100.264028</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>98.883759</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>28028</td>\n",
       "      <td>GCF_001941465.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Corynebacterium flavescens</td>\n",
       "      <td>OJ8</td>\n",
       "      <td>2758653.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Corynebacterium flavescens</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999901</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999901</td>\n",
       "      <td>50.766405</td>\n",
       "      <td>0.999901</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999901</td>\n",
       "      <td>48.833515</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1197717</td>\n",
       "      <td>GCF_001701045.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Cloacibacillus porcorum</td>\n",
       "      <td>CL-84 (T)</td>\n",
       "      <td>3585187.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Cloacibacillus porcorum</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999960</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999960</td>\n",
       "      <td>99.919357</td>\n",
       "      <td>0.999943</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999943</td>\n",
       "      <td>98.855180</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1389020</td>\n",
       "      <td>GCF_021442325.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Agromyces marinus</td>\n",
       "      <td>DSM 26151</td>\n",
       "      <td>3202016.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Agromyces marinus</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999974</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999974</td>\n",
       "      <td>99.276504</td>\n",
       "      <td>0.999974</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999974</td>\n",
       "      <td>98.883394</td>\n",
       "      <td>100.00</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>190148</td>\n",
       "      <td>GCF_001693515.2</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bradyrhizobium paxllaeri</td>\n",
       "      <td>LMTR 21</td>\n",
       "      <td>8377719.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Bradyrhizobium paxllaeri</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999984</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999984</td>\n",
       "      <td>50.402800</td>\n",
       "      <td>0.999984</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999984</td>\n",
       "      <td>49.757890</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1761016</td>\n",
       "      <td>GCF_003368325.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Paraburkholderia caffeinilytica</td>\n",
       "      <td>CF1</td>\n",
       "      <td>8324638.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Paraburkholderia caffeinilytica</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999977</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999977</td>\n",
       "      <td>20.064410</td>\n",
       "      <td>0.999488</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999488</td>\n",
       "      <td>19.822158</td>\n",
       "      <td>99.95</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>392416</td>\n",
       "      <td>GCF_001951175.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Companilactobacillus crustorum</td>\n",
       "      <td>MN047</td>\n",
       "      <td>2316479.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Companilactobacillus crustorum</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999683</td>\n",
       "      <td>0.999965</td>\n",
       "      <td>0.999719</td>\n",
       "      <td>10.241042</td>\n",
       "      <td>0.999683</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>0.999733</td>\n",
       "      <td>9.902845</td>\n",
       "      <td>99.97</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>67258</td>\n",
       "      <td>GCF_002804165.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Streptomyces cavourensis</td>\n",
       "      <td>1AS2a</td>\n",
       "      <td>7600475.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Streptomyces cavourensis</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>49.725614</td>\n",
       "      <td>0.999983</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999983</td>\n",
       "      <td>47.023317</td>\n",
       "      <td>100.00</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>945844</td>\n",
       "      <td>GCF_003143515.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Massilia oculi</td>\n",
       "      <td>CCUG 43427</td>\n",
       "      <td>5844653.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Massilia oculi</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>99.814100</td>\n",
       "      <td>0.999962</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999962</td>\n",
       "      <td>98.874955</td>\n",
       "      <td>100.00</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1815583</td>\n",
       "      <td>GCF_014297595.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Neisseria musculi</td>\n",
       "      <td>NW831</td>\n",
       "      <td>2933071.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Neisseria musculi</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999963</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999963</td>\n",
       "      <td>50.761172</td>\n",
       "      <td>0.999963</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999963</td>\n",
       "      <td>48.780799</td>\n",
       "      <td>100.00</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>577</td>\n",
       "      <td>GCF_012029655.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Raoultella terrigena</td>\n",
       "      <td>JH01</td>\n",
       "      <td>5598454.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Raoultella terrigena</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>99.720803</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>98.528117</td>\n",
       "      <td>100.00</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>336203</td>\n",
       "      <td>GCF_015074805.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Sphingobium fuliginis</td>\n",
       "      <td>KK22</td>\n",
       "      <td>4984172.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Sphingobium fuliginis</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999919</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999919</td>\n",
       "      <td>50.386814</td>\n",
       "      <td>0.824513</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.824513</td>\n",
       "      <td>33.643599</td>\n",
       "      <td>82.38</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>46354</td>\n",
       "      <td>GCF_000147695.2</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Thermoanaerobacter wiegelii Rt8.B1</td>\n",
       "      <td>Rt8.B1</td>\n",
       "      <td>2785056.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Thermoanaerobacter wiegelii</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>101.209415</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>98.311759</td>\n",
       "      <td>99.99</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>2358</td>\n",
       "      <td>GCF_000266945.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Desulfomonile tiedjei DSM 6799</td>\n",
       "      <td>DSM 6799</td>\n",
       "      <td>6527027.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Desulfomonile tiedjei</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999981</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999981</td>\n",
       "      <td>20.117984</td>\n",
       "      <td>0.999980</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999980</td>\n",
       "      <td>19.713291</td>\n",
       "      <td>100.00</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>2733484</td>\n",
       "      <td>GCF_017868775.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Halomonas sulfidoxydans</td>\n",
       "      <td>MCCC 1A11059</td>\n",
       "      <td>4491948.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Halomonas sulfidoxydans</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999982</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999982</td>\n",
       "      <td>49.732874</td>\n",
       "      <td>0.999982</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999982</td>\n",
       "      <td>49.468670</td>\n",
       "      <td>100.00</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Taxonomy ID Assembly Accession ID         Source Database  \\\n",
       "0          1246       GCF_007954605.1  SOURCE_DATABASE_REFSEQ   \n",
       "2        927083       GCF_000737325.1  SOURCE_DATABASE_REFSEQ   \n",
       "3         63363       GCF_000008625.1  SOURCE_DATABASE_REFSEQ   \n",
       "4        658062       GCF_000014005.1  SOURCE_DATABASE_REFSEQ   \n",
       "5       1293412       GCF_006542355.1  SOURCE_DATABASE_REFSEQ   \n",
       "6       2203724       GCF_004102045.2  SOURCE_DATABASE_REFSEQ   \n",
       "7        664643       GCF_024347035.1  SOURCE_DATABASE_REFSEQ   \n",
       "8           623       GCF_000006925.2  SOURCE_DATABASE_REFSEQ   \n",
       "9        154981       GCF_025137635.1  SOURCE_DATABASE_REFSEQ   \n",
       "10      2305508       GCF_003576455.2  SOURCE_DATABASE_REFSEQ   \n",
       "11          436       GCF_014843995.1  SOURCE_DATABASE_REFSEQ   \n",
       "12       136241       GCF_000238995.1  SOURCE_DATABASE_REFSEQ   \n",
       "13          380       GCF_003177055.1  SOURCE_DATABASE_REFSEQ   \n",
       "14       445709       GCF_001017775.3  SOURCE_DATABASE_REFSEQ   \n",
       "15       413403       GCF_024347075.1  SOURCE_DATABASE_REFSEQ   \n",
       "16      1851148       GCF_001999965.1  SOURCE_DATABASE_REFSEQ   \n",
       "17        28028       GCF_001941465.1  SOURCE_DATABASE_REFSEQ   \n",
       "18      1197717       GCF_001701045.1  SOURCE_DATABASE_REFSEQ   \n",
       "20      1389020       GCF_021442325.1  SOURCE_DATABASE_REFSEQ   \n",
       "21       190148       GCF_001693515.2  SOURCE_DATABASE_REFSEQ   \n",
       "22      1761016       GCF_003368325.1  SOURCE_DATABASE_REFSEQ   \n",
       "23       392416       GCF_001951175.1  SOURCE_DATABASE_REFSEQ   \n",
       "25        67258       GCF_002804165.1  SOURCE_DATABASE_REFSEQ   \n",
       "29       945844       GCF_003143515.1  SOURCE_DATABASE_REFSEQ   \n",
       "30      1815583       GCF_014297595.1  SOURCE_DATABASE_REFSEQ   \n",
       "32          577       GCF_012029655.1  SOURCE_DATABASE_REFSEQ   \n",
       "33       336203       GCF_015074805.1  SOURCE_DATABASE_REFSEQ   \n",
       "36        46354       GCF_000147695.2  SOURCE_DATABASE_REFSEQ   \n",
       "39         2358       GCF_000266945.1  SOURCE_DATABASE_REFSEQ   \n",
       "43      2733484       GCF_017868775.1  SOURCE_DATABASE_REFSEQ   \n",
       "\n",
       "   Is Representative   Assembly Level  \\\n",
       "0               True  Complete Genome   \n",
       "2               True  Complete Genome   \n",
       "3               True  Complete Genome   \n",
       "4               True  Complete Genome   \n",
       "5               True  Complete Genome   \n",
       "6               True  Complete Genome   \n",
       "7               True  Complete Genome   \n",
       "8               True  Complete Genome   \n",
       "9               True  Complete Genome   \n",
       "10              True  Complete Genome   \n",
       "11              True  Complete Genome   \n",
       "12              True  Complete Genome   \n",
       "13              True  Complete Genome   \n",
       "14              True  Complete Genome   \n",
       "15              True  Complete Genome   \n",
       "16              True  Complete Genome   \n",
       "17              True  Complete Genome   \n",
       "18              True  Complete Genome   \n",
       "20              True  Complete Genome   \n",
       "21              True  Complete Genome   \n",
       "22              True  Complete Genome   \n",
       "23              True  Complete Genome   \n",
       "25              True  Complete Genome   \n",
       "29              True  Complete Genome   \n",
       "30              True  Complete Genome   \n",
       "32              True  Complete Genome   \n",
       "33              True  Complete Genome   \n",
       "36              True  Complete Genome   \n",
       "39              True  Complete Genome   \n",
       "43              True  Complete Genome   \n",
       "\n",
       "                         Organism of Assembly        Strain  Total Length  \\\n",
       "0                          Leuconostoc lactis       CBA3625     1791608.0   \n",
       "2                   Sandaracinus amylolyticus     DSM 53668    10327335.0   \n",
       "3                        Aquifex aeolicus VF5           VF5     1590791.0   \n",
       "4   Candidatus Koribacter versatilis Ellin345      Ellin345     5650368.0   \n",
       "5                         Swingsia samuiensis          AH83     2193407.0   \n",
       "6                          Oenococcus sicerae     UCMA15228     1664394.0   \n",
       "7                        Vibrio plantisponsor     CECT 7581     4509247.0   \n",
       "8               Shigella flexneri 2a str. 301           301     4828820.0   \n",
       "9                Aliiroseovarius crassostreae          S044     3367659.0   \n",
       "10                     Mucilaginibacter celer       HYN0043     7149106.0   \n",
       "11                     Novacetimonas hansenii          C110     3549186.0   \n",
       "12        Mycoplasma haemocanis str. Illinois      Illinois      919992.0   \n",
       "13           Sinorhizobium fredii CCBAU 25509   CCBAU 25509     6810864.0   \n",
       "14                      Pandoraea thiooxydans     DSM 25325     4464186.0   \n",
       "15                               Vibrio rarus     LMG 23674     3877193.0   \n",
       "16               Limihaloglobus sulfuriphilus     SM-Chi-D1     3868357.0   \n",
       "17                 Corynebacterium flavescens           OJ8     2758653.0   \n",
       "18                    Cloacibacillus porcorum     CL-84 (T)     3585187.0   \n",
       "20                          Agromyces marinus     DSM 26151     3202016.0   \n",
       "21                   Bradyrhizobium paxllaeri       LMTR 21     8377719.0   \n",
       "22            Paraburkholderia caffeinilytica           CF1     8324638.0   \n",
       "23             Companilactobacillus crustorum         MN047     2316479.0   \n",
       "25                   Streptomyces cavourensis         1AS2a     7600475.0   \n",
       "29                             Massilia oculi    CCUG 43427     5844653.0   \n",
       "30                          Neisseria musculi         NW831     2933071.0   \n",
       "32                       Raoultella terrigena          JH01     5598454.0   \n",
       "33                      Sphingobium fuliginis          KK22     4984172.0   \n",
       "36         Thermoanaerobacter wiegelii Rt8.B1        Rt8.B1     2785056.0   \n",
       "39             Desulfomonile tiedjei DSM 6799      DSM 6799     6527027.0   \n",
       "43                    Halomonas sulfidoxydans  MCCC 1A11059     4491948.0   \n",
       "\n",
       "    Downloaded                           Species  ...  Secondary Breadth  \\\n",
       "0         True                Leuconostoc lactis  ...           0.999951   \n",
       "2         True         Sandaracinus amylolyticus  ...           1.000000   \n",
       "3         True                  Aquifex aeolicus  ...           0.999932   \n",
       "4         True  Candidatus Koribacter versatilis  ...           0.999977   \n",
       "5         True               Swingsia samuiensis  ...           0.999929   \n",
       "6         True                Oenococcus sicerae  ...           0.999871   \n",
       "7         True              Vibrio plantisponsor  ...           0.999960   \n",
       "8         True                 Shigella flexneri  ...           0.999904   \n",
       "9         True      Aliiroseovarius crassostreae  ...           0.999794   \n",
       "10        True            Mucilaginibacter celer  ...           0.999760   \n",
       "11        True            Novacetimonas hansenii  ...           0.999419   \n",
       "12        True             Mycoplasma haemocanis  ...           0.999188   \n",
       "13        True              Sinorhizobium fredii  ...           0.999976   \n",
       "14        True             Pandoraea thiooxydans  ...           0.999933   \n",
       "15        True                      Vibrio rarus  ...           0.999865   \n",
       "16        True      Limihaloglobus sulfuriphilus  ...           0.999950   \n",
       "17        True        Corynebacterium flavescens  ...           0.999901   \n",
       "18        True           Cloacibacillus porcorum  ...           0.999960   \n",
       "20        True                 Agromyces marinus  ...           0.999974   \n",
       "21        True          Bradyrhizobium paxllaeri  ...           0.999984   \n",
       "22        True   Paraburkholderia caffeinilytica  ...           0.999977   \n",
       "23        True    Companilactobacillus crustorum  ...           0.999683   \n",
       "25        True          Streptomyces cavourensis  ...           0.999994   \n",
       "29        True                    Massilia oculi  ...           1.000000   \n",
       "30        True                 Neisseria musculi  ...           0.999963   \n",
       "32        True              Raoultella terrigena  ...           0.999997   \n",
       "33        True             Sphingobium fuliginis  ...           0.999919   \n",
       "36        True       Thermoanaerobacter wiegelii  ...           0.999947   \n",
       "39        True             Desulfomonile tiedjei  ...           0.999981   \n",
       "43        True           Halomonas sulfidoxydans  ...           0.999982   \n",
       "\n",
       "    Secondary Expected  Secondary Score  Secondary Depth  Primary Breadth  \\\n",
       "0             1.000000         0.999951        99.860177         0.999951   \n",
       "2             1.000000         1.000000        49.606000         0.999992   \n",
       "3             1.000000         0.999932        49.586588         0.999932   \n",
       "4             1.000000         0.999977        98.981260         0.999977   \n",
       "5             1.000000         0.999929       100.964187         0.999929   \n",
       "6             1.000000         0.999871        49.581504         0.999871   \n",
       "7             1.000000         0.999960        50.684451         0.999960   \n",
       "8             1.000000         0.999904        22.716979         0.999904   \n",
       "9             1.000000         0.999794        19.906297         0.999794   \n",
       "10            1.000000         0.999760        19.861267         0.999760   \n",
       "11            0.999955         0.999464        10.000399         0.999419   \n",
       "12            1.000000         0.999188        19.812887         0.999188   \n",
       "13            1.000000         0.999976       103.333486         0.999976   \n",
       "14            1.000000         0.999933        19.914034         0.999933   \n",
       "15            1.000000         0.999865        20.375531         0.999865   \n",
       "16            1.000000         0.999950       100.264028         0.999950   \n",
       "17            1.000000         0.999901        50.766405         0.999901   \n",
       "18            1.000000         0.999960        99.919357         0.999943   \n",
       "20            1.000000         0.999974        99.276504         0.999974   \n",
       "21            1.000000         0.999984        50.402800         0.999984   \n",
       "22            1.000000         0.999977        20.064410         0.999488   \n",
       "23            0.999965         0.999719        10.241042         0.999683   \n",
       "25            1.000000         0.999994        49.725614         0.999983   \n",
       "29            1.000000         1.000000        99.814100         0.999962   \n",
       "30            1.000000         0.999963        50.761172         0.999963   \n",
       "32            1.000000         0.999997        99.720803         0.999997   \n",
       "33            1.000000         0.999919        50.386814         0.824513   \n",
       "36            1.000000         0.999947       101.209415         0.999947   \n",
       "39            1.000000         0.999981        20.117984         0.999980   \n",
       "43            1.000000         0.999982        49.732874         0.999982   \n",
       "\n",
       "    Primary Expected  Primary Score  Primary Depth  \\\n",
       "0           1.000000       0.999951      98.808125   \n",
       "2           1.000000       0.999992      49.413092   \n",
       "3           1.000000       0.999932      49.447579   \n",
       "4           1.000000       0.999977      98.947459   \n",
       "5           1.000000       0.999929      99.078553   \n",
       "6           1.000000       0.999871      49.436986   \n",
       "7           1.000000       0.999960      49.405227   \n",
       "8           1.000000       0.999904      19.765296   \n",
       "9           1.000000       0.999794      19.789472   \n",
       "10          1.000000       0.999760      19.789550   \n",
       "11          0.999949       0.999470       9.888364   \n",
       "12          1.000000       0.999188      19.806581   \n",
       "13          1.000000       0.999976     100.712422   \n",
       "14          1.000000       0.999933      19.786825   \n",
       "15          1.000000       0.999865      19.762609   \n",
       "16          1.000000       0.999950      98.883759   \n",
       "17          1.000000       0.999901      48.833515   \n",
       "18          1.000000       0.999943      98.855180   \n",
       "20          1.000000       0.999974      98.883394   \n",
       "21          1.000000       0.999984      49.757890   \n",
       "22          1.000000       0.999488      19.822158   \n",
       "23          0.999950       0.999733       9.902845   \n",
       "25          1.000000       0.999983      47.023317   \n",
       "29          1.000000       0.999962      98.874955   \n",
       "30          1.000000       0.999963      48.780799   \n",
       "32          1.000000       0.999997      98.528117   \n",
       "33          1.000000       0.824513      33.643599   \n",
       "36          1.000000       0.999947      98.311759   \n",
       "39          1.000000       0.999980      19.713291   \n",
       "43          1.000000       0.999982      49.468670   \n",
       "\n",
       "    Combined PS and ANI (Sqrt(ANI)xPSx100)  Presence/Absence  \n",
       "0                                    99.99           Present  \n",
       "2                                   100.00           Present  \n",
       "3                                    99.99           Present  \n",
       "4                                   100.00           Present  \n",
       "5                                    99.99           Present  \n",
       "6                                    99.99           Present  \n",
       "7                                   100.00           Present  \n",
       "8                                    99.95           Present  \n",
       "9                                    99.98           Present  \n",
       "10                                   99.98           Present  \n",
       "11                                   99.95           Present  \n",
       "12                                   99.92           Present  \n",
       "13                                   99.99           Present  \n",
       "14                                   99.99           Present  \n",
       "15                                   99.99           Present  \n",
       "16                                   99.99           Present  \n",
       "17                                   99.99           Present  \n",
       "18                                   99.99           Present  \n",
       "20                                  100.00           Present  \n",
       "21                                   99.99           Present  \n",
       "22                                   99.95           Present  \n",
       "23                                   99.97           Present  \n",
       "25                                  100.00           Present  \n",
       "29                                  100.00           Present  \n",
       "30                                  100.00           Present  \n",
       "32                                  100.00           Present  \n",
       "33                                   82.38           Present  \n",
       "36                                   99.99           Present  \n",
       "39                                  100.00           Present  \n",
       "43                                  100.00           Present  \n",
       "\n",
       "[30 rows x 21 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "downloaded_assemblies[downloaded_assemblies['Presence/Absence'] == 'Present']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "3b8512fd-b936-4e52-86d8-466c5676d5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "ground_truth_abd = {\n",
    "\"Leuconostoc lactis\": 10.0,\n",
    "\"Cloacibacillus porcorum\": 10.0,\n",
    "\"Limihaloglobus sulfuriphilus\": 10.0,\n",
    "\"Sinorhizobium fredii\": 10.0,\n",
    "\"Massilia oculi\": 10.0,\n",
    "\"Raoultella terrigena\": 10.0,\n",
    "\"Agromyces marinus\": 10.0,\n",
    "\"Thermoanaerobacter wiegelii\": 10.0,\n",
    "\"Candidatus Koribacter versatilis\": 10.0,\n",
    "\"Swingsia samuiensis\": 10.0,\n",
    "\"Sphingobium fuliginis\": 5.0,\n",
    "\"Aquifex aeolicus\": 5.0,\n",
    "\"Corynebacterium flavescens\": 5.0,\n",
    "\"Oenococcus sicerae\": 5.0,\n",
    "\"Vibrio plantisponsor\": 5.0,\n",
    "\"Bradyrhizobium paxllaeri\": 5.0,\n",
    "\"Sandaracinus amylolyticus\": 5.0,\n",
    "\"Halomonas sulfidoxydans\": 5.0,\n",
    "\"Neisseria musculi\": 5.0,\n",
    "\"Streptomyces cavourensis\": 5.0,\n",
    "\"Desulfomonile tiedjei\": 2.0,\n",
    "\"Mycoplasma haemocanis\": 2.0,\n",
    "\"Paraburkholderia caffeinilytica\": 2.0,\n",
    "\"Shigella flexneri\": 2.0,\n",
    "\"Vibrio rarus\": 2.0,\n",
    "\"Pandoraea thiooxydans\": 2.0,\n",
    "\"Aliiroseovarius crassostreae\": 2.0,\n",
    "\"Mucilaginibacter celer\": 2.0,\n",
    "\"Novacetimonas hansenii\": 1.0,\n",
    "\"Companilactobacillus crustorum\": 1.0,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "cbaa30f6-cb33-4432-844a-c49d5e2f1c7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ground_truth_set = set(ground_truth_abd.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "1054491d-1dc8-4927-99e8-9f5248b0d170",
   "metadata": {},
   "outputs": [],
   "source": [
    "positive_calls = set(downloaded_assemblies[downloaded_assemblies['Presence/Absence'] == 'Present']['Species'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a929c391-14ee-4dfa-a1f3-5b9512d70689",
   "metadata": {},
   "outputs": [],
   "source": [
    "#positive_calls = set(downloaded_assemblies[downloaded_assemblies['Combined CS2 and ANI (Sqrt(ANI)xCS2x100)'] >= 97]['Species'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ac936dc5-1f7c-488f-be95-69e916946435",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(positive_calls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "7e65f836-5fbe-4658-804f-015ea087f988",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ground_truth_set.intersection(positive_calls))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "c067fa94-005c-4907-9ad8-feac96258c78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ground_truth_set - positive_calls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "c91cf46d-c61c-4d97-959f-f03015890096",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(positive_calls - ground_truth_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "2cec6c88-06f2-40a8-a73b-73aeacd9b560",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Assembly Accession ID</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>Is Representative</th>\n",
       "      <th>Assembly Level</th>\n",
       "      <th>Organism of Assembly</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Total Length</th>\n",
       "      <th>Downloaded</th>\n",
       "      <th>Species</th>\n",
       "      <th>...</th>\n",
       "      <th>Secondary Breadth</th>\n",
       "      <th>Secondary Expected</th>\n",
       "      <th>Secondary Score</th>\n",
       "      <th>Secondary Depth</th>\n",
       "      <th>Primary Breadth</th>\n",
       "      <th>Primary Expected</th>\n",
       "      <th>Primary Score</th>\n",
       "      <th>Primary Depth</th>\n",
       "      <th>Combined PS and ANI (Sqrt(ANI)xPSx100)</th>\n",
       "      <th>Presence/Absence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Taxonomy ID, Assembly Accession ID, Source Database, Is Representative, Assembly Level, Organism of Assembly, Strain, Total Length, Downloaded, Species, Consensus ANI, Secondary Breadth, Secondary Expected, Secondary Score, Secondary Depth, Primary Breadth, Primary Expected, Primary Score, Primary Depth, Combined PS and ANI (Sqrt(ANI)xPSx100), Presence/Absence]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 21 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "downloaded_assemblies[downloaded_assemblies['Species'].isin(list(positive_calls - ground_truth_set))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dca57ad9-06c2-40c7-9db2-1f0934b1e984",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
