{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "251cef2a-37b5-46e6-9fda-4546d1c14315",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import os\n",
    "import argparse\n",
    "import sys\n",
    "import subprocess\n",
    "import math\n",
    "import pickle\n",
    "from collections import defaultdict\n",
    "import json\n",
    "import pandas as pd\n",
    "import math\n",
    "from ete3 import NCBITaxa\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3911e5c3-e6b4-41f6-a06b-0fc2337e0e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.reference_finder import prepare_reference_genomes\n",
    "from utils.alignment import run_minimap2, sort_samfile, run_bwa, samtools_calculate_depth\n",
    "from utils.summary import alignment_1_summary, alignment_2_summary, merge_reference_fasta, call_present_absent\n",
    "from utils.ani import samtools_merged_consensus, ani_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2b2917e3-e287-47cb-a01e-5df068b70a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_species_taxid(taxid, ncbi_taxa_db, valid_kingdom):\n",
    "    lineage = ncbi_taxa_db.get_lineage(taxid)\n",
    "    if bool(set(lineage) & valid_kingdom):\n",
    "        taxid2rank_dict = ncbi_taxa_db.get_rank(lineage)\n",
    "        for lineage_taxid in taxid2rank_dict:\n",
    "            if taxid2rank_dict[lineage_taxid] == 'species':\n",
    "                return lineage_taxid\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f15d60c-3721-4449-a4ba-f523ce4a32b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_kraken2_taxonomy(kraken2_report, min_frac, ncbi_taxa_db, valid_kingdom):\n",
    "    classification_result_df = pd.read_csv(kraken2_report, sep='\\t', names=['Abundance', 'Cumulative Count', 'Count', 'Rank', 'Taxid', 'Taxname']) \n",
    "    taxids = classification_result_df[(classification_result_df['Rank'] == 'S') & (classification_result_df['Abundance'] >= min_frac*100)]['Taxid'].tolist()\n",
    "    valid_taxids = []\n",
    "    for taxid in taxids:\n",
    "        species_taxid = get_species_taxid(taxid, ncbi_taxa_db, valid_kingdom)\n",
    "        if species_taxid is not None:\n",
    "            valid_taxids.append(species_taxid)\n",
    "    return valid_taxids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b5a8b208-4aa2-481c-8e3c-927c6634ff8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_taxa_db = NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6747fe53-a872-4a7a-a56f-9a81745db890",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_fastq_1 = '../test_data/Zymo_illumina/SRR11207265_1.downsampled.fastq'\n",
    "input_fastq_2 = '../test_data/Zymo_illumina/SRR11207265_2.downsampled.fastq'\n",
    "kraken2_report = \"../test_data/Zymo_illumina/SRR11207265.report\"\n",
    "threads = 40\n",
    "min_mapq = 20\n",
    "min_coverage_score = 0.7\n",
    "min_frac = 0.0005"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "1f4939e6-c628-4af1-99f8-e76945f76e50",
   "metadata": {},
   "outputs": [],
   "source": [
    "working_directory = '../test_output_downsampled_0002'\n",
    "if not os.path.exists(working_directory):\n",
    "    os.mkdir(working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9f8c9842-3089-4a3f-8230-de0ba518830a",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxid_queries = filter_kraken2_taxonomy(kraken2_report,\n",
    "                                        min_frac=min_frac, \n",
    "                                        ncbi_taxa_db=ncbi_taxa_db, \n",
    "                                        valid_kingdom={2, 4751, 2157, 10239})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "6e1de4ec-7aa2-45c2-adb5-454bbe04b719",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(taxid_queries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f5bf1e11-2199-42ac-9550-d3af818b2d66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28901      \t GCF_000006945.2 \t Salmonella enterica subsp. e.. \t LT2        \t Complete Genome\n",
      "54736      \t GCF_000439255.1 \t Salmonella bongori N268-08     \t N268-08    \t Complete Genome\n",
      "562        \t GCF_000008865.2 \t Escherichia coli O157:H7 str.. \t Sakai su.. \t Complete Genome\n",
      "208962     \t GCF_016904755.1 \t Escherichia albertii           \t Sample 1.. \t Complete Genome\n",
      "1499973    \t GCF_002900365.1 \t Escherichia marmotae           \t HT073016   \t Complete Genome\n",
      "564        \t GCF_013892435.1 \t Escherichia fergusonii         \t RHB19-C0.. \t Complete Genome\n",
      "2044467    \t GCF_005843885.1 \t Escherichia sp. E4742          \t E4742      \t Complete Genome\n",
      "546        \t GCF_003812345.1 \t Citrobacter freundii           \t FDAARGOS.. \t Complete Genome\n",
      "623        \t GCF_000006925.2 \t Shigella flexneri 2a str. 30.. \t 301        \t Complete Genome\n",
      "621        \t GCF_001027225.1 \t Shigella boydii                \t ATCC 921.. \t Complete Genome\n",
      "622        \t GCF_002741615.1 \t Shigella dysenteriae           \t BU53M1     \t Complete Genome\n",
      "624        \t GCF_013374815.1 \t Shigella sonnei                \t SE6-1      \t Complete Genome\n",
      "573        \t GCF_000240185.1 \t Klebsiella pneumoniae subsp... \t HS11286    \t Complete Genome\n",
      "287        \t GCF_000006765.1 \t Pseudomonas aeruginosa PAO1    \t PAO1       \t Complete Genome\n",
      "294        \t GCF_000730425.1 \t Pseudomonas fluorescens        \t UK4        \t Complete Genome\n",
      "303        \t GCF_000412675.1 \t Pseudomonas putida NBRC 1416.. \t NBRC 141.. \t Complete Genome\n",
      "316        \t GCF_019704535.1 \t Stutzerimonas stutzeri         \t F2a        \t Complete Genome\n",
      "1423       \t GCF_000009045.1 \t Bacillus subtilis subsp. sub.. \t 168        \t Complete Genome\n",
      "72361      \t GCF_003667885.1 \t Bacillus vallismortis          \t Bac111     \t Complete Genome\n",
      "227866     \t GCF_012225885.1 \t Bacillus tequilensis           \t EA-CB001.. \t Complete Genome\n",
      "72360      \t GCF_012648005.1 \t Bacillus mojavensis            \t UCMB5075   \t Complete Genome\n",
      "260554     \t GCF_004006435.1 \t Bacillus halotolerans          \t ZB201702   \t Complete Genome\n",
      "96241      \t GCF_006094475.1 \t Bacillus subtilis subsp. spi.. \t ATCC 663.. \t Complete Genome\n",
      "492670     \t GCF_000769555.1 \t Bacillus velezensis            \t JS25R      \t Complete Genome\n",
      "483913     \t GCF_003148415.1 \t Bacillus inaquosorum           \t KCTC 134.. \t Complete Genome\n",
      "1963032    \t GCF_000959025.1 \t Bacillus intestinalis          \t T30        \t Complete Genome\n",
      "1280       \t GCF_000013425.1 \t Staphylococcus aureus subsp... \t NCTC 832.. \t Complete Genome\n",
      "2699836    \t GCF_022493035.1 \t Staphylococcus roterodami      \t BDS-53E    \t Complete Genome\n",
      "1639       \t GCF_000196035.1 \t Listeria monocytogenes EGD-e   \t EGD-e      \t Complete Genome\n",
      "1351       \t GCF_001598635.1 \t Enterococcus faecalis          \t LD33       \t Complete Genome\n",
      "1613       \t GCF_022819245.1 \t Limosilactobacillus fermentu.. \t SCB0035    \t Complete Genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "31 archives were successfully processed.\n"
     ]
    }
   ],
   "source": [
    "reference_metadata = prepare_reference_genomes(taxid_queries, working_directory, ncbi_taxa_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3b6276e5-36bd-4cfd-a37a-907f3c66c8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = reference_metadata[reference_metadata['Downloaded']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7aacc8f4-314a-41a0-8b1d-c16f3fd9e1de",
   "metadata": {},
   "outputs": [],
   "source": [
    "for assembly_id in downloaded_assemblies['Assembly Accession ID']:\n",
    "    reference_fasta = os.path.join(working_directory, 'reference_genomes', f'{assembly_id}.fasta')\n",
    "    run_bwa(input_fastq_1, input_fastq_2, reference_fasta, assembly_id, working_directory, threads=threads)\n",
    "    sort_samfile(assembly_id, working_directory, min_mapq=min_mapq, threads=threads)\n",
    "    samtools_calculate_depth(assembly_id, working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e4316f7e-c07a-4078-9511-3faf482968d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = alignment_1_summary(downloaded_assemblies, working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f6cf7b01-2f5c-4975-8725-3a9f5717a40e",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_assemblies = list(downloaded_assemblies[downloaded_assemblies['Coverage Score'] >= min_coverage_score]['Assembly Accession ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5b7a62e5-2ffc-495e-85d4-852339ed1256",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_fasta = merge_reference_fasta(filtered_assemblies, working_directory)\n",
    "run_bwa(input_fastq_1, input_fastq_2, reference_fasta, 'merged', working_directory, threads=threads)\n",
    "sort_samfile('merged', working_directory, min_mapq=0, threads=threads)\n",
    "samtools_calculate_depth('merged', working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "766ba8bc-2cc6-4d24-866b-9e6c2b30128e",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = alignment_2_summary(downloaded_assemblies, working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9736777b-5376-4ba9-b750-c8f628ad2ef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_record_dict = samtools_merged_consensus(working_directory, threads)\n",
    "downloaded_assemblies = ani_summary(downloaded_assemblies, consensus_record_dict, working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "052662c7-a05d-49f7-a88e-e54eb58910e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = call_present_absent(downloaded_assemblies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fb5825f7-1eaa-45b2-a984-fc47188e5fdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies.sort_values(['CS2'], ascending=False).to_csv(os.path.join(working_directory, 'reference_alignment.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d9addeba-3d0b-415e-90e0-8349530e74aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Assembly Accession ID</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>Is Representative</th>\n",
       "      <th>Assembly Level</th>\n",
       "      <th>Organism of Assembly</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Downloaded</th>\n",
       "      <th>Species</th>\n",
       "      <th>Breadth Coverage</th>\n",
       "      <th>Expected Coverage</th>\n",
       "      <th>Coverage Score</th>\n",
       "      <th>Depth Coverage</th>\n",
       "      <th>BC2</th>\n",
       "      <th>EC2</th>\n",
       "      <th>CS2</th>\n",
       "      <th>DC2</th>\n",
       "      <th>Consensus ANI</th>\n",
       "      <th>Combined CS2 and ANI (Sqrt(ANI)xCS2x100)</th>\n",
       "      <th>Presence/Absence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>28901</td>\n",
       "      <td>GCF_000006945.2</td>\n",
       "      <td>Washington University Genome Sequencing Center</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>LT2</td>\n",
       "      <td>True</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>0.907640</td>\n",
       "      <td>0.999430</td>\n",
       "      <td>0.908158</td>\n",
       "      <td>8.230061</td>\n",
       "      <td>0.912043</td>\n",
       "      <td>0.997296</td>\n",
       "      <td>0.914515</td>\n",
       "      <td>6.483257</td>\n",
       "      <td>0.989812</td>\n",
       "      <td>90.98</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>54736</td>\n",
       "      <td>GCF_000439255.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Salmonella bongori N268-08</td>\n",
       "      <td>N268-08</td>\n",
       "      <td>True</td>\n",
       "      <td>Salmonella bongori</td>\n",
       "      <td>0.743137</td>\n",
       "      <td>0.997109</td>\n",
       "      <td>0.745292</td>\n",
       "      <td>7.866824</td>\n",
       "      <td>0.022666</td>\n",
       "      <td>0.077416</td>\n",
       "      <td>0.292787</td>\n",
       "      <td>3.554857</td>\n",
       "      <td>0.967281</td>\n",
       "      <td>28.80</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>562</td>\n",
       "      <td>GCF_000008865.2</td>\n",
       "      <td>GIRC</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia coli O157:H7 str. Sakai</td>\n",
       "      <td>Sakai substr. RIMD 0509952</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>0.751993</td>\n",
       "      <td>0.998130</td>\n",
       "      <td>0.753402</td>\n",
       "      <td>8.353340</td>\n",
       "      <td>0.372702</td>\n",
       "      <td>0.659305</td>\n",
       "      <td>0.565296</td>\n",
       "      <td>2.889042</td>\n",
       "      <td>0.988459</td>\n",
       "      <td>56.20</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>208962</td>\n",
       "      <td>GCF_016904755.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia albertii</td>\n",
       "      <td>Sample 167</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia albertii</td>\n",
       "      <td>0.733252</td>\n",
       "      <td>0.997331</td>\n",
       "      <td>0.735214</td>\n",
       "      <td>8.081780</td>\n",
       "      <td>0.012326</td>\n",
       "      <td>0.029806</td>\n",
       "      <td>0.413548</td>\n",
       "      <td>2.454837</td>\n",
       "      <td>0.966724</td>\n",
       "      <td>40.66</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1499973</td>\n",
       "      <td>GCF_002900365.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia marmotae</td>\n",
       "      <td>HT073016</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia marmotae</td>\n",
       "      <td>0.730958</td>\n",
       "      <td>0.997523</td>\n",
       "      <td>0.732773</td>\n",
       "      <td>8.209433</td>\n",
       "      <td>0.019633</td>\n",
       "      <td>0.059327</td>\n",
       "      <td>0.330929</td>\n",
       "      <td>3.115106</td>\n",
       "      <td>0.965896</td>\n",
       "      <td>32.52</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>564</td>\n",
       "      <td>GCF_013892435.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia fergusonii</td>\n",
       "      <td>RHB19-C05</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia fergusonii</td>\n",
       "      <td>0.614433</td>\n",
       "      <td>0.993791</td>\n",
       "      <td>0.618272</td>\n",
       "      <td>8.270563</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2044467</td>\n",
       "      <td>GCF_005843885.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia sp. E4742</td>\n",
       "      <td>E4742</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia sp. E4742</td>\n",
       "      <td>0.734095</td>\n",
       "      <td>0.997641</td>\n",
       "      <td>0.735830</td>\n",
       "      <td>8.240902</td>\n",
       "      <td>0.038282</td>\n",
       "      <td>0.086245</td>\n",
       "      <td>0.443874</td>\n",
       "      <td>2.355988</td>\n",
       "      <td>0.971706</td>\n",
       "      <td>43.75</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>546</td>\n",
       "      <td>GCF_003812345.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Citrobacter freundii</td>\n",
       "      <td>FDAARGOS_549</td>\n",
       "      <td>True</td>\n",
       "      <td>Citrobacter freundii</td>\n",
       "      <td>0.478365</td>\n",
       "      <td>0.968910</td>\n",
       "      <td>0.493714</td>\n",
       "      <td>7.255633</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>623</td>\n",
       "      <td>GCF_000006925.2</td>\n",
       "      <td>Microbial Genome Center of ChMPH</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Shigella flexneri 2a str. 301</td>\n",
       "      <td>301</td>\n",
       "      <td>True</td>\n",
       "      <td>Shigella flexneri</td>\n",
       "      <td>0.794622</td>\n",
       "      <td>0.998802</td>\n",
       "      <td>0.795576</td>\n",
       "      <td>8.465387</td>\n",
       "      <td>0.387796</td>\n",
       "      <td>0.633569</td>\n",
       "      <td>0.612083</td>\n",
       "      <td>2.588805</td>\n",
       "      <td>0.990129</td>\n",
       "      <td>60.91</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>621</td>\n",
       "      <td>GCF_001027225.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Shigella boydii</td>\n",
       "      <td>ATCC 9210</td>\n",
       "      <td>True</td>\n",
       "      <td>Shigella boydii</td>\n",
       "      <td>0.821040</td>\n",
       "      <td>0.999083</td>\n",
       "      <td>0.821793</td>\n",
       "      <td>8.519167</td>\n",
       "      <td>0.459718</td>\n",
       "      <td>0.715478</td>\n",
       "      <td>0.642533</td>\n",
       "      <td>2.734124</td>\n",
       "      <td>0.991279</td>\n",
       "      <td>63.97</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>622</td>\n",
       "      <td>GCF_002741615.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Shigella dysenteriae</td>\n",
       "      <td>BU53M1</td>\n",
       "      <td>True</td>\n",
       "      <td>Shigella dysenteriae</td>\n",
       "      <td>0.733746</td>\n",
       "      <td>0.998104</td>\n",
       "      <td>0.735139</td>\n",
       "      <td>8.542506</td>\n",
       "      <td>0.260939</td>\n",
       "      <td>0.461075</td>\n",
       "      <td>0.565935</td>\n",
       "      <td>2.369026</td>\n",
       "      <td>0.988213</td>\n",
       "      <td>56.26</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>624</td>\n",
       "      <td>GCF_013374815.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Shigella sonnei</td>\n",
       "      <td>SE6-1</td>\n",
       "      <td>True</td>\n",
       "      <td>Shigella sonnei</td>\n",
       "      <td>0.887203</td>\n",
       "      <td>0.999406</td>\n",
       "      <td>0.887730</td>\n",
       "      <td>8.372347</td>\n",
       "      <td>0.608641</td>\n",
       "      <td>0.869903</td>\n",
       "      <td>0.699665</td>\n",
       "      <td>3.350823</td>\n",
       "      <td>0.991290</td>\n",
       "      <td>69.66</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>573</td>\n",
       "      <td>GCF_000240185.1</td>\n",
       "      <td>Bioinformatics &amp; Microbial Genomics, State Ke...</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae HS11286</td>\n",
       "      <td>HS11286</td>\n",
       "      <td>True</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>0.295883</td>\n",
       "      <td>0.869641</td>\n",
       "      <td>0.340235</td>\n",
       "      <td>6.885986</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>287</td>\n",
       "      <td>GCF_000006765.1</td>\n",
       "      <td>PathoGenesis Corporation</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>True</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>0.945966</td>\n",
       "      <td>0.987809</td>\n",
       "      <td>0.957641</td>\n",
       "      <td>4.658703</td>\n",
       "      <td>0.951679</td>\n",
       "      <td>0.987619</td>\n",
       "      <td>0.963609</td>\n",
       "      <td>4.614541</td>\n",
       "      <td>0.994470</td>\n",
       "      <td>96.09</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>294</td>\n",
       "      <td>GCF_000730425.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pseudomonas fluorescens</td>\n",
       "      <td>UK4</td>\n",
       "      <td>True</td>\n",
       "      <td>Pseudomonas fluorescens</td>\n",
       "      <td>0.180846</td>\n",
       "      <td>0.474090</td>\n",
       "      <td>0.381459</td>\n",
       "      <td>3.553416</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>303</td>\n",
       "      <td>GCF_000412675.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pseudomonas putida NBRC 14164</td>\n",
       "      <td>NBRC 14164</td>\n",
       "      <td>True</td>\n",
       "      <td>Pseudomonas putida</td>\n",
       "      <td>0.197449</td>\n",
       "      <td>0.500223</td>\n",
       "      <td>0.394722</td>\n",
       "      <td>3.512750</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>316</td>\n",
       "      <td>GCF_019704535.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Stutzerimonas stutzeri</td>\n",
       "      <td>F2a</td>\n",
       "      <td>True</td>\n",
       "      <td>Stutzerimonas stutzeri</td>\n",
       "      <td>0.241283</td>\n",
       "      <td>0.590584</td>\n",
       "      <td>0.408550</td>\n",
       "      <td>3.701112</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1423</td>\n",
       "      <td>GCF_000009045.1</td>\n",
       "      <td>BSNR</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>168</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>0.762025</td>\n",
       "      <td>0.882021</td>\n",
       "      <td>0.863954</td>\n",
       "      <td>2.804649</td>\n",
       "      <td>0.002772</td>\n",
       "      <td>0.005465</td>\n",
       "      <td>0.507204</td>\n",
       "      <td>1.976979</td>\n",
       "      <td>0.989899</td>\n",
       "      <td>50.46</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>72361</td>\n",
       "      <td>GCF_003667885.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus vallismortis</td>\n",
       "      <td>Bac111</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus vallismortis</td>\n",
       "      <td>0.774163</td>\n",
       "      <td>0.885038</td>\n",
       "      <td>0.874723</td>\n",
       "      <td>2.794134</td>\n",
       "      <td>0.002715</td>\n",
       "      <td>0.003215</td>\n",
       "      <td>0.844591</td>\n",
       "      <td>1.185895</td>\n",
       "      <td>0.995716</td>\n",
       "      <td>84.28</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>227866</td>\n",
       "      <td>GCF_012225885.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus tequilensis</td>\n",
       "      <td>EA-CB0015</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus tequilensis</td>\n",
       "      <td>0.741291</td>\n",
       "      <td>0.873268</td>\n",
       "      <td>0.848871</td>\n",
       "      <td>2.786549</td>\n",
       "      <td>0.001899</td>\n",
       "      <td>0.002372</td>\n",
       "      <td>0.800479</td>\n",
       "      <td>1.250722</td>\n",
       "      <td>0.992905</td>\n",
       "      <td>79.76</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>72360</td>\n",
       "      <td>GCF_012648005.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus mojavensis</td>\n",
       "      <td>UCMB5075</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus mojavensis</td>\n",
       "      <td>0.683937</td>\n",
       "      <td>0.832525</td>\n",
       "      <td>0.821522</td>\n",
       "      <td>2.612655</td>\n",
       "      <td>0.002909</td>\n",
       "      <td>0.003533</td>\n",
       "      <td>0.823596</td>\n",
       "      <td>1.216320</td>\n",
       "      <td>0.996587</td>\n",
       "      <td>82.22</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>260554</td>\n",
       "      <td>GCF_004006435.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus halotolerans</td>\n",
       "      <td>ZB201702</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus halotolerans</td>\n",
       "      <td>0.677839</td>\n",
       "      <td>0.830789</td>\n",
       "      <td>0.815898</td>\n",
       "      <td>2.620952</td>\n",
       "      <td>0.002948</td>\n",
       "      <td>0.003706</td>\n",
       "      <td>0.795317</td>\n",
       "      <td>1.259677</td>\n",
       "      <td>0.997138</td>\n",
       "      <td>79.42</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>96241</td>\n",
       "      <td>GCF_006094475.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus subtilis subsp. spizizenii ATCC 6633 ...</td>\n",
       "      <td>ATCC 6633</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus spizizenii</td>\n",
       "      <td>0.905485</td>\n",
       "      <td>0.923127</td>\n",
       "      <td>0.980888</td>\n",
       "      <td>2.833352</td>\n",
       "      <td>0.709671</td>\n",
       "      <td>0.724661</td>\n",
       "      <td>0.979315</td>\n",
       "      <td>1.817361</td>\n",
       "      <td>0.998095</td>\n",
       "      <td>97.84</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>492670</td>\n",
       "      <td>GCF_000769555.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus velezensis</td>\n",
       "      <td>JS25R</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus velezensis</td>\n",
       "      <td>0.202223</td>\n",
       "      <td>0.383401</td>\n",
       "      <td>0.527445</td>\n",
       "      <td>2.391079</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>483913</td>\n",
       "      <td>GCF_003148415.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus inaquosorum</td>\n",
       "      <td>KCTC 13429</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus inaquosorum</td>\n",
       "      <td>0.763588</td>\n",
       "      <td>0.884493</td>\n",
       "      <td>0.863306</td>\n",
       "      <td>2.826641</td>\n",
       "      <td>0.003876</td>\n",
       "      <td>0.015586</td>\n",
       "      <td>0.248670</td>\n",
       "      <td>4.053022</td>\n",
       "      <td>0.987319</td>\n",
       "      <td>24.71</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1963032</td>\n",
       "      <td>GCF_000959025.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus intestinalis</td>\n",
       "      <td>T30</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus intestinalis</td>\n",
       "      <td>0.908677</td>\n",
       "      <td>0.924061</td>\n",
       "      <td>0.983351</td>\n",
       "      <td>2.836854</td>\n",
       "      <td>0.706171</td>\n",
       "      <td>0.719776</td>\n",
       "      <td>0.981098</td>\n",
       "      <td>1.801468</td>\n",
       "      <td>0.998096</td>\n",
       "      <td>98.02</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1280</td>\n",
       "      <td>GCF_000013425.1</td>\n",
       "      <td>University of Oklahoma Health Sciences Center</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>NCTC 8325</td>\n",
       "      <td>True</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>0.866442</td>\n",
       "      <td>0.950488</td>\n",
       "      <td>0.911576</td>\n",
       "      <td>3.468753</td>\n",
       "      <td>0.866030</td>\n",
       "      <td>0.947037</td>\n",
       "      <td>0.914463</td>\n",
       "      <td>3.392590</td>\n",
       "      <td>0.992470</td>\n",
       "      <td>91.10</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>2699836</td>\n",
       "      <td>GCF_022493035.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Staphylococcus roterodami</td>\n",
       "      <td>BDS-53E</td>\n",
       "      <td>True</td>\n",
       "      <td>Staphylococcus roterodami</td>\n",
       "      <td>0.701952</td>\n",
       "      <td>0.891262</td>\n",
       "      <td>0.787593</td>\n",
       "      <td>3.160850</td>\n",
       "      <td>0.031307</td>\n",
       "      <td>0.062633</td>\n",
       "      <td>0.499844</td>\n",
       "      <td>2.065971</td>\n",
       "      <td>0.971264</td>\n",
       "      <td>49.26</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1639</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "      <td>European Consortium</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Listeria monocytogenes EGD-e</td>\n",
       "      <td>EGD-e</td>\n",
       "      <td>True</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>0.817813</td>\n",
       "      <td>0.892041</td>\n",
       "      <td>0.916789</td>\n",
       "      <td>2.721834</td>\n",
       "      <td>0.825120</td>\n",
       "      <td>0.888395</td>\n",
       "      <td>0.928776</td>\n",
       "      <td>2.657477</td>\n",
       "      <td>0.952034</td>\n",
       "      <td>90.62</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1351</td>\n",
       "      <td>GCF_001598635.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Enterococcus faecalis</td>\n",
       "      <td>LD33</td>\n",
       "      <td>True</td>\n",
       "      <td>Enterococcus faecalis</td>\n",
       "      <td>0.869026</td>\n",
       "      <td>0.989935</td>\n",
       "      <td>0.877861</td>\n",
       "      <td>5.291694</td>\n",
       "      <td>0.877028</td>\n",
       "      <td>0.990105</td>\n",
       "      <td>0.885793</td>\n",
       "      <td>5.262746</td>\n",
       "      <td>0.991390</td>\n",
       "      <td>88.20</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1613</td>\n",
       "      <td>GCF_022819245.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Limosilactobacillus fermentum</td>\n",
       "      <td>SCB0035</td>\n",
       "      <td>True</td>\n",
       "      <td>Limosilactobacillus fermentum</td>\n",
       "      <td>0.765433</td>\n",
       "      <td>0.912251</td>\n",
       "      <td>0.839059</td>\n",
       "      <td>3.178842</td>\n",
       "      <td>0.799532</td>\n",
       "      <td>0.921203</td>\n",
       "      <td>0.867922</td>\n",
       "      <td>3.177840</td>\n",
       "      <td>0.992856</td>\n",
       "      <td>86.48</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Taxonomy ID Assembly Accession ID  \\\n",
       "0         28901       GCF_000006945.2   \n",
       "1         54736       GCF_000439255.1   \n",
       "2           562       GCF_000008865.2   \n",
       "3        208962       GCF_016904755.1   \n",
       "4       1499973       GCF_002900365.1   \n",
       "5           564       GCF_013892435.1   \n",
       "6       2044467       GCF_005843885.1   \n",
       "7           546       GCF_003812345.1   \n",
       "8           623       GCF_000006925.2   \n",
       "9           621       GCF_001027225.1   \n",
       "10          622       GCF_002741615.1   \n",
       "11          624       GCF_013374815.1   \n",
       "12          573       GCF_000240185.1   \n",
       "13          287       GCF_000006765.1   \n",
       "14          294       GCF_000730425.1   \n",
       "15          303       GCF_000412675.1   \n",
       "16          316       GCF_019704535.1   \n",
       "17         1423       GCF_000009045.1   \n",
       "18        72361       GCF_003667885.1   \n",
       "19       227866       GCF_012225885.1   \n",
       "20        72360       GCF_012648005.1   \n",
       "21       260554       GCF_004006435.1   \n",
       "22        96241       GCF_006094475.1   \n",
       "23       492670       GCF_000769555.1   \n",
       "24       483913       GCF_003148415.1   \n",
       "25      1963032       GCF_000959025.1   \n",
       "26         1280       GCF_000013425.1   \n",
       "27      2699836       GCF_022493035.1   \n",
       "28         1639       GCF_000196035.1   \n",
       "29         1351       GCF_001598635.1   \n",
       "30         1613       GCF_022819245.1   \n",
       "\n",
       "                                      Source Database  Is Representative  \\\n",
       "0      Washington University Genome Sequencing Center               True   \n",
       "1                                         NCBI RefSeq               True   \n",
       "2                                                GIRC               True   \n",
       "3                                         NCBI RefSeq               True   \n",
       "4                                         NCBI RefSeq              False   \n",
       "5                                         NCBI RefSeq               True   \n",
       "6                                         NCBI RefSeq              False   \n",
       "7                                         NCBI RefSeq               True   \n",
       "8                    Microbial Genome Center of ChMPH               True   \n",
       "9                                         NCBI RefSeq              False   \n",
       "10                                        NCBI RefSeq              False   \n",
       "11                                        NCBI RefSeq               True   \n",
       "12   Bioinformatics & Microbial Genomics, State Ke...               True   \n",
       "13                           PathoGenesis Corporation               True   \n",
       "14                                        NCBI RefSeq              False   \n",
       "15                                        NCBI RefSeq               True   \n",
       "16                                        NCBI RefSeq               True   \n",
       "17                                               BSNR               True   \n",
       "18                                        NCBI RefSeq               True   \n",
       "19                                        NCBI RefSeq               True   \n",
       "20                                        NCBI RefSeq               True   \n",
       "21                                        NCBI RefSeq               True   \n",
       "22                                        NCBI RefSeq               True   \n",
       "23                                        NCBI RefSeq               True   \n",
       "24                                        NCBI RefSeq              False   \n",
       "25                                        NCBI RefSeq              False   \n",
       "26      University of Oklahoma Health Sciences Center               True   \n",
       "27                                        NCBI RefSeq              False   \n",
       "28                                European Consortium               True   \n",
       "29                                        NCBI RefSeq              False   \n",
       "30                                        NCBI RefSeq               True   \n",
       "\n",
       "     Assembly Level                               Organism of Assembly  \\\n",
       "0   Complete Genome  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "1   Complete Genome                         Salmonella bongori N268-08   \n",
       "2   Complete Genome                Escherichia coli O157:H7 str. Sakai   \n",
       "3   Complete Genome                               Escherichia albertii   \n",
       "4   Complete Genome                               Escherichia marmotae   \n",
       "5   Complete Genome                             Escherichia fergusonii   \n",
       "6   Complete Genome                              Escherichia sp. E4742   \n",
       "7   Complete Genome                               Citrobacter freundii   \n",
       "8   Complete Genome                      Shigella flexneri 2a str. 301   \n",
       "9   Complete Genome                                    Shigella boydii   \n",
       "10  Complete Genome                               Shigella dysenteriae   \n",
       "11  Complete Genome                                    Shigella sonnei   \n",
       "12  Complete Genome    Klebsiella pneumoniae subsp. pneumoniae HS11286   \n",
       "13  Complete Genome                        Pseudomonas aeruginosa PAO1   \n",
       "14  Complete Genome                            Pseudomonas fluorescens   \n",
       "15  Complete Genome                      Pseudomonas putida NBRC 14164   \n",
       "16  Complete Genome                             Stutzerimonas stutzeri   \n",
       "17  Complete Genome         Bacillus subtilis subsp. subtilis str. 168   \n",
       "18  Complete Genome                              Bacillus vallismortis   \n",
       "19  Complete Genome                               Bacillus tequilensis   \n",
       "20  Complete Genome                                Bacillus mojavensis   \n",
       "21  Complete Genome                              Bacillus halotolerans   \n",
       "22  Complete Genome  Bacillus subtilis subsp. spizizenii ATCC 6633 ...   \n",
       "23  Complete Genome                                Bacillus velezensis   \n",
       "24  Complete Genome                               Bacillus inaquosorum   \n",
       "25  Complete Genome                              Bacillus intestinalis   \n",
       "26  Complete Genome      Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "27  Complete Genome                          Staphylococcus roterodami   \n",
       "28  Complete Genome                       Listeria monocytogenes EGD-e   \n",
       "29  Complete Genome                              Enterococcus faecalis   \n",
       "30  Complete Genome                      Limosilactobacillus fermentum   \n",
       "\n",
       "                        Strain  Downloaded                        Species  \\\n",
       "0                          LT2        True            Salmonella enterica   \n",
       "1                      N268-08        True             Salmonella bongori   \n",
       "2   Sakai substr. RIMD 0509952        True               Escherichia coli   \n",
       "3                   Sample 167        True           Escherichia albertii   \n",
       "4                     HT073016        True           Escherichia marmotae   \n",
       "5                    RHB19-C05        True         Escherichia fergusonii   \n",
       "6                        E4742        True          Escherichia sp. E4742   \n",
       "7                 FDAARGOS_549        True           Citrobacter freundii   \n",
       "8                          301        True              Shigella flexneri   \n",
       "9                    ATCC 9210        True                Shigella boydii   \n",
       "10                      BU53M1        True           Shigella dysenteriae   \n",
       "11                       SE6-1        True                Shigella sonnei   \n",
       "12                     HS11286        True          Klebsiella pneumoniae   \n",
       "13                        PAO1        True         Pseudomonas aeruginosa   \n",
       "14                         UK4        True        Pseudomonas fluorescens   \n",
       "15                  NBRC 14164        True             Pseudomonas putida   \n",
       "16                         F2a        True         Stutzerimonas stutzeri   \n",
       "17                         168        True              Bacillus subtilis   \n",
       "18                      Bac111        True          Bacillus vallismortis   \n",
       "19                   EA-CB0015        True           Bacillus tequilensis   \n",
       "20                    UCMB5075        True            Bacillus mojavensis   \n",
       "21                    ZB201702        True          Bacillus halotolerans   \n",
       "22                   ATCC 6633        True            Bacillus spizizenii   \n",
       "23                       JS25R        True            Bacillus velezensis   \n",
       "24                  KCTC 13429        True           Bacillus inaquosorum   \n",
       "25                         T30        True          Bacillus intestinalis   \n",
       "26                   NCTC 8325        True          Staphylococcus aureus   \n",
       "27                     BDS-53E        True      Staphylococcus roterodami   \n",
       "28                       EGD-e        True         Listeria monocytogenes   \n",
       "29                        LD33        True          Enterococcus faecalis   \n",
       "30                     SCB0035        True  Limosilactobacillus fermentum   \n",
       "\n",
       "    Breadth Coverage  Expected Coverage  Coverage Score  Depth Coverage  \\\n",
       "0           0.907640           0.999430        0.908158        8.230061   \n",
       "1           0.743137           0.997109        0.745292        7.866824   \n",
       "2           0.751993           0.998130        0.753402        8.353340   \n",
       "3           0.733252           0.997331        0.735214        8.081780   \n",
       "4           0.730958           0.997523        0.732773        8.209433   \n",
       "5           0.614433           0.993791        0.618272        8.270563   \n",
       "6           0.734095           0.997641        0.735830        8.240902   \n",
       "7           0.478365           0.968910        0.493714        7.255633   \n",
       "8           0.794622           0.998802        0.795576        8.465387   \n",
       "9           0.821040           0.999083        0.821793        8.519167   \n",
       "10          0.733746           0.998104        0.735139        8.542506   \n",
       "11          0.887203           0.999406        0.887730        8.372347   \n",
       "12          0.295883           0.869641        0.340235        6.885986   \n",
       "13          0.945966           0.987809        0.957641        4.658703   \n",
       "14          0.180846           0.474090        0.381459        3.553416   \n",
       "15          0.197449           0.500223        0.394722        3.512750   \n",
       "16          0.241283           0.590584        0.408550        3.701112   \n",
       "17          0.762025           0.882021        0.863954        2.804649   \n",
       "18          0.774163           0.885038        0.874723        2.794134   \n",
       "19          0.741291           0.873268        0.848871        2.786549   \n",
       "20          0.683937           0.832525        0.821522        2.612655   \n",
       "21          0.677839           0.830789        0.815898        2.620952   \n",
       "22          0.905485           0.923127        0.980888        2.833352   \n",
       "23          0.202223           0.383401        0.527445        2.391079   \n",
       "24          0.763588           0.884493        0.863306        2.826641   \n",
       "25          0.908677           0.924061        0.983351        2.836854   \n",
       "26          0.866442           0.950488        0.911576        3.468753   \n",
       "27          0.701952           0.891262        0.787593        3.160850   \n",
       "28          0.817813           0.892041        0.916789        2.721834   \n",
       "29          0.869026           0.989935        0.877861        5.291694   \n",
       "30          0.765433           0.912251        0.839059        3.178842   \n",
       "\n",
       "         BC2       EC2       CS2       DC2  Consensus ANI  \\\n",
       "0   0.912043  0.997296  0.914515  6.483257       0.989812   \n",
       "1   0.022666  0.077416  0.292787  3.554857       0.967281   \n",
       "2   0.372702  0.659305  0.565296  2.889042       0.988459   \n",
       "3   0.012326  0.029806  0.413548  2.454837       0.966724   \n",
       "4   0.019633  0.059327  0.330929  3.115106       0.965896   \n",
       "5   0.000000  0.000000  0.000000  0.000000       0.000000   \n",
       "6   0.038282  0.086245  0.443874  2.355988       0.971706   \n",
       "7   0.000000  0.000000  0.000000  0.000000       0.000000   \n",
       "8   0.387796  0.633569  0.612083  2.588805       0.990129   \n",
       "9   0.459718  0.715478  0.642533  2.734124       0.991279   \n",
       "10  0.260939  0.461075  0.565935  2.369026       0.988213   \n",
       "11  0.608641  0.869903  0.699665  3.350823       0.991290   \n",
       "12  0.000000  0.000000  0.000000  0.000000       0.000000   \n",
       "13  0.951679  0.987619  0.963609  4.614541       0.994470   \n",
       "14  0.000000  0.000000  0.000000  0.000000       0.000000   \n",
       "15  0.000000  0.000000  0.000000  0.000000       0.000000   \n",
       "16  0.000000  0.000000  0.000000  0.000000       0.000000   \n",
       "17  0.002772  0.005465  0.507204  1.976979       0.989899   \n",
       "18  0.002715  0.003215  0.844591  1.185895       0.995716   \n",
       "19  0.001899  0.002372  0.800479  1.250722       0.992905   \n",
       "20  0.002909  0.003533  0.823596  1.216320       0.996587   \n",
       "21  0.002948  0.003706  0.795317  1.259677       0.997138   \n",
       "22  0.709671  0.724661  0.979315  1.817361       0.998095   \n",
       "23  0.000000  0.000000  0.000000  0.000000       0.000000   \n",
       "24  0.003876  0.015586  0.248670  4.053022       0.987319   \n",
       "25  0.706171  0.719776  0.981098  1.801468       0.998096   \n",
       "26  0.866030  0.947037  0.914463  3.392590       0.992470   \n",
       "27  0.031307  0.062633  0.499844  2.065971       0.971264   \n",
       "28  0.825120  0.888395  0.928776  2.657477       0.952034   \n",
       "29  0.877028  0.990105  0.885793  5.262746       0.991390   \n",
       "30  0.799532  0.921203  0.867922  3.177840       0.992856   \n",
       "\n",
       "    Combined CS2 and ANI (Sqrt(ANI)xCS2x100) Presence/Absence  \n",
       "0                                      90.98          Present  \n",
       "1                                      28.80           Absent  \n",
       "2                                      56.20          Present  \n",
       "3                                      40.66           Absent  \n",
       "4                                      32.52           Absent  \n",
       "5                                       0.00           Absent  \n",
       "6                                      43.75          Present  \n",
       "7                                       0.00           Absent  \n",
       "8                                      60.91          Present  \n",
       "9                                      63.97          Present  \n",
       "10                                     56.26          Present  \n",
       "11                                     69.66          Present  \n",
       "12                                      0.00           Absent  \n",
       "13                                     96.09          Present  \n",
       "14                                      0.00           Absent  \n",
       "15                                      0.00           Absent  \n",
       "16                                      0.00           Absent  \n",
       "17                                     50.46          Present  \n",
       "18                                     84.28          Present  \n",
       "19                                     79.76          Present  \n",
       "20                                     82.22          Present  \n",
       "21                                     79.42          Present  \n",
       "22                                     97.84          Present  \n",
       "23                                      0.00           Absent  \n",
       "24                                     24.71          Present  \n",
       "25                                     98.02          Present  \n",
       "26                                     91.10          Present  \n",
       "27                                     49.26          Present  \n",
       "28                                     90.62          Present  \n",
       "29                                     88.20          Present  \n",
       "30                                     86.48          Present  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "downloaded_assemblies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "335180ef-eedf-41f7-848e-336ccfa0be2b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
