{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1c2eecdc-31a9-4f82-b2c1-2538d6380aff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "from collections import defaultdict\n",
    "from Bio import Entrez\n",
    "import time\n",
    "\n",
    "import pandas as pd\n",
    "import json\n",
    "from ete3 import NCBITaxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f79aa254-97a1-4867-a88c-5675a5347495",
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.insert(0, '../utils')\n",
    "from reference_finder import download_reference_genome, unpack, cat_reference_genome\n",
    "from alignment import run_minimap2, sort_samfile, samtools_calculate_depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "08b1fd79-11d4-4d26-87f2-2a09a3447ca7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "22e46ecd-2fa2-487f-883b-06b1f0db5a73",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_species_taxid(taxid, ncbi_taxa_db, valid_kingdom={2, 4751, 2157, 10239}):\n",
    "    lineage = ncbi_taxa_db.get_lineage(taxid)\n",
    "    if bool(set(lineage) & valid_kingdom):\n",
    "        taxid2rank_dict = ncbi_taxa_db.get_rank(lineage)\n",
    "        for lineage_taxid in taxid2rank_dict:\n",
    "            if taxid2rank_dict[lineage_taxid] == 'species':\n",
    "                return lineage_taxid\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aca8de42-c709-497c-b573-9f3b091756a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_seqscreen_taxonomy(seqscreen_output, min_frac, ncbi_taxa_db, valid_kingdom):\n",
    "    classification_result_df = pd.read_csv(os.path.join(seqscreen_output, \n",
    "                                                        'taxonomic_identification', \n",
    "                                                        'taxonomic_assignment', \n",
    "                                                        'taxonomic_results.txt'), \n",
    "                                           sep='\\t')\n",
    "    total_read_count, _ = classification_result_df.shape\n",
    "    \n",
    "    taxid_count_dict = defaultdict(int)\n",
    "    taxid_species_lookup = dict()\n",
    "    error_count = 0\n",
    "    for taxid in classification_result_df['taxid']:\n",
    "        try:\n",
    "            taxid = int(taxid)\n",
    "            try:\n",
    "                species_taxid = taxid_species_lookup[taxid]\n",
    "            except KeyError:\n",
    "                species_taxid = get_species_taxid(taxid, ncbi_taxa_db, valid_kingdom)\n",
    "                taxid_species_lookup[taxid] = species_taxid\n",
    "\n",
    "            if species_taxid is not None:\n",
    "                taxid_count_dict[species_taxid] += 1\n",
    "        except ValueError:\n",
    "            error_count += 1\n",
    "\n",
    "    taxid_queries = []\n",
    "    for taxid in taxid_count_dict:\n",
    "        if taxid_count_dict[taxid] >= min_frac * total_read_count:\n",
    "            taxid_queries.append(taxid)\n",
    "            \n",
    "    return taxid_queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5d2ba1d0-ccca-4f0e-b3fb-f652e8be287a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_reference_genomes(taxid_queries, output_directory, ncbi_taxa_db):\n",
    "    working_dir = os.path.join(output_directory, 'ncbi_downloads')\n",
    "    if not os.path.exists(working_dir):\n",
    "        os.mkdir(working_dir)\n",
    "        \n",
    "    download_result = []\n",
    "    for taxid in taxid_queries:\n",
    "        download_result.append(download_reference_genome(taxid, working_dir))\n",
    "        \n",
    "    unpack(working_dir, output_directory)\n",
    "    reference_metadata = pd.DataFrame(download_result,\n",
    "                                      columns=['Taxonomy ID', \n",
    "                                               'Assembly Accession ID', \n",
    "                                               'Source Database', \n",
    "                                               'Is Representative', \n",
    "                                               'Assembly Level', \n",
    "                                               'Organism of Assembly',\n",
    "                                               'Strain',\n",
    "                                               'Downloaded'])\n",
    "    \n",
    "    taxonomy_name = []\n",
    "    for taxid in reference_metadata['Taxonomy ID']:\n",
    "        taxonomy_name.append(ncbi_taxa_db.get_taxid_translator([taxid])[taxid])\n",
    "    reference_metadata['Species'] = taxonomy_name\n",
    "    \n",
    "    reference_metadata.to_csv(os.path.join(output_directory, 'reference_metadata.csv'), index=False)\n",
    "    cat_reference_genome(reference_metadata, output_directory, reference_genome_path=os.path.join(output_directory, 'reference_genomes'))\n",
    "    \n",
    "    return reference_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c83020db-749d-4676-b2da-bb483e9e28f6",
   "metadata": {},
   "source": [
    "## Load ETE3 NCBITaxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8719997e-dc12-4c48-aa76-87452061ae1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "database_path = '/home/dbs/SeqScreenDB_21.4'\n",
    "\n",
    "ete3db = os.path.join(database_path, \"reference_inference\", \"taxa.sqlite\")\n",
    "sequences_db_f = os.path.join(database_path, \"bowtie2\", \"blacklist.seqs.nt.fna\")\n",
    "mapping_f = os.path.join(database_path, \"reference_inference\", \"taxid2seqid.pickle\")\n",
    "\n",
    "ncbi_taxa_db = NCBITaxa(dbfile=ete3db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e24b0e98-1811-4a70-8eeb-58c1c4c3ddba",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_directory = \"/home/Users/yl181/seqscreen_nano/test\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e25e0405-273a-4e88-a44a-d8f2bde5f7e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_fasta = '/home/Users/yl181/seqscreen_nano/input_datasets/ERR3152364.downsampled.fasta'\n",
    "seqscreen_output = \"/home/Users/yl181/seqscreen_nano/output_datasets/ZymoBIOMICS.STD.Even.ont.seqscreen\"\n",
    "threads = 40\n",
    "min_frac = 0.002"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ad2b11f6-e9f9-464f-bf04-92596715da42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ncbi.update_taxonomy_database()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb8cab7-2569-443a-8c11-f1abd13932da",
   "metadata": {},
   "source": [
    "## Filtering Seqscreen Taxonomic Assignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe727ba4-7f9d-40ff-9aeb-711e6be1c344",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxid_queries = filter_seqscreen_taxonomy(seqscreen_output, \n",
    "                                          min_frac=min_frac, \n",
    "                                          ncbi_taxa_db=ncbi_taxa_db, \n",
    "                                          valid_kingdom={2, 4751, 2157, 10239})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a8c51de-5916-4e66-b620-4b0697aee0a3",
   "metadata": {},
   "source": [
    "## Fetch Reference Genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30379c84-a5fb-4c57-9295-cf96c4e74700",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_metadata = prepare_reference_genomes(taxid_queries, output_directory, ncbi_taxa_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aef231fe-a8dd-4847-9e4a-fdc261326174",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_metadata = pd.read_csv(os.path.join(output_directory, 'reference_metadata.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b15290e5-ba4e-40f7-b126-49cc5f9963e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = reference_metadata[reference_metadata['Downloaded']]\n",
    "for assembly_id in downloaded_assemblies['Assembly Accession ID']:\n",
    "    reference_fasta = os.path.join(output_directory, 'reference_genomes', f'{assembly_id}.fasta')\n",
    "    run_minimap2(input_fasta, reference_fasta, assembly_id, output_directory, threads=threads)\n",
    "    sort_samfile(assembly_id, output_directory, threads)\n",
    "    samtools_calculate_depth(assembly_id, output_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d04504f6-ab98-4268-b337-10b4b8068a28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Assembly Accession ID</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>Is Representative</th>\n",
       "      <th>Assembly Level</th>\n",
       "      <th>Organism of Assembly</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Downloaded</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1280</td>\n",
       "      <td>GCF_000013425.1</td>\n",
       "      <td>University of Oklahoma Health Sciences Center</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>NCTC 8325</td>\n",
       "      <td>True</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1613</td>\n",
       "      <td>GCF_022819245.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Limosilactobacillus fermentum</td>\n",
       "      <td>SCB0035</td>\n",
       "      <td>True</td>\n",
       "      <td>Limosilactobacillus fermentum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1351</td>\n",
       "      <td>GCF_001598635.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Enterococcus faecalis</td>\n",
       "      <td>LD33</td>\n",
       "      <td>True</td>\n",
       "      <td>Enterococcus faecalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1423</td>\n",
       "      <td>GCF_000009045.1</td>\n",
       "      <td>BSNR</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>168</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>287</td>\n",
       "      <td>GCF_000006765.1</td>\n",
       "      <td>PathoGenesis Corporation</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>True</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5207</td>\n",
       "      <td>GCA_022832995.1</td>\n",
       "      <td>Broad Institute</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>VNII</td>\n",
       "      <td>True</td>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1639</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "      <td>European Consortium</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Listeria monocytogenes EGD-e</td>\n",
       "      <td>EGD-e</td>\n",
       "      <td>True</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1642</td>\n",
       "      <td>GCF_009648575.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Listeria innocua</td>\n",
       "      <td>CFSAN044836</td>\n",
       "      <td>True</td>\n",
       "      <td>Listeria innocua</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>562</td>\n",
       "      <td>GCF_000008865.2</td>\n",
       "      <td>GIRC</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia coli O157:H7 str. Sakai</td>\n",
       "      <td>Sakai substr. RIMD 0509952</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1638</td>\n",
       "      <td>GCF_000763515.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Listeria ivanovii subsp. ivanovii</td>\n",
       "      <td>WSLC 3010</td>\n",
       "      <td>True</td>\n",
       "      <td>Listeria ivanovii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>28901</td>\n",
       "      <td>GCF_000006945.2</td>\n",
       "      <td>Washington University Genome Sequencing Center</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>LT2</td>\n",
       "      <td>True</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1392</td>\n",
       "      <td>GCF_000008445.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus anthracis str. 'Ames Ancestor'</td>\n",
       "      <td>Ames Ancestor; A2084</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus anthracis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>623</td>\n",
       "      <td>GCF_000006925.2</td>\n",
       "      <td>Microbial Genome Center of ChMPH</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Shigella flexneri 2a str. 301</td>\n",
       "      <td>301</td>\n",
       "      <td>True</td>\n",
       "      <td>Shigella flexneri</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>573</td>\n",
       "      <td>GCF_000240185.1</td>\n",
       "      <td>Bioinformatics &amp; Microbial Genomics, State Ke...</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae HS11286</td>\n",
       "      <td>HS11286</td>\n",
       "      <td>True</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4932</td>\n",
       "      <td>GCF_000146045.2</td>\n",
       "      <td>Saccharomyces Genome Database</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Saccharomyces cerevisiae S288C</td>\n",
       "      <td>S288C</td>\n",
       "      <td>True</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>96241</td>\n",
       "      <td>GCF_006094475.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bacillus subtilis subsp. spizizenii ATCC 6633 ...</td>\n",
       "      <td>ATCC 6633</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacillus spizizenii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1590</td>\n",
       "      <td>GCF_003269405.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Lactiplantibacillus plantarum</td>\n",
       "      <td>SK151</td>\n",
       "      <td>True</td>\n",
       "      <td>Lactiplantibacillus plantarum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1352</td>\n",
       "      <td>GCF_009734005.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>SRR24</td>\n",
       "      <td>True</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>294</td>\n",
       "      <td>GCF_000730425.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pseudomonas fluorescens</td>\n",
       "      <td>UK4</td>\n",
       "      <td>True</td>\n",
       "      <td>Pseudomonas fluorescens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1643</td>\n",
       "      <td>GCF_900187315.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Listeria welshimeri</td>\n",
       "      <td>NCTC11857</td>\n",
       "      <td>True</td>\n",
       "      <td>Listeria welshimeri</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>176275</td>\n",
       "      <td>GCA_014607475.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Beauveria bassiana</td>\n",
       "      <td>HN6</td>\n",
       "      <td>True</td>\n",
       "      <td>Beauveria bassiana</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Taxonomy ID Assembly Accession ID  \\\n",
       "0          1280       GCF_000013425.1   \n",
       "1          1613       GCF_022819245.1   \n",
       "2          1351       GCF_001598635.1   \n",
       "3          1423       GCF_000009045.1   \n",
       "4           287       GCF_000006765.1   \n",
       "5          5207       GCA_022832995.1   \n",
       "6          1639       GCF_000196035.1   \n",
       "7          1642       GCF_009648575.1   \n",
       "8           562       GCF_000008865.2   \n",
       "9          1638       GCF_000763515.1   \n",
       "10        28901       GCF_000006945.2   \n",
       "11         1392       GCF_000008445.1   \n",
       "12          623       GCF_000006925.2   \n",
       "13          573       GCF_000240185.1   \n",
       "14         4932       GCF_000146045.2   \n",
       "15        96241       GCF_006094475.1   \n",
       "16         1590       GCF_003269405.1   \n",
       "17         1352       GCF_009734005.1   \n",
       "18          294       GCF_000730425.1   \n",
       "19         1643       GCF_900187315.1   \n",
       "20       176275       GCA_014607475.1   \n",
       "\n",
       "                                      Source Database  Is Representative  \\\n",
       "0       University of Oklahoma Health Sciences Center               True   \n",
       "1                                         NCBI RefSeq               True   \n",
       "2                                         NCBI RefSeq              False   \n",
       "3                                                BSNR               True   \n",
       "4                            PathoGenesis Corporation               True   \n",
       "5                                     Broad Institute              False   \n",
       "6                                 European Consortium               True   \n",
       "7                                         NCBI RefSeq               True   \n",
       "8                                                GIRC               True   \n",
       "9                                         NCBI RefSeq               True   \n",
       "10     Washington University Genome Sequencing Center               True   \n",
       "11                                        NCBI RefSeq               True   \n",
       "12                   Microbial Genome Center of ChMPH               True   \n",
       "13   Bioinformatics & Microbial Genomics, State Ke...               True   \n",
       "14                      Saccharomyces Genome Database               True   \n",
       "15                                        NCBI RefSeq               True   \n",
       "16                                        NCBI RefSeq               True   \n",
       "17                                        NCBI RefSeq               True   \n",
       "18                                        NCBI RefSeq              False   \n",
       "19                                        NCBI RefSeq               True   \n",
       "20                                                NaN              False   \n",
       "\n",
       "     Assembly Level                               Organism of Assembly  \\\n",
       "0   Complete Genome      Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "1   Complete Genome                      Limosilactobacillus fermentum   \n",
       "2   Complete Genome                              Enterococcus faecalis   \n",
       "3   Complete Genome         Bacillus subtilis subsp. subtilis str. 168   \n",
       "4   Complete Genome                        Pseudomonas aeruginosa PAO1   \n",
       "5   Complete Genome                            Cryptococcus neoformans   \n",
       "6   Complete Genome                       Listeria monocytogenes EGD-e   \n",
       "7   Complete Genome                                   Listeria innocua   \n",
       "8   Complete Genome                Escherichia coli O157:H7 str. Sakai   \n",
       "9   Complete Genome                  Listeria ivanovii subsp. ivanovii   \n",
       "10  Complete Genome  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "11  Complete Genome            Bacillus anthracis str. 'Ames Ancestor'   \n",
       "12  Complete Genome                      Shigella flexneri 2a str. 301   \n",
       "13  Complete Genome    Klebsiella pneumoniae subsp. pneumoniae HS11286   \n",
       "14  Complete Genome                     Saccharomyces cerevisiae S288C   \n",
       "15  Complete Genome  Bacillus subtilis subsp. spizizenii ATCC 6633 ...   \n",
       "16  Complete Genome                      Lactiplantibacillus plantarum   \n",
       "17  Complete Genome                               Enterococcus faecium   \n",
       "18  Complete Genome                            Pseudomonas fluorescens   \n",
       "19  Complete Genome                                Listeria welshimeri   \n",
       "20  Complete Genome                                 Beauveria bassiana   \n",
       "\n",
       "                        Strain  Downloaded                        Species  \n",
       "0                    NCTC 8325        True          Staphylococcus aureus  \n",
       "1                      SCB0035        True  Limosilactobacillus fermentum  \n",
       "2                         LD33        True          Enterococcus faecalis  \n",
       "3                          168        True              Bacillus subtilis  \n",
       "4                         PAO1        True         Pseudomonas aeruginosa  \n",
       "5                         VNII        True        Cryptococcus neoformans  \n",
       "6                        EGD-e        True         Listeria monocytogenes  \n",
       "7                  CFSAN044836        True               Listeria innocua  \n",
       "8   Sakai substr. RIMD 0509952        True               Escherichia coli  \n",
       "9                    WSLC 3010        True              Listeria ivanovii  \n",
       "10                         LT2        True            Salmonella enterica  \n",
       "11        Ames Ancestor; A2084        True             Bacillus anthracis  \n",
       "12                         301        True              Shigella flexneri  \n",
       "13                     HS11286        True          Klebsiella pneumoniae  \n",
       "14                       S288C        True       Saccharomyces cerevisiae  \n",
       "15                   ATCC 6633        True            Bacillus spizizenii  \n",
       "16                       SK151        True  Lactiplantibacillus plantarum  \n",
       "17                       SRR24        True           Enterococcus faecium  \n",
       "18                         UK4        True        Pseudomonas fluorescens  \n",
       "19                   NCTC11857        True            Listeria welshimeri  \n",
       "20                         HN6        True             Beauveria bassiana  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reference_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10cc7a31-bc25-49ca-b380-7b1d2a63393f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
