{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "251cef2a-37b5-46e6-9fda-4546d1c14315",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import os\n",
    "import argparse\n",
    "import sys\n",
    "import subprocess\n",
    "import math\n",
    "import pickle\n",
    "from collections import defaultdict\n",
    "import json\n",
    "import pandas as pd\n",
    "import math\n",
    "from ete3 import NCBITaxa\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3911e5c3-e6b4-41f6-a06b-0fc2337e0e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.reference_finder import prepare_reference_genomes\n",
    "from utils.alignment import run_minimap2, sort_samfile, run_bwa, samtools_calculate_depth\n",
    "from utils.summary import alignment_1_summary, alignment_2_summary, merge_reference_fasta, call_present_absent\n",
    "from utils.ani import samtools_merged_consensus, ani_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2b2917e3-e287-47cb-a01e-5df068b70a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_species_taxid(taxid, ncbi_taxa_db, valid_kingdom):\n",
    "    lineage = ncbi_taxa_db.get_lineage(taxid)\n",
    "    if bool(set(lineage) & valid_kingdom):\n",
    "        taxid2rank_dict = ncbi_taxa_db.get_rank(lineage)\n",
    "        for lineage_taxid in taxid2rank_dict:\n",
    "            if taxid2rank_dict[lineage_taxid] == 'species':\n",
    "                return lineage_taxid\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f15d60c-3721-4449-a4ba-f523ce4a32b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_kraken2_taxonomy(kraken2_report, min_frac, ncbi_taxa_db, valid_kingdom):\n",
    "    classification_result_df = pd.read_csv(kraken2_report, sep='\\t', names=['Abundance', 'Cumulative Count', 'Count', 'Rank', 'Taxid', 'Taxname']) \n",
    "    taxids = classification_result_df[(classification_result_df['Rank'] == 'S') & (classification_result_df['Abundance'] >= min_frac*100)]['Taxid'].tolist()\n",
    "    valid_taxids = []\n",
    "    for taxid in taxids:\n",
    "        species_taxid = get_species_taxid(taxid, ncbi_taxa_db, valid_kingdom)\n",
    "        if species_taxid is not None:\n",
    "            valid_taxids.append(species_taxid)\n",
    "    return valid_taxids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "63a51027-5f3b-4191-a6a1-2a02927dd7f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pathogen_taxids(pathogen_list_f, ncbi_taxa_db, valid_kingdom):\n",
    "    pathogen_df = pd.read_csv(pathogen_list_f)\n",
    "    pathogen_df = pathogen_df.set_index('PathogenTaxID')\n",
    "    valid_taxids = []\n",
    "    for taxid in pathogen_df.index.to_list():\n",
    "        species_taxid = get_species_taxid(taxid, ncbi_taxa_db, valid_kingdom)\n",
    "        if species_taxid is not None:\n",
    "            valid_taxids.append(species_taxid)\n",
    "    return valid_taxids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "66cb6cb6-0526-4a9d-bb73-0300b4b74fbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def unzip_fastq(input_fastq, working_directory):\n",
    "    temp_input_dir = os.path.join(working_directory, 'input_reads')\n",
    "    if not os.path.exists(temp_input_dir):\n",
    "        os.mkdir(temp_input_dir)\n",
    "        \n",
    "    subprocess.run(['cp',\n",
    "                    input_fastq,\n",
    "                    temp_input_dir],\n",
    "                   check=True)\n",
    "    \n",
    "    filename = input_fastq.split('/')[-1]\n",
    "    subprocess.run(['gunzip', '-d', os.path.join(temp_input_dir, filename)], check=True)\n",
    "    \n",
    "    filename = input_fastq.split('/')[-1][:-3]\n",
    "    return os.path.join(temp_input_dir, filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b5a8b208-4aa2-481c-8e3c-927c6634ff8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_taxa_db = NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6747fe53-a872-4a7a-a56f-9a81745db890",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_fastq_1 = '/data/ww_data/final_ww_data/fastp/wastewater metagenome/SRR11088368_1.fastq.gz'\n",
    "input_fastq_2 = '/data/ww_data/final_ww_data/fastp/wastewater metagenome/SRR11088368_2.fastq.gz'\n",
    "kraken2_report = '/data/ww_data/final_ww_data/kraken/wastewater metagenome/SRR11088368_report.txt'\n",
    "pathogen_list_f = '/home/Users/yl181/ww_pathogen/metadata/filtered_pathogen.csv'\n",
    "working_directory = '/scratch0/yl181/ww_pathogen/test_run'\n",
    "threads = 40\n",
    "min_mapq = 20\n",
    "min_coverage_score = 0.7\n",
    "min_frac = 0.0005"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1f4939e6-c628-4af1-99f8-e76945f76e50",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(working_directory):\n",
    "    os.mkdir(working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cc7259c0-7d2c-47e9-9a40-0eeb769e73d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gzip: /scratch0/yl181/ww_pathogen/test_run/input_reads/SRR11088368_1.fastq already exists;\tnot overwritten\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command '['gunzip', '-d', '/scratch0/yl181/ww_pathogen/test_run/input_reads/SRR11088368_1.fastq.gz']' returned non-zero exit status 2.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [10], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m input_fastq_1 \u001b[38;5;241m=\u001b[39m \u001b[43munzip_fastq\u001b[49m\u001b[43m(\u001b[49m\u001b[43minput_fastq_1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mworking_directory\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn [6], line 12\u001b[0m, in \u001b[0;36munzip_fastq\u001b[0;34m(input_fastq, working_directory)\u001b[0m\n\u001b[1;32m      6\u001b[0m subprocess\u001b[38;5;241m.\u001b[39mrun([\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcp\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      7\u001b[0m                 input_fastq,\n\u001b[1;32m      8\u001b[0m                 temp_input_dir],\n\u001b[1;32m      9\u001b[0m                check\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m     11\u001b[0m filename \u001b[38;5;241m=\u001b[39m input_fastq\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\n\u001b[0;32m---> 12\u001b[0m \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgunzip\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m-d\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtemp_input_dir\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m     14\u001b[0m filename \u001b[38;5;241m=\u001b[39m input_fastq\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m][:\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m3\u001b[39m]\n\u001b[1;32m     15\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(temp_input_dir, filename)\n",
      "File \u001b[0;32m~/miniconda3/envs/ww_pathogen/lib/python3.9/subprocess.py:528\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    526\u001b[0m     retcode \u001b[38;5;241m=\u001b[39m process\u001b[38;5;241m.\u001b[39mpoll()\n\u001b[1;32m    527\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m check \u001b[38;5;129;01mand\u001b[39;00m retcode:\n\u001b[0;32m--> 528\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(retcode, process\u001b[38;5;241m.\u001b[39margs,\n\u001b[1;32m    529\u001b[0m                                  output\u001b[38;5;241m=\u001b[39mstdout, stderr\u001b[38;5;241m=\u001b[39mstderr)\n\u001b[1;32m    530\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m CompletedProcess(process\u001b[38;5;241m.\u001b[39margs, retcode, stdout, stderr)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command '['gunzip', '-d', '/scratch0/yl181/ww_pathogen/test_run/input_reads/SRR11088368_1.fastq.gz']' returned non-zero exit status 2."
     ]
    }
   ],
   "source": [
    "input_fastq_1 = unzip_fastq(input_fastq_1, working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19cf7f02-2c5d-4c5a-98fd-000fc1729efa",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_fastq_2 = unzip_fastq(input_fastq_2, working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9f8c9842-3089-4a3f-8230-de0ba518830a",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxid_queries = filter_kraken2_taxonomy(kraken2_report,\n",
    "                                        min_frac=min_frac, \n",
    "                                        ncbi_taxa_db=ncbi_taxa_db, \n",
    "                                        valid_kingdom={2, 4751, 2157, 10239})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6e1de4ec-7aa2-45c2-adb5-454bbe04b719",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(taxid_queries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a2b29907-910d-424e-83d1-04d39b7a327c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathogen_taxids = get_pathogen_taxids(pathogen_list_f,\n",
    "                                      ncbi_taxa_db=ncbi_taxa_db, \n",
    "                                      valid_kingdom={2, 4751, 2157, 10239})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f0d21859-f743-4e55-a0ef-f5c32f45a607",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxid_queries = list(set(taxid_queries).intersection(set(pathogen_taxids)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4a8a6776-6ed4-4547-93b6-1856effe31d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(taxid_queries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f5bf1e11-2199-42ac-9550-d3af818b2d66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40324      \t GCF_900475405.1 \t Stenotrophomonas maltophilia   \t NCTC1025.. \t Complete Genome\n",
      "645        \t GCF_012931585.1 \t Aeromonas salmonicida          \t SRW-OG1    \t Complete Genome\n",
      "651        \t GCF_020423125.1 \t Aeromonas media                \t TR3_1      \t Complete Genome\n",
      "33039      \t GCA_023674525.1 \t [Ruminococcus] torques         \t NB2A-14-.. \t Chromosome\n",
      "1680       \t GCF_003030905.1 \t Bifidobacterium adolescentis   \t 1-11       \t Complete Genome\n",
      "40214      \t GCF_003335165.1 \t Acinetobacter johnsonii        \t LXL_C1     \t Complete Genome\n",
      "40215      \t GCF_018336855.1 \t Acinetobacter junii            \t YR7        \t Complete Genome\n",
      "2479767    \t GCF_005848555.1 \t Blautia sp. SC05B48            \t SC05B48    \t Complete Genome\n",
      "2743575    \t GCF_013374795.1 \t Acinetobacter sp. NEB 394      \t NEB 394    \t Complete Genome\n",
      "1307       \t GCF_000026745.1 \t Streptococcus suis BM407       \t BM407      \t Complete Genome\n",
      "649756     \t GCF_001998765.1 \t Anaerostipes hadrus            \t BPB5       \t Complete Genome\n",
      "285        \t GCF_014076415.1 \t Comamonas testosteroni         \t T5-67      \t Complete Genome\n",
      "287        \t GCF_000006765.1 \t Pseudomonas aeruginosa PAO1    \t PAO1       \t Complete Genome\n",
      "546        \t GCF_003812345.1 \t Citrobacter freundii           \t FDAARGOS.. \t Complete Genome\n",
      "28198      \t GCF_003660105.1 \t Aliarcobacter cryaerophilus .. \t ATCC 431.. \t Complete Genome\n",
      "562        \t GCF_000008865.2 \t Escherichia coli O157:H7 str.. \t Sakai su.. \t Complete Genome\n",
      "74426      \t GCF_002736145.1 \t Collinsella aerofaciens        \t indica     \t Complete Genome\n",
      "165179     \t GCF_020735445.1 \t Prevotella copri DSM 18205     \t FDAARGOS.. \t Complete Genome\n",
      "573        \t GCF_000240185.1 \t Klebsiella pneumoniae subsp... \t HS11286    \t Complete Genome\n",
      "239935     \t GCF_009731575.1 \t Akkermansia muciniphila        \t JCM 3089.. \t Complete Genome\n",
      "225992     \t GCF_002056725.1 \t Comamonas kerstersii           \t 8943       \t Complete Genome\n",
      "237258     \t GCF_003860565.1 \t Cloacibacterium normanense     \t NRS-1      \t Complete Genome\n",
      "853        \t GCF_002586945.1 \t Faecalibacterium prausnitzii   \t Indica     \t Complete Genome\n",
      "470        \t GCF_008632635.1 \t Acinetobacter baumannii        \t K09-14     \t Complete Genome\n",
      "166486     \t GCF_900537995.1 \t Roseburia intestinalis L1-82   \t L1-82      \t Complete Genome\n",
      "410072     \t GCF_016904155.1 \t Coprococcus comes              \t FDAARGOS.. \t Complete Genome\n",
      "876        \t GCF_017815575.1 \t Desulfovibrio desulfuricans    \t L4         \t Complete Genome\n",
      "216816     \t GCF_000196555.1 \t Bifidobacterium longum subsp.. \t JCM 1217   \t Complete Genome\n",
      "158836     \t GCF_019048625.1 \t Enterobacter hormaechei        \t FDAARGOS.. \t Complete Genome\n",
      "1463165    \t GCF_016415705.1 \t Klebsiella quasipneumoniae     \t KqPF26     \t Complete Genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "30 archives were successfully processed.\n"
     ]
    }
   ],
   "source": [
    "reference_metadata = prepare_reference_genomes(taxid_queries, working_directory, ncbi_taxa_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5c51a0c3-59eb-4ad6-93ba-5496fcca90b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Assembly Accession ID</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>Is Representative</th>\n",
       "      <th>Assembly Level</th>\n",
       "      <th>Organism of Assembly</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Total Length</th>\n",
       "      <th>Downloaded</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>40324</td>\n",
       "      <td>GCF_900475405.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Stenotrophomonas maltophilia</td>\n",
       "      <td>NCTC10258</td>\n",
       "      <td>4481118</td>\n",
       "      <td>True</td>\n",
       "      <td>Stenotrophomonas maltophilia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>645</td>\n",
       "      <td>GCF_012931585.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Aeromonas salmonicida</td>\n",
       "      <td>SRW-OG1</td>\n",
       "      <td>4621725</td>\n",
       "      <td>True</td>\n",
       "      <td>Aeromonas salmonicida</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>651</td>\n",
       "      <td>GCF_020423125.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Aeromonas media</td>\n",
       "      <td>TR3_1</td>\n",
       "      <td>4531033</td>\n",
       "      <td>True</td>\n",
       "      <td>Aeromonas media</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>33039</td>\n",
       "      <td>GCA_023674525.1</td>\n",
       "      <td>SOURCE_DATABASE_GENBANK</td>\n",
       "      <td>False</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>[Ruminococcus] torques</td>\n",
       "      <td>NB2A-14-FMU</td>\n",
       "      <td>2838928</td>\n",
       "      <td>True</td>\n",
       "      <td>[Ruminococcus] torques</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1680</td>\n",
       "      <td>GCF_003030905.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bifidobacterium adolescentis</td>\n",
       "      <td>1-11</td>\n",
       "      <td>2192428</td>\n",
       "      <td>True</td>\n",
       "      <td>Bifidobacterium adolescentis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>40214</td>\n",
       "      <td>GCF_003335165.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Acinetobacter johnsonii</td>\n",
       "      <td>LXL_C1</td>\n",
       "      <td>3398706</td>\n",
       "      <td>True</td>\n",
       "      <td>Acinetobacter johnsonii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>40215</td>\n",
       "      <td>GCF_018336855.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Acinetobacter junii</td>\n",
       "      <td>YR7</td>\n",
       "      <td>3438557</td>\n",
       "      <td>True</td>\n",
       "      <td>Acinetobacter junii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2479767</td>\n",
       "      <td>GCF_005848555.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Blautia sp. SC05B48</td>\n",
       "      <td>SC05B48</td>\n",
       "      <td>3731576</td>\n",
       "      <td>True</td>\n",
       "      <td>Blautia sp. SC05B48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2743575</td>\n",
       "      <td>GCF_013374795.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Acinetobacter sp. NEB 394</td>\n",
       "      <td>NEB 394</td>\n",
       "      <td>3842079</td>\n",
       "      <td>True</td>\n",
       "      <td>Acinetobacter sp. NEB 394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1307</td>\n",
       "      <td>GCF_000026745.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Streptococcus suis BM407</td>\n",
       "      <td>BM407</td>\n",
       "      <td>2170808</td>\n",
       "      <td>True</td>\n",
       "      <td>Streptococcus suis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>649756</td>\n",
       "      <td>GCF_001998765.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Anaerostipes hadrus</td>\n",
       "      <td>BPB5</td>\n",
       "      <td>3172613</td>\n",
       "      <td>True</td>\n",
       "      <td>Anaerostipes hadrus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>285</td>\n",
       "      <td>GCF_014076415.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Comamonas testosteroni</td>\n",
       "      <td>T5-67</td>\n",
       "      <td>5947116</td>\n",
       "      <td>True</td>\n",
       "      <td>Comamonas testosteroni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>287</td>\n",
       "      <td>GCF_000006765.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>6264404</td>\n",
       "      <td>True</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>546</td>\n",
       "      <td>GCF_003812345.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Citrobacter freundii</td>\n",
       "      <td>FDAARGOS_549</td>\n",
       "      <td>5102161</td>\n",
       "      <td>True</td>\n",
       "      <td>Citrobacter freundii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>28198</td>\n",
       "      <td>GCF_003660105.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Aliarcobacter cryaerophilus ATCC 43158</td>\n",
       "      <td>ATCC 43158</td>\n",
       "      <td>2108340</td>\n",
       "      <td>True</td>\n",
       "      <td>Aliarcobacter cryaerophilus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>562</td>\n",
       "      <td>GCF_000008865.2</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia coli O157:H7 str. Sakai</td>\n",
       "      <td>Sakai substr. RIMD 0509952</td>\n",
       "      <td>5594605</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>74426</td>\n",
       "      <td>GCF_002736145.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Collinsella aerofaciens</td>\n",
       "      <td>indica</td>\n",
       "      <td>2306349</td>\n",
       "      <td>True</td>\n",
       "      <td>Collinsella aerofaciens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>165179</td>\n",
       "      <td>GCF_020735445.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Prevotella copri DSM 18205</td>\n",
       "      <td>FDAARGOS_1573</td>\n",
       "      <td>3551614</td>\n",
       "      <td>True</td>\n",
       "      <td>Prevotella copri</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>573</td>\n",
       "      <td>GCF_000240185.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae HS11286</td>\n",
       "      <td>HS11286</td>\n",
       "      <td>5682322</td>\n",
       "      <td>True</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>239935</td>\n",
       "      <td>GCF_009731575.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Akkermansia muciniphila</td>\n",
       "      <td>JCM 30893</td>\n",
       "      <td>2878459</td>\n",
       "      <td>True</td>\n",
       "      <td>Akkermansia muciniphila</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>225992</td>\n",
       "      <td>GCF_002056725.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Comamonas kerstersii</td>\n",
       "      <td>8943</td>\n",
       "      <td>3547915</td>\n",
       "      <td>True</td>\n",
       "      <td>Comamonas kerstersii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>237258</td>\n",
       "      <td>GCF_003860565.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Cloacibacterium normanense</td>\n",
       "      <td>NRS-1</td>\n",
       "      <td>2759562</td>\n",
       "      <td>True</td>\n",
       "      <td>Cloacibacterium normanense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>853</td>\n",
       "      <td>GCF_002586945.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Faecalibacterium prausnitzii</td>\n",
       "      <td>Indica</td>\n",
       "      <td>2868932</td>\n",
       "      <td>True</td>\n",
       "      <td>Faecalibacterium prausnitzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>470</td>\n",
       "      <td>GCF_008632635.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>K09-14</td>\n",
       "      <td>3980230</td>\n",
       "      <td>True</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>166486</td>\n",
       "      <td>GCF_900537995.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Roseburia intestinalis L1-82</td>\n",
       "      <td>L1-82</td>\n",
       "      <td>4493348</td>\n",
       "      <td>True</td>\n",
       "      <td>Roseburia intestinalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>410072</td>\n",
       "      <td>GCF_016904155.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Coprococcus comes</td>\n",
       "      <td>FDAARGOS_1339</td>\n",
       "      <td>3373067</td>\n",
       "      <td>True</td>\n",
       "      <td>Coprococcus comes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>876</td>\n",
       "      <td>GCF_017815575.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Desulfovibrio desulfuricans</td>\n",
       "      <td>L4</td>\n",
       "      <td>3554892</td>\n",
       "      <td>True</td>\n",
       "      <td>Desulfovibrio desulfuricans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>216816</td>\n",
       "      <td>GCF_000196555.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bifidobacterium longum subsp. longum JCM 1217</td>\n",
       "      <td>JCM 1217</td>\n",
       "      <td>2385164</td>\n",
       "      <td>True</td>\n",
       "      <td>Bifidobacterium longum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>158836</td>\n",
       "      <td>GCF_019048625.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Enterobacter hormaechei</td>\n",
       "      <td>FDAARGOS 1435</td>\n",
       "      <td>4762440</td>\n",
       "      <td>True</td>\n",
       "      <td>Enterobacter hormaechei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1463165</td>\n",
       "      <td>GCF_016415705.1</td>\n",
       "      <td>SOURCE_DATABASE_REFSEQ</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Klebsiella quasipneumoniae</td>\n",
       "      <td>KqPF26</td>\n",
       "      <td>5391123</td>\n",
       "      <td>True</td>\n",
       "      <td>Klebsiella quasipneumoniae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Taxonomy ID Assembly Accession ID          Source Database  \\\n",
       "0         40324       GCF_900475405.1   SOURCE_DATABASE_REFSEQ   \n",
       "1           645       GCF_012931585.1   SOURCE_DATABASE_REFSEQ   \n",
       "2           651       GCF_020423125.1   SOURCE_DATABASE_REFSEQ   \n",
       "3         33039       GCA_023674525.1  SOURCE_DATABASE_GENBANK   \n",
       "4          1680       GCF_003030905.1   SOURCE_DATABASE_REFSEQ   \n",
       "5         40214       GCF_003335165.1   SOURCE_DATABASE_REFSEQ   \n",
       "6         40215       GCF_018336855.1   SOURCE_DATABASE_REFSEQ   \n",
       "7       2479767       GCF_005848555.1   SOURCE_DATABASE_REFSEQ   \n",
       "8       2743575       GCF_013374795.1   SOURCE_DATABASE_REFSEQ   \n",
       "9          1307       GCF_000026745.1   SOURCE_DATABASE_REFSEQ   \n",
       "10       649756       GCF_001998765.1   SOURCE_DATABASE_REFSEQ   \n",
       "11          285       GCF_014076415.1   SOURCE_DATABASE_REFSEQ   \n",
       "12          287       GCF_000006765.1   SOURCE_DATABASE_REFSEQ   \n",
       "13          546       GCF_003812345.1   SOURCE_DATABASE_REFSEQ   \n",
       "14        28198       GCF_003660105.1   SOURCE_DATABASE_REFSEQ   \n",
       "15          562       GCF_000008865.2   SOURCE_DATABASE_REFSEQ   \n",
       "16        74426       GCF_002736145.1   SOURCE_DATABASE_REFSEQ   \n",
       "17       165179       GCF_020735445.1   SOURCE_DATABASE_REFSEQ   \n",
       "18          573       GCF_000240185.1   SOURCE_DATABASE_REFSEQ   \n",
       "19       239935       GCF_009731575.1   SOURCE_DATABASE_REFSEQ   \n",
       "20       225992       GCF_002056725.1   SOURCE_DATABASE_REFSEQ   \n",
       "21       237258       GCF_003860565.1   SOURCE_DATABASE_REFSEQ   \n",
       "22          853       GCF_002586945.1   SOURCE_DATABASE_REFSEQ   \n",
       "23          470       GCF_008632635.1   SOURCE_DATABASE_REFSEQ   \n",
       "24       166486       GCF_900537995.1   SOURCE_DATABASE_REFSEQ   \n",
       "25       410072       GCF_016904155.1   SOURCE_DATABASE_REFSEQ   \n",
       "26          876       GCF_017815575.1   SOURCE_DATABASE_REFSEQ   \n",
       "27       216816       GCF_000196555.1   SOURCE_DATABASE_REFSEQ   \n",
       "28       158836       GCF_019048625.1   SOURCE_DATABASE_REFSEQ   \n",
       "29      1463165       GCF_016415705.1   SOURCE_DATABASE_REFSEQ   \n",
       "\n",
       "    Is Representative   Assembly Level  \\\n",
       "0                True  Complete Genome   \n",
       "1                True  Complete Genome   \n",
       "2                True  Complete Genome   \n",
       "3               False       Chromosome   \n",
       "4                True  Complete Genome   \n",
       "5               False  Complete Genome   \n",
       "6                True  Complete Genome   \n",
       "7               False  Complete Genome   \n",
       "8               False  Complete Genome   \n",
       "9                True  Complete Genome   \n",
       "10              False  Complete Genome   \n",
       "11              False  Complete Genome   \n",
       "12               True  Complete Genome   \n",
       "13               True  Complete Genome   \n",
       "14               True  Complete Genome   \n",
       "15               True  Complete Genome   \n",
       "16               True  Complete Genome   \n",
       "17               True  Complete Genome   \n",
       "18               True  Complete Genome   \n",
       "19               True  Complete Genome   \n",
       "20               True  Complete Genome   \n",
       "21               True  Complete Genome   \n",
       "22              False  Complete Genome   \n",
       "23               True  Complete Genome   \n",
       "24               True  Complete Genome   \n",
       "25               True  Complete Genome   \n",
       "26               True  Complete Genome   \n",
       "27               True  Complete Genome   \n",
       "28               True  Complete Genome   \n",
       "29               True  Complete Genome   \n",
       "\n",
       "                               Organism of Assembly  \\\n",
       "0                      Stenotrophomonas maltophilia   \n",
       "1                             Aeromonas salmonicida   \n",
       "2                                   Aeromonas media   \n",
       "3                            [Ruminococcus] torques   \n",
       "4                      Bifidobacterium adolescentis   \n",
       "5                           Acinetobacter johnsonii   \n",
       "6                               Acinetobacter junii   \n",
       "7                               Blautia sp. SC05B48   \n",
       "8                         Acinetobacter sp. NEB 394   \n",
       "9                          Streptococcus suis BM407   \n",
       "10                              Anaerostipes hadrus   \n",
       "11                           Comamonas testosteroni   \n",
       "12                      Pseudomonas aeruginosa PAO1   \n",
       "13                             Citrobacter freundii   \n",
       "14           Aliarcobacter cryaerophilus ATCC 43158   \n",
       "15              Escherichia coli O157:H7 str. Sakai   \n",
       "16                          Collinsella aerofaciens   \n",
       "17                       Prevotella copri DSM 18205   \n",
       "18  Klebsiella pneumoniae subsp. pneumoniae HS11286   \n",
       "19                          Akkermansia muciniphila   \n",
       "20                             Comamonas kerstersii   \n",
       "21                       Cloacibacterium normanense   \n",
       "22                     Faecalibacterium prausnitzii   \n",
       "23                          Acinetobacter baumannii   \n",
       "24                     Roseburia intestinalis L1-82   \n",
       "25                                Coprococcus comes   \n",
       "26                      Desulfovibrio desulfuricans   \n",
       "27    Bifidobacterium longum subsp. longum JCM 1217   \n",
       "28                          Enterobacter hormaechei   \n",
       "29                       Klebsiella quasipneumoniae   \n",
       "\n",
       "                        Strain  Total Length  Downloaded  \\\n",
       "0                    NCTC10258       4481118        True   \n",
       "1                      SRW-OG1       4621725        True   \n",
       "2                        TR3_1       4531033        True   \n",
       "3                  NB2A-14-FMU       2838928        True   \n",
       "4                         1-11       2192428        True   \n",
       "5                       LXL_C1       3398706        True   \n",
       "6                          YR7       3438557        True   \n",
       "7                      SC05B48       3731576        True   \n",
       "8                      NEB 394       3842079        True   \n",
       "9                        BM407       2170808        True   \n",
       "10                        BPB5       3172613        True   \n",
       "11                       T5-67       5947116        True   \n",
       "12                        PAO1       6264404        True   \n",
       "13                FDAARGOS_549       5102161        True   \n",
       "14                  ATCC 43158       2108340        True   \n",
       "15  Sakai substr. RIMD 0509952       5594605        True   \n",
       "16                      indica       2306349        True   \n",
       "17               FDAARGOS_1573       3551614        True   \n",
       "18                     HS11286       5682322        True   \n",
       "19                   JCM 30893       2878459        True   \n",
       "20                        8943       3547915        True   \n",
       "21                       NRS-1       2759562        True   \n",
       "22                      Indica       2868932        True   \n",
       "23                      K09-14       3980230        True   \n",
       "24                       L1-82       4493348        True   \n",
       "25               FDAARGOS_1339       3373067        True   \n",
       "26                          L4       3554892        True   \n",
       "27                    JCM 1217       2385164        True   \n",
       "28               FDAARGOS 1435       4762440        True   \n",
       "29                      KqPF26       5391123        True   \n",
       "\n",
       "                         Species  \n",
       "0   Stenotrophomonas maltophilia  \n",
       "1          Aeromonas salmonicida  \n",
       "2                Aeromonas media  \n",
       "3         [Ruminococcus] torques  \n",
       "4   Bifidobacterium adolescentis  \n",
       "5        Acinetobacter johnsonii  \n",
       "6            Acinetobacter junii  \n",
       "7            Blautia sp. SC05B48  \n",
       "8      Acinetobacter sp. NEB 394  \n",
       "9             Streptococcus suis  \n",
       "10           Anaerostipes hadrus  \n",
       "11        Comamonas testosteroni  \n",
       "12        Pseudomonas aeruginosa  \n",
       "13          Citrobacter freundii  \n",
       "14   Aliarcobacter cryaerophilus  \n",
       "15              Escherichia coli  \n",
       "16       Collinsella aerofaciens  \n",
       "17              Prevotella copri  \n",
       "18         Klebsiella pneumoniae  \n",
       "19       Akkermansia muciniphila  \n",
       "20          Comamonas kerstersii  \n",
       "21    Cloacibacterium normanense  \n",
       "22  Faecalibacterium prausnitzii  \n",
       "23       Acinetobacter baumannii  \n",
       "24        Roseburia intestinalis  \n",
       "25             Coprococcus comes  \n",
       "26   Desulfovibrio desulfuricans  \n",
       "27        Bifidobacterium longum  \n",
       "28       Enterobacter hormaechei  \n",
       "29    Klebsiella quasipneumoniae  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reference_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "3b6276e5-36bd-4cfd-a37a-907f3c66c8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = reference_metadata[reference_metadata['Downloaded']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "7aacc8f4-314a-41a0-8b1d-c16f3fd9e1de",
   "metadata": {},
   "outputs": [],
   "source": [
    "for assembly_id in downloaded_assemblies['Assembly Accession ID']:\n",
    "    reference_fasta = os.path.join(working_directory, 'reference_genomes', f'{assembly_id}.fasta')\n",
    "    run_bwa(input_fastq_1, input_fastq_2, reference_fasta, assembly_id, working_directory, threads=threads)\n",
    "    sort_samfile(assembly_id, working_directory, min_mapq=min_mapq, threads=threads)\n",
    "    samtools_calculate_depth(assembly_id, working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "e4316f7e-c07a-4078-9511-3faf482968d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = alignment_1_summary(downloaded_assemblies, working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f6cf7b01-2f5c-4975-8725-3a9f5717a40e",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_assemblies = list(downloaded_assemblies[downloaded_assemblies['Coverage Score'] >= min_coverage_score]['Assembly Accession ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "5b7a62e5-2ffc-495e-85d4-852339ed1256",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_fasta = merge_reference_fasta(filtered_assemblies, working_directory)\n",
    "run_bwa(input_fastq_1, input_fastq_2, reference_fasta, 'merged', working_directory, threads=threads)\n",
    "sort_samfile('merged', working_directory, min_mapq=0, threads=threads)\n",
    "samtools_calculate_depth('merged', working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "766ba8bc-2cc6-4d24-866b-9e6c2b30128e",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = alignment_2_summary(downloaded_assemblies, working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "9736777b-5376-4ba9-b750-c8f628ad2ef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_record_dict = samtools_merged_consensus(working_directory, threads)\n",
    "downloaded_assemblies = ani_summary(downloaded_assemblies, consensus_record_dict, working_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "052662c7-a05d-49f7-a88e-e54eb58910e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies = call_present_absent(downloaded_assemblies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "fb5825f7-1eaa-45b2-a984-fc47188e5fdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_assemblies.sort_values(['CS2'], ascending=False).to_csv(os.path.join(working_directory, 'reference_alignment.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "d9addeba-3d0b-415e-90e0-8349530e74aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Assembly Accession ID</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>Is Representative</th>\n",
       "      <th>Assembly Level</th>\n",
       "      <th>Organism of Assembly</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Downloaded</th>\n",
       "      <th>Species</th>\n",
       "      <th>Breadth Coverage</th>\n",
       "      <th>Expected Coverage</th>\n",
       "      <th>Coverage Score</th>\n",
       "      <th>Depth Coverage</th>\n",
       "      <th>BC2</th>\n",
       "      <th>EC2</th>\n",
       "      <th>CS2</th>\n",
       "      <th>DC2</th>\n",
       "      <th>Consensus ANI</th>\n",
       "      <th>Combined CS2 and ANI (Sqrt(ANI)xCS2x100)</th>\n",
       "      <th>Presence/Absence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>40324</td>\n",
       "      <td>GCF_900475405.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Stenotrophomonas maltophilia</td>\n",
       "      <td>NCTC10258</td>\n",
       "      <td>True</td>\n",
       "      <td>Stenotrophomonas maltophilia</td>\n",
       "      <td>0.839573</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.839573</td>\n",
       "      <td>34.563671</td>\n",
       "      <td>0.814941</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.814941</td>\n",
       "      <td>23.577827</td>\n",
       "      <td>0.981539</td>\n",
       "      <td>80.74</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>645</td>\n",
       "      <td>GCF_012931585.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Aeromonas salmonicida</td>\n",
       "      <td>SRW-OG1</td>\n",
       "      <td>True</td>\n",
       "      <td>Aeromonas salmonicida</td>\n",
       "      <td>0.841355</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>0.841356</td>\n",
       "      <td>15.705711</td>\n",
       "      <td>0.718890</td>\n",
       "      <td>0.986014</td>\n",
       "      <td>0.729087</td>\n",
       "      <td>5.939255</td>\n",
       "      <td>0.987220</td>\n",
       "      <td>72.44</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>651</td>\n",
       "      <td>GCF_020423125.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Aeromonas media</td>\n",
       "      <td>TR3_1</td>\n",
       "      <td>True</td>\n",
       "      <td>Aeromonas media</td>\n",
       "      <td>0.871588</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.871588</td>\n",
       "      <td>18.324055</td>\n",
       "      <td>0.842731</td>\n",
       "      <td>0.998330</td>\n",
       "      <td>0.844141</td>\n",
       "      <td>7.588203</td>\n",
       "      <td>0.983200</td>\n",
       "      <td>83.70</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>33039</td>\n",
       "      <td>GCA_023674525.1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>False</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>[Ruminococcus] torques</td>\n",
       "      <td>NB2A-14-FMU</td>\n",
       "      <td>True</td>\n",
       "      <td>[Ruminococcus] torques</td>\n",
       "      <td>0.638248</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>0.638250</td>\n",
       "      <td>20.186668</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1680</td>\n",
       "      <td>GCF_003030905.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bifidobacterium adolescentis</td>\n",
       "      <td>1-11</td>\n",
       "      <td>True</td>\n",
       "      <td>Bifidobacterium adolescentis</td>\n",
       "      <td>0.969784</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.969784</td>\n",
       "      <td>49.626209</td>\n",
       "      <td>0.970498</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.970498</td>\n",
       "      <td>39.889433</td>\n",
       "      <td>0.995705</td>\n",
       "      <td>96.84</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>40214</td>\n",
       "      <td>GCF_003335165.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Acinetobacter johnsonii</td>\n",
       "      <td>LXL_C1</td>\n",
       "      <td>True</td>\n",
       "      <td>Acinetobacter johnsonii</td>\n",
       "      <td>0.955982</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.955982</td>\n",
       "      <td>79.342853</td>\n",
       "      <td>0.974916</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.974916</td>\n",
       "      <td>33.517974</td>\n",
       "      <td>0.989357</td>\n",
       "      <td>96.97</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>40215</td>\n",
       "      <td>GCF_018336855.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Acinetobacter junii</td>\n",
       "      <td>YR7</td>\n",
       "      <td>True</td>\n",
       "      <td>Acinetobacter junii</td>\n",
       "      <td>0.703081</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.703081</td>\n",
       "      <td>32.917163</td>\n",
       "      <td>0.671086</td>\n",
       "      <td>0.999798</td>\n",
       "      <td>0.671221</td>\n",
       "      <td>12.675992</td>\n",
       "      <td>0.955793</td>\n",
       "      <td>65.62</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2479767</td>\n",
       "      <td>GCF_005848555.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Blautia sp. SC05B48</td>\n",
       "      <td>SC05B48</td>\n",
       "      <td>True</td>\n",
       "      <td>Blautia sp. SC05B48</td>\n",
       "      <td>0.939978</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.939978</td>\n",
       "      <td>28.135163</td>\n",
       "      <td>0.954549</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.954549</td>\n",
       "      <td>23.922561</td>\n",
       "      <td>0.983677</td>\n",
       "      <td>94.67</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2743575</td>\n",
       "      <td>GCF_013374795.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Acinetobacter sp. NEB 394</td>\n",
       "      <td>NEB 394</td>\n",
       "      <td>True</td>\n",
       "      <td>Acinetobacter sp. NEB 394</td>\n",
       "      <td>0.959521</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.959521</td>\n",
       "      <td>79.441439</td>\n",
       "      <td>0.990409</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.990409</td>\n",
       "      <td>56.647163</td>\n",
       "      <td>0.995534</td>\n",
       "      <td>98.82</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1307</td>\n",
       "      <td>GCF_000026745.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Streptococcus suis BM407</td>\n",
       "      <td>BM407</td>\n",
       "      <td>True</td>\n",
       "      <td>Streptococcus suis</td>\n",
       "      <td>0.641112</td>\n",
       "      <td>0.999732</td>\n",
       "      <td>0.641284</td>\n",
       "      <td>12.830374</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>649756</td>\n",
       "      <td>GCF_001998765.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Anaerostipes hadrus</td>\n",
       "      <td>BPB5</td>\n",
       "      <td>True</td>\n",
       "      <td>Anaerostipes hadrus</td>\n",
       "      <td>0.955947</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.955947</td>\n",
       "      <td>23.247111</td>\n",
       "      <td>0.973811</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.973811</td>\n",
       "      <td>19.278097</td>\n",
       "      <td>0.995118</td>\n",
       "      <td>97.14</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>285</td>\n",
       "      <td>GCF_014076415.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Comamonas testosteroni</td>\n",
       "      <td>T5-67</td>\n",
       "      <td>True</td>\n",
       "      <td>Comamonas testosteroni</td>\n",
       "      <td>0.729633</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.729633</td>\n",
       "      <td>53.984637</td>\n",
       "      <td>0.700041</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.700041</td>\n",
       "      <td>40.119623</td>\n",
       "      <td>0.968364</td>\n",
       "      <td>68.89</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>287</td>\n",
       "      <td>GCF_000006765.1</td>\n",
       "      <td>PathoGenesis Corporation</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>True</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>0.750096</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.750096</td>\n",
       "      <td>23.007642</td>\n",
       "      <td>0.704004</td>\n",
       "      <td>0.999804</td>\n",
       "      <td>0.704142</td>\n",
       "      <td>12.128048</td>\n",
       "      <td>0.989756</td>\n",
       "      <td>70.05</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>546</td>\n",
       "      <td>GCF_003812345.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Citrobacter freundii</td>\n",
       "      <td>FDAARGOS_549</td>\n",
       "      <td>True</td>\n",
       "      <td>Citrobacter freundii</td>\n",
       "      <td>0.916487</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.916487</td>\n",
       "      <td>26.468455</td>\n",
       "      <td>0.892323</td>\n",
       "      <td>0.998064</td>\n",
       "      <td>0.894054</td>\n",
       "      <td>7.000806</td>\n",
       "      <td>0.985482</td>\n",
       "      <td>88.75</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>28198</td>\n",
       "      <td>GCF_003660105.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Aliarcobacter cryaerophilus ATCC 43158</td>\n",
       "      <td>ATCC 43158</td>\n",
       "      <td>True</td>\n",
       "      <td>Aliarcobacter cryaerophilus</td>\n",
       "      <td>0.932623</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.932623</td>\n",
       "      <td>20.159384</td>\n",
       "      <td>0.945659</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.945659</td>\n",
       "      <td>18.186905</td>\n",
       "      <td>0.977474</td>\n",
       "      <td>93.49</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>562</td>\n",
       "      <td>GCF_000008865.2</td>\n",
       "      <td>GIRC</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Escherichia coli O157:H7 str. Sakai</td>\n",
       "      <td>Sakai substr. RIMD 0509952</td>\n",
       "      <td>True</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>0.896436</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.896436</td>\n",
       "      <td>42.209379</td>\n",
       "      <td>0.916219</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.916219</td>\n",
       "      <td>33.477121</td>\n",
       "      <td>0.990711</td>\n",
       "      <td>91.20</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>74426</td>\n",
       "      <td>GCF_002736145.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Collinsella aerofaciens</td>\n",
       "      <td>indica</td>\n",
       "      <td>True</td>\n",
       "      <td>Collinsella aerofaciens</td>\n",
       "      <td>0.944198</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.944198</td>\n",
       "      <td>32.480432</td>\n",
       "      <td>0.954109</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.954109</td>\n",
       "      <td>26.068877</td>\n",
       "      <td>0.980516</td>\n",
       "      <td>94.48</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>165179</td>\n",
       "      <td>GCF_020735445.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Prevotella copri DSM 18205</td>\n",
       "      <td>FDAARGOS_1573</td>\n",
       "      <td>True</td>\n",
       "      <td>Prevotella copri</td>\n",
       "      <td>0.741253</td>\n",
       "      <td>0.995963</td>\n",
       "      <td>0.744258</td>\n",
       "      <td>7.436305</td>\n",
       "      <td>0.746695</td>\n",
       "      <td>0.986719</td>\n",
       "      <td>0.756745</td>\n",
       "      <td>5.787401</td>\n",
       "      <td>0.986355</td>\n",
       "      <td>75.16</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>573</td>\n",
       "      <td>GCF_000240185.1</td>\n",
       "      <td>Bioinformatics &amp; Microbial Genomics, State Ke...</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae HS11286</td>\n",
       "      <td>HS11286</td>\n",
       "      <td>True</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>0.951345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.951345</td>\n",
       "      <td>30.103833</td>\n",
       "      <td>0.958143</td>\n",
       "      <td>0.999979</td>\n",
       "      <td>0.958164</td>\n",
       "      <td>11.217542</td>\n",
       "      <td>0.995492</td>\n",
       "      <td>95.60</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>239935</td>\n",
       "      <td>GCF_009731575.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Akkermansia muciniphila</td>\n",
       "      <td>JCM 30893</td>\n",
       "      <td>True</td>\n",
       "      <td>Akkermansia muciniphila</td>\n",
       "      <td>0.983008</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.983008</td>\n",
       "      <td>51.916972</td>\n",
       "      <td>0.994885</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.994885</td>\n",
       "      <td>49.239218</td>\n",
       "      <td>0.996074</td>\n",
       "      <td>99.29</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>225992</td>\n",
       "      <td>GCF_002056725.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Comamonas kerstersii</td>\n",
       "      <td>8943</td>\n",
       "      <td>True</td>\n",
       "      <td>Comamonas kerstersii</td>\n",
       "      <td>0.733389</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.733389</td>\n",
       "      <td>67.342723</td>\n",
       "      <td>0.694253</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.694253</td>\n",
       "      <td>39.381015</td>\n",
       "      <td>0.965437</td>\n",
       "      <td>68.22</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>237258</td>\n",
       "      <td>GCF_003860565.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Cloacibacterium normanense</td>\n",
       "      <td>NRS-1</td>\n",
       "      <td>True</td>\n",
       "      <td>Cloacibacterium normanense</td>\n",
       "      <td>0.908765</td>\n",
       "      <td>0.999990</td>\n",
       "      <td>0.908774</td>\n",
       "      <td>12.638790</td>\n",
       "      <td>0.914433</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>0.914480</td>\n",
       "      <td>10.808966</td>\n",
       "      <td>0.985893</td>\n",
       "      <td>90.80</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>853</td>\n",
       "      <td>GCF_002586945.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Faecalibacterium prausnitzii</td>\n",
       "      <td>Indica</td>\n",
       "      <td>True</td>\n",
       "      <td>Faecalibacterium prausnitzii</td>\n",
       "      <td>0.965445</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.965445</td>\n",
       "      <td>47.055026</td>\n",
       "      <td>0.974128</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.974128</td>\n",
       "      <td>41.691542</td>\n",
       "      <td>0.989486</td>\n",
       "      <td>96.90</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>470</td>\n",
       "      <td>GCF_008632635.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>K09-14</td>\n",
       "      <td>True</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>0.575722</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.575722</td>\n",
       "      <td>27.502425</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>166486</td>\n",
       "      <td>GCF_900537995.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Roseburia intestinalis L1-82</td>\n",
       "      <td>L1-82</td>\n",
       "      <td>True</td>\n",
       "      <td>Roseburia intestinalis</td>\n",
       "      <td>0.613842</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>0.613843</td>\n",
       "      <td>21.759672</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>410072</td>\n",
       "      <td>GCF_016904155.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Coprococcus comes</td>\n",
       "      <td>FDAARGOS_1339</td>\n",
       "      <td>True</td>\n",
       "      <td>Coprococcus comes</td>\n",
       "      <td>0.928913</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.928913</td>\n",
       "      <td>22.712329</td>\n",
       "      <td>0.931943</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.931943</td>\n",
       "      <td>16.955882</td>\n",
       "      <td>0.992883</td>\n",
       "      <td>92.86</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>876</td>\n",
       "      <td>GCF_017815575.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Desulfovibrio desulfuricans</td>\n",
       "      <td>L4</td>\n",
       "      <td>True</td>\n",
       "      <td>Desulfovibrio desulfuricans</td>\n",
       "      <td>0.871071</td>\n",
       "      <td>0.999988</td>\n",
       "      <td>0.871081</td>\n",
       "      <td>13.015867</td>\n",
       "      <td>0.872297</td>\n",
       "      <td>0.999341</td>\n",
       "      <td>0.872873</td>\n",
       "      <td>8.397204</td>\n",
       "      <td>0.979759</td>\n",
       "      <td>86.40</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>216816</td>\n",
       "      <td>GCF_000196555.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Bifidobacterium longum subsp. longum JCM 1217</td>\n",
       "      <td>JCM 1217</td>\n",
       "      <td>True</td>\n",
       "      <td>Bifidobacterium longum</td>\n",
       "      <td>0.944944</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.944944</td>\n",
       "      <td>36.286933</td>\n",
       "      <td>0.958789</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.958789</td>\n",
       "      <td>22.268293</td>\n",
       "      <td>0.997280</td>\n",
       "      <td>95.75</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>158836</td>\n",
       "      <td>GCF_019048625.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Enterobacter hormaechei</td>\n",
       "      <td>FDAARGOS 1435</td>\n",
       "      <td>True</td>\n",
       "      <td>Enterobacter hormaechei</td>\n",
       "      <td>0.913310</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.913310</td>\n",
       "      <td>27.952961</td>\n",
       "      <td>0.882106</td>\n",
       "      <td>0.999063</td>\n",
       "      <td>0.882933</td>\n",
       "      <td>7.904936</td>\n",
       "      <td>0.986232</td>\n",
       "      <td>87.68</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1463165</td>\n",
       "      <td>GCF_016415705.1</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>True</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Klebsiella quasipneumoniae</td>\n",
       "      <td>KqPF26</td>\n",
       "      <td>True</td>\n",
       "      <td>Klebsiella quasipneumoniae</td>\n",
       "      <td>0.968364</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.968364</td>\n",
       "      <td>30.549331</td>\n",
       "      <td>0.936350</td>\n",
       "      <td>0.999623</td>\n",
       "      <td>0.936703</td>\n",
       "      <td>8.418771</td>\n",
       "      <td>0.990976</td>\n",
       "      <td>93.25</td>\n",
       "      <td>Present</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Taxonomy ID Assembly Accession ID  \\\n",
       "0         40324       GCF_900475405.1   \n",
       "1           645       GCF_012931585.1   \n",
       "2           651       GCF_020423125.1   \n",
       "3         33039       GCA_023674525.1   \n",
       "4          1680       GCF_003030905.1   \n",
       "5         40214       GCF_003335165.1   \n",
       "6         40215       GCF_018336855.1   \n",
       "7       2479767       GCF_005848555.1   \n",
       "8       2743575       GCF_013374795.1   \n",
       "9          1307       GCF_000026745.1   \n",
       "10       649756       GCF_001998765.1   \n",
       "11          285       GCF_014076415.1   \n",
       "12          287       GCF_000006765.1   \n",
       "13          546       GCF_003812345.1   \n",
       "14        28198       GCF_003660105.1   \n",
       "15          562       GCF_000008865.2   \n",
       "16        74426       GCF_002736145.1   \n",
       "17       165179       GCF_020735445.1   \n",
       "18          573       GCF_000240185.1   \n",
       "19       239935       GCF_009731575.1   \n",
       "20       225992       GCF_002056725.1   \n",
       "21       237258       GCF_003860565.1   \n",
       "22          853       GCF_002586945.1   \n",
       "23          470       GCF_008632635.1   \n",
       "24       166486       GCF_900537995.1   \n",
       "25       410072       GCF_016904155.1   \n",
       "26          876       GCF_017815575.1   \n",
       "27       216816       GCF_000196555.1   \n",
       "28       158836       GCF_019048625.1   \n",
       "29      1463165       GCF_016415705.1   \n",
       "\n",
       "                                      Source Database  Is Representative  \\\n",
       "0                                         NCBI RefSeq               True   \n",
       "1                                         NCBI RefSeq               True   \n",
       "2                                         NCBI RefSeq               True   \n",
       "3                                                NCBI              False   \n",
       "4                                         NCBI RefSeq               True   \n",
       "5                                         NCBI RefSeq              False   \n",
       "6                                         NCBI RefSeq               True   \n",
       "7                                         NCBI RefSeq              False   \n",
       "8                                         NCBI RefSeq              False   \n",
       "9                                         NCBI RefSeq               True   \n",
       "10                                        NCBI RefSeq              False   \n",
       "11                                        NCBI RefSeq              False   \n",
       "12                           PathoGenesis Corporation               True   \n",
       "13                                        NCBI RefSeq               True   \n",
       "14                                        NCBI RefSeq               True   \n",
       "15                                               GIRC               True   \n",
       "16                                        NCBI RefSeq               True   \n",
       "17                                        NCBI RefSeq               True   \n",
       "18   Bioinformatics & Microbial Genomics, State Ke...               True   \n",
       "19                                        NCBI RefSeq               True   \n",
       "20                                        NCBI RefSeq               True   \n",
       "21                                        NCBI RefSeq               True   \n",
       "22                                        NCBI RefSeq              False   \n",
       "23                                        NCBI RefSeq               True   \n",
       "24                                        NCBI RefSeq               True   \n",
       "25                                        NCBI RefSeq               True   \n",
       "26                                        NCBI RefSeq               True   \n",
       "27                                        NCBI RefSeq               True   \n",
       "28                                        NCBI RefSeq               True   \n",
       "29                                        NCBI RefSeq               True   \n",
       "\n",
       "     Assembly Level                             Organism of Assembly  \\\n",
       "0   Complete Genome                     Stenotrophomonas maltophilia   \n",
       "1   Complete Genome                            Aeromonas salmonicida   \n",
       "2   Complete Genome                                  Aeromonas media   \n",
       "3        Chromosome                           [Ruminococcus] torques   \n",
       "4   Complete Genome                     Bifidobacterium adolescentis   \n",
       "5   Complete Genome                          Acinetobacter johnsonii   \n",
       "6   Complete Genome                              Acinetobacter junii   \n",
       "7   Complete Genome                              Blautia sp. SC05B48   \n",
       "8   Complete Genome                        Acinetobacter sp. NEB 394   \n",
       "9   Complete Genome                         Streptococcus suis BM407   \n",
       "10  Complete Genome                              Anaerostipes hadrus   \n",
       "11  Complete Genome                           Comamonas testosteroni   \n",
       "12  Complete Genome                      Pseudomonas aeruginosa PAO1   \n",
       "13  Complete Genome                             Citrobacter freundii   \n",
       "14  Complete Genome           Aliarcobacter cryaerophilus ATCC 43158   \n",
       "15  Complete Genome              Escherichia coli O157:H7 str. Sakai   \n",
       "16  Complete Genome                          Collinsella aerofaciens   \n",
       "17  Complete Genome                       Prevotella copri DSM 18205   \n",
       "18  Complete Genome  Klebsiella pneumoniae subsp. pneumoniae HS11286   \n",
       "19  Complete Genome                          Akkermansia muciniphila   \n",
       "20  Complete Genome                             Comamonas kerstersii   \n",
       "21  Complete Genome                       Cloacibacterium normanense   \n",
       "22  Complete Genome                     Faecalibacterium prausnitzii   \n",
       "23  Complete Genome                          Acinetobacter baumannii   \n",
       "24  Complete Genome                     Roseburia intestinalis L1-82   \n",
       "25  Complete Genome                                Coprococcus comes   \n",
       "26  Complete Genome                      Desulfovibrio desulfuricans   \n",
       "27  Complete Genome    Bifidobacterium longum subsp. longum JCM 1217   \n",
       "28  Complete Genome                          Enterobacter hormaechei   \n",
       "29  Complete Genome                       Klebsiella quasipneumoniae   \n",
       "\n",
       "                        Strain  Downloaded                       Species  \\\n",
       "0                    NCTC10258        True  Stenotrophomonas maltophilia   \n",
       "1                      SRW-OG1        True         Aeromonas salmonicida   \n",
       "2                        TR3_1        True               Aeromonas media   \n",
       "3                  NB2A-14-FMU        True        [Ruminococcus] torques   \n",
       "4                         1-11        True  Bifidobacterium adolescentis   \n",
       "5                       LXL_C1        True       Acinetobacter johnsonii   \n",
       "6                          YR7        True           Acinetobacter junii   \n",
       "7                      SC05B48        True           Blautia sp. SC05B48   \n",
       "8                      NEB 394        True     Acinetobacter sp. NEB 394   \n",
       "9                        BM407        True            Streptococcus suis   \n",
       "10                        BPB5        True           Anaerostipes hadrus   \n",
       "11                       T5-67        True        Comamonas testosteroni   \n",
       "12                        PAO1        True        Pseudomonas aeruginosa   \n",
       "13                FDAARGOS_549        True          Citrobacter freundii   \n",
       "14                  ATCC 43158        True   Aliarcobacter cryaerophilus   \n",
       "15  Sakai substr. RIMD 0509952        True              Escherichia coli   \n",
       "16                      indica        True       Collinsella aerofaciens   \n",
       "17               FDAARGOS_1573        True              Prevotella copri   \n",
       "18                     HS11286        True         Klebsiella pneumoniae   \n",
       "19                   JCM 30893        True       Akkermansia muciniphila   \n",
       "20                        8943        True          Comamonas kerstersii   \n",
       "21                       NRS-1        True    Cloacibacterium normanense   \n",
       "22                      Indica        True  Faecalibacterium prausnitzii   \n",
       "23                      K09-14        True       Acinetobacter baumannii   \n",
       "24                       L1-82        True        Roseburia intestinalis   \n",
       "25               FDAARGOS_1339        True             Coprococcus comes   \n",
       "26                          L4        True   Desulfovibrio desulfuricans   \n",
       "27                    JCM 1217        True        Bifidobacterium longum   \n",
       "28               FDAARGOS 1435        True       Enterobacter hormaechei   \n",
       "29                      KqPF26        True    Klebsiella quasipneumoniae   \n",
       "\n",
       "    Breadth Coverage  Expected Coverage  Coverage Score  Depth Coverage  \\\n",
       "0           0.839573           1.000000        0.839573       34.563671   \n",
       "1           0.841355           0.999998        0.841356       15.705711   \n",
       "2           0.871588           1.000000        0.871588       18.324055   \n",
       "3           0.638248           0.999997        0.638250       20.186668   \n",
       "4           0.969784           1.000000        0.969784       49.626209   \n",
       "5           0.955982           1.000000        0.955982       79.342853   \n",
       "6           0.703081           1.000000        0.703081       32.917163   \n",
       "7           0.939978           1.000000        0.939978       28.135163   \n",
       "8           0.959521           1.000000        0.959521       79.441439   \n",
       "9           0.641112           0.999732        0.641284       12.830374   \n",
       "10          0.955947           1.000000        0.955947       23.247111   \n",
       "11          0.729633           1.000000        0.729633       53.984637   \n",
       "12          0.750096           1.000000        0.750096       23.007642   \n",
       "13          0.916487           1.000000        0.916487       26.468455   \n",
       "14          0.932623           1.000000        0.932623       20.159384   \n",
       "15          0.896436           1.000000        0.896436       42.209379   \n",
       "16          0.944198           1.000000        0.944198       32.480432   \n",
       "17          0.741253           0.995963        0.744258        7.436305   \n",
       "18          0.951345           1.000000        0.951345       30.103833   \n",
       "19          0.983008           1.000000        0.983008       51.916972   \n",
       "20          0.733389           1.000000        0.733389       67.342723   \n",
       "21          0.908765           0.999990        0.908774       12.638790   \n",
       "22          0.965445           1.000000        0.965445       47.055026   \n",
       "23          0.575722           1.000000        0.575722       27.502425   \n",
       "24          0.613842           0.999998        0.613843       21.759672   \n",
       "25          0.928913           1.000000        0.928913       22.712329   \n",
       "26          0.871071           0.999988        0.871081       13.015867   \n",
       "27          0.944944           1.000000        0.944944       36.286933   \n",
       "28          0.913310           1.000000        0.913310       27.952961   \n",
       "29          0.968364           1.000000        0.968364       30.549331   \n",
       "\n",
       "         BC2       EC2       CS2        DC2  Consensus ANI  \\\n",
       "0   0.814941  1.000000  0.814941  23.577827       0.981539   \n",
       "1   0.718890  0.986014  0.729087   5.939255       0.987220   \n",
       "2   0.842731  0.998330  0.844141   7.588203       0.983200   \n",
       "3   0.000000  0.000000  0.000000   0.000000       0.000000   \n",
       "4   0.970498  1.000000  0.970498  39.889433       0.995705   \n",
       "5   0.974916  1.000000  0.974916  33.517974       0.989357   \n",
       "6   0.671086  0.999798  0.671221  12.675992       0.955793   \n",
       "7   0.954549  1.000000  0.954549  23.922561       0.983677   \n",
       "8   0.990409  1.000000  0.990409  56.647163       0.995534   \n",
       "9   0.000000  0.000000  0.000000   0.000000       0.000000   \n",
       "10  0.973811  1.000000  0.973811  19.278097       0.995118   \n",
       "11  0.700041  1.000000  0.700041  40.119623       0.968364   \n",
       "12  0.704004  0.999804  0.704142  12.128048       0.989756   \n",
       "13  0.892323  0.998064  0.894054   7.000806       0.985482   \n",
       "14  0.945659  1.000000  0.945659  18.186905       0.977474   \n",
       "15  0.916219  1.000000  0.916219  33.477121       0.990711   \n",
       "16  0.954109  1.000000  0.954109  26.068877       0.980516   \n",
       "17  0.746695  0.986719  0.756745   5.787401       0.986355   \n",
       "18  0.958143  0.999979  0.958164  11.217542       0.995492   \n",
       "19  0.994885  1.000000  0.994885  49.239218       0.996074   \n",
       "20  0.694253  1.000000  0.694253  39.381015       0.965437   \n",
       "21  0.914433  0.999949  0.914480  10.808966       0.985893   \n",
       "22  0.974128  1.000000  0.974128  41.691542       0.989486   \n",
       "23  0.000000  0.000000  0.000000   0.000000       0.000000   \n",
       "24  0.000000  0.000000  0.000000   0.000000       0.000000   \n",
       "25  0.931943  1.000000  0.931943  16.955882       0.992883   \n",
       "26  0.872297  0.999341  0.872873   8.397204       0.979759   \n",
       "27  0.958789  1.000000  0.958789  22.268293       0.997280   \n",
       "28  0.882106  0.999063  0.882933   7.904936       0.986232   \n",
       "29  0.936350  0.999623  0.936703   8.418771       0.990976   \n",
       "\n",
       "    Combined CS2 and ANI (Sqrt(ANI)xCS2x100) Presence/Absence  \n",
       "0                                      80.74          Present  \n",
       "1                                      72.44          Present  \n",
       "2                                      83.70          Present  \n",
       "3                                       0.00           Absent  \n",
       "4                                      96.84          Present  \n",
       "5                                      96.97          Present  \n",
       "6                                      65.62          Present  \n",
       "7                                      94.67          Present  \n",
       "8                                      98.82          Present  \n",
       "9                                       0.00           Absent  \n",
       "10                                     97.14          Present  \n",
       "11                                     68.89          Present  \n",
       "12                                     70.05          Present  \n",
       "13                                     88.75          Present  \n",
       "14                                     93.49          Present  \n",
       "15                                     91.20          Present  \n",
       "16                                     94.48          Present  \n",
       "17                                     75.16          Present  \n",
       "18                                     95.60          Present  \n",
       "19                                     99.29          Present  \n",
       "20                                     68.22          Present  \n",
       "21                                     90.80          Present  \n",
       "22                                     96.90          Present  \n",
       "23                                      0.00           Absent  \n",
       "24                                      0.00           Absent  \n",
       "25                                     92.86          Present  \n",
       "26                                     86.40          Present  \n",
       "27                                     95.75          Present  \n",
       "28                                     87.68          Present  \n",
       "29                                     93.25          Present  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "downloaded_assemblies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "335180ef-eedf-41f7-848e-336ccfa0be2b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
